// @quri/squiggle-lang/dist/index.js
import { SqProject } from "./public/SqProject/index.js";
export { SqTextAreaInput, SqCheckboxInput, SqSelectInput, SqTextInput, } from "./public/SqValue/SqInput.js";
import { SqLambdaValue, SqNumberValue, SqStringValue, } from "./public/SqValue/index.js";
export { SqCompileError, SqFrame, SqOtherError, SqRuntimeError, } from "./public/SqError.js";
export { SqDistributionError } from "./public/SqValue/SqDistribution/SqDistributionError.js";
export { SqAbstractDistribution, SqDistributionTag, SqPointSetDistribution, SqSampleSetDistribution, SqSymbolicDistribution, } from "./public/SqValue/SqDistribution/index.js";
export { SqLambda } from "./public/SqValue/SqLambda.js";
export { SqDictValue } from "./public/SqValue/index.js";
export { SqDistFnPlot, SqDistributionsPlot, SqNumericFnPlot, SqRelativeValuesPlot, SqScatterPlot, } from "./public/SqValue/SqPlot.js";
export { SqTableChart } from "./public/SqValue/SqTableChart.js";
export { SqCalculator } from "./public/SqValue/SqCalculator.js";
export { SqDict } from "./public/SqValue/SqDict.js";
export { SqLinearScale, SqLogScale, SqPowerScale, SqSymlogScale, } from "./public/SqValue/SqScale.js";
export { SqValuePath } from "./public/SqValuePath.js";
export { parse } from "./public/parse.js";
export { fmap as resultMap } from "./utility/result.js";
export { defaultEnv as defaultEnvironment } from "./dist/env.js";
export { SqLambdaValue, SqNumberValue, SqProject, SqStringValue, };
export async function run(code, options) {
    const project = SqProject.create();
    project.setSource("main", code);
    if (options?.environment) {
        project.setEnvironment(options.environment);
    }
    await project.run("main");
    return project.getOutput("main");
}
export function sq(strings, ...rest) {
    if (rest.length) {
        throw new Error("Extrapolation in sq`` template literals is forbidden");
    }
    return strings.join("");
}
