// @quri/squiggle-lang/dist/operationError.js
export class OperationError {
    constructor() {
        this.type = "OperationError";
    }
}
export class DivisionByZeroError extends OperationError {
    toString() {
        return "Cannot divide by zero";
    }
}
export class ComplexNumberError extends OperationError {
    toString() {
        return "Operation returned complex result";
    }
}
export class InfinityError extends OperationError {
    toString() {
        return "Operation returned positive infinity";
    }
}
export class NegativeInfinityError extends OperationError {
    toString() {
        return "Operation returned negative infinity";
    }
}
export class SampleMapNeedsNtoNFunction extends OperationError {
    toString() {
        return "SampleMap needs a function that converts a number to a number";
    }
}
export class PdfInvalidError extends OperationError {
    toString() {
        return "This Pdf is invalid";
    }
}
export class OtherOperationError extends OperationError {
    constructor(value) {
        super();
        this.value = value;
    }
    toString() {
        return this.value;
    }
}
