// @quri/squiggle-lang/dist/operation.js
import { Ok } from "./utility/result.js";
import * as Result from "./utility/result.js";
import { ComplexNumberError, DivisionByZeroError, NegativeInfinityError, } from "./operationError.js";
export const Convolution = {
    fromAlgebraicOperation(op) {
        if (op === "Add" || op === "Subtract" || op === "Multiply") {
            return op;
        }
        return undefined;
    },
    canDoAlgebraicOperation(op) {
        return Convolution.fromAlgebraicOperation(op) !== undefined;
    },
    toFn(t) {
        switch (t) {
            case "Add":
                return (a, b) => a + b;
            case "Subtract":
                return (a, b) => a - b;
            case "Multiply":
                return (a, b) => a * b;
            default:
                throw new Error("This should never happen");
        }
    },
};
const power = (a, b) => {
    const result = a ** b;
    if (Number.isNaN(result)) {
        if (Number.isNaN(a) || Number.isNaN(b)) {
            return Ok(result);
        }
        return Result.Err(new ComplexNumberError());
    }
    return Ok(result);
};
const add = (a, b) => Ok(a + b);
const subtract = (a, b) => Ok(a - b);
const multiply = (a, b) => Ok(a * b);
const divide = (a, b) => {
    if (b !== 0) {
        return Ok(a / b);
    }
    else {
        return Result.Err(new DivisionByZeroError());
    }
};
const logarithm = (a, b) => {
    if (b === 1) {
        return Result.Err(new DivisionByZeroError());
    }
    else if (b === 0) {
        return Ok(0);
    }
    else if (a > 0 && b > 0) {
        return Ok(Math.log(a) / Math.log(b));
    }
    else if (a === 0) {
        return Result.Err(new NegativeInfinityError());
    }
    else {
        return Result.Err(new ComplexNumberError());
    }
};
const buildLogarithmWithThreshold = (threshold) => {
    const fn = (a, b) => {
        if (a < threshold) {
            return Ok(0);
        }
        else {
            return logarithm(a, b);
        }
    };
    return fn;
};
export const Algebraic = {
    toFn(x) {
        if (x === "Add") {
            return add;
        }
        else if (x === "Subtract") {
            return subtract;
        }
        else if (x === "Multiply") {
            return multiply;
        }
        else if (x === "Power") {
            return power;
        }
        else if (x === "Divide") {
            return divide;
        }
        else if (x === "Logarithm") {
            return logarithm;
        }
        else if (x.NAME === "LogarithmWithThreshold") {
            return buildLogarithmWithThreshold(x.VAL);
        }
        else {
            throw new Error(`Unknown operation ${x}`);
        }
    },
    toString(x) {
        if (x === "Add") {
            return "+";
        }
        else if (x === "Subtract") {
            return "-";
        }
        else if (x === "Multiply") {
            return "*";
        }
        else if (x === "Power") {
            return "**";
        }
        else if (x === "Divide") {
            return "/";
        }
        else if (x === "Logarithm") {
            return "log";
        }
        else if (x.NAME === "LogarithmWithThreshold") {
            return "log";
        }
        else {
            throw new Error(`Unknown operation ${x}`);
        }
    },
};
export const Scale = {
    toFn(x) {
        if (x === "Multiply") {
            return multiply;
        }
        else if (x === "Divide") {
            return divide;
        }
        else if (x === "Power") {
            return power;
        }
        else if (x === "Logarithm") {
            return logarithm;
        }
        else if (x.NAME === "LogarithmWithThreshold") {
            return buildLogarithmWithThreshold(x.VAL);
        }
        else {
            throw new Error(`Unknown scale operation ${x}`);
        }
    },
    toIntegralSumCacheFn(x) {
        if (x === "Multiply") {
            return (a, b) => a * b;
        }
        else if (x === "Divide") {
            return (a, b) => a / b;
        }
        else {
            return undefined;
        }
    },
    toIntegralCacheFn(x) {
        return undefined;
    },
};
