// @quri/squiggle-lang/src/rescript/FR/FR_List.bs.js
'use strict';

var Belt_Array = require("@rescript/std/lib/js/belt_Array.js");
var E_A$QuriSquiggleLang = require("../Utility/E/E_A.bs.js");
var E_O$QuriSquiggleLang = require("../Utility/E/E_O.bs.js");
var E_R$QuriSquiggleLang = require("../Utility/E/E_R.bs.js");
var E_Float$QuriSquiggleLang = require("../Utility/E/E_Float.bs.js");
var Reducer_Lambda$QuriSquiggleLang = require("../Reducer/Reducer_Lambda.bs.js");
var FunctionRegistry_Core$QuriSquiggleLang = require("../FunctionRegistry/FunctionRegistry_Core.bs.js");
var FunctionRegistry_Helpers$QuriSquiggleLang = require("../FunctionRegistry/FunctionRegistry_Helpers.bs.js");

var nameSpace = "List";

function length(v) {
  return {
          TAG: 6,
          _0: E_A$QuriSquiggleLang.length(v),
          [Symbol.for("name")]: "IEvNumber"
        };
}

function makeFromNumber(n, value) {
  return {
          TAG: 0,
          _0: Belt_Array.make(E_Float$QuriSquiggleLang.toInt(n), value),
          [Symbol.for("name")]: "IEvArray"
        };
}

function upTo(low, high) {
  return {
          TAG: 0,
          _0: E_A$QuriSquiggleLang.fmap(E_A$QuriSquiggleLang.Floats.range(low, high, E_Float$QuriSquiggleLang.toInt(high - low + 1.0)), FunctionRegistry_Helpers$QuriSquiggleLang.Wrappers.evNumber),
          [Symbol.for("name")]: "IEvArray"
        };
}

function first(v) {
  return E_O$QuriSquiggleLang.toResult(E_A$QuriSquiggleLang.first(v), "No first element");
}

function last(v) {
  return E_O$QuriSquiggleLang.toResult(E_A$QuriSquiggleLang.last(v), "No last element");
}

function reverse(array) {
  return {
          TAG: 0,
          _0: Belt_Array.reverse(array),
          [Symbol.for("name")]: "IEvArray"
        };
}

function map(array, eLambdaValue, context, reducer) {
  return FunctionRegistry_Helpers$QuriSquiggleLang.Wrappers.evArray(E_A$QuriSquiggleLang.fmap(array, (function (elem) {
                    return Reducer_Lambda$QuriSquiggleLang.doLambdaCall(eLambdaValue, [elem], context, reducer);
                  })));
}

function reduce(aValueArray, initialValue, aLambdaValue, context, reducer) {
  return E_A$QuriSquiggleLang.reduce(aValueArray, initialValue, (function (acc, elem) {
                return Reducer_Lambda$QuriSquiggleLang.doLambdaCall(aLambdaValue, [
                            acc,
                            elem
                          ], context, reducer);
              }));
}

function reduceReverse(aValueArray, initialValue, aLambdaValue, context, reducer) {
  return E_A$QuriSquiggleLang.reduceReverse(aValueArray, initialValue, (function (acc, elem) {
                return Reducer_Lambda$QuriSquiggleLang.doLambdaCall(aLambdaValue, [
                            acc,
                            elem
                          ], context, reducer);
              }));
}

function filter(aValueArray, aLambdaValue, context, reducer) {
  return FunctionRegistry_Helpers$QuriSquiggleLang.Wrappers.evArray(aValueArray.filter(function (elem) {
                  var result = Reducer_Lambda$QuriSquiggleLang.doLambdaCall(aLambdaValue, [elem], context, reducer);
                  if (typeof result === "number" || !(result.TAG === /* IEvBool */1 && result._0)) {
                    return false;
                  } else {
                    return true;
                  }
                }));
}

var Internals = {
  length: length,
  makeFromNumber: makeFromNumber,
  upTo: upTo,
  first: first,
  last: last,
  reverse: reverse,
  map: map,
  reduce: reduce,
  reduceReverse: reduceReverse,
  filter: filter
};

var library = [
  FunctionRegistry_Core$QuriSquiggleLang.$$Function.make("length", nameSpace, true, [FunctionRegistry_Core$QuriSquiggleLang.FnDefinition.make("length", [{
                TAG: 2,
                _0: /* FRTypeAny */9,
                [Symbol.for("name")]: "FRTypeArray"
              }], (function (inputs, param, param$1) {
                if (inputs.length !== 1) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                var array = inputs[0];
                if (typeof array === "number" || array.TAG !== /* IEvArray */0) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                } else {
                  return {
                          TAG: 0,
                          _0: {
                            TAG: 6,
                            _0: E_A$QuriSquiggleLang.length(array._0),
                            [Symbol.for("name")]: "IEvNumber"
                          },
                          [Symbol.for("name")]: "Ok"
                        };
                }
              }), undefined)], ["List.length([1,4,5])"], /* EvtNumber */6, undefined, undefined, undefined),
  FunctionRegistry_Core$QuriSquiggleLang.$$Function.make("make", nameSpace, true, [FunctionRegistry_Core$QuriSquiggleLang.FnDefinition.make("make", [
              /* FRTypeNumber */0,
              /* FRTypeAny */9
            ], (function (inputs, param, param$1) {
                if (inputs.length !== 2) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                var number = inputs[0];
                if (typeof number === "number") {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                if (number.TAG !== /* IEvNumber */6) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                var value = inputs[1];
                return {
                        TAG: 0,
                        _0: makeFromNumber(number._0, value),
                        [Symbol.for("name")]: "Ok"
                      };
              }), undefined)], ["List.make(2, \"testValue\")"], /* EvtArray */0, undefined, undefined, undefined),
  FunctionRegistry_Core$QuriSquiggleLang.$$Function.make("upTo", nameSpace, true, [FunctionRegistry_Core$QuriSquiggleLang.FnDefinition.make("upTo", [
              /* FRTypeNumber */0,
              /* FRTypeNumber */0
            ], (function (inputs, param, param$1) {
                if (inputs.length !== 2) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                var low = inputs[0];
                if (typeof low === "number") {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                if (low.TAG !== /* IEvNumber */6) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                var high = inputs[1];
                if (typeof high === "number" || high.TAG !== /* IEvNumber */6) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                } else {
                  return {
                          TAG: 0,
                          _0: upTo(low._0, high._0),
                          [Symbol.for("name")]: "Ok"
                        };
                }
              }), undefined)], ["List.upTo(1,4)"], /* EvtArray */0, undefined, undefined, undefined),
  FunctionRegistry_Core$QuriSquiggleLang.$$Function.make("first", nameSpace, true, [FunctionRegistry_Core$QuriSquiggleLang.FnDefinition.make("first", [{
                TAG: 2,
                _0: /* FRTypeAny */9,
                [Symbol.for("name")]: "FRTypeArray"
              }], (function (inputs, param, param$1) {
                if (inputs.length !== 1) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                var array = inputs[0];
                if (typeof array === "number" || array.TAG !== /* IEvArray */0) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                } else {
                  return E_R$QuriSquiggleLang.errMap(E_O$QuriSquiggleLang.toResult(E_A$QuriSquiggleLang.first(array._0), "No first element"), FunctionRegistry_Helpers$QuriSquiggleLang.wrapError);
                }
              }), undefined)], ["List.first([1,4,5])"], undefined, undefined, undefined, undefined),
  FunctionRegistry_Core$QuriSquiggleLang.$$Function.make("last", nameSpace, true, [FunctionRegistry_Core$QuriSquiggleLang.FnDefinition.make("last", [{
                TAG: 2,
                _0: /* FRTypeAny */9,
                [Symbol.for("name")]: "FRTypeArray"
              }], (function (inputs, param, param$1) {
                if (inputs.length !== 1) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                var array = inputs[0];
                if (typeof array === "number" || array.TAG !== /* IEvArray */0) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                } else {
                  return E_R$QuriSquiggleLang.errMap(E_O$QuriSquiggleLang.toResult(E_A$QuriSquiggleLang.last(array._0), "No last element"), FunctionRegistry_Helpers$QuriSquiggleLang.wrapError);
                }
              }), undefined)], ["List.last([1,4,5])"], undefined, undefined, undefined, undefined),
  FunctionRegistry_Core$QuriSquiggleLang.$$Function.make("reverse", nameSpace, false, [FunctionRegistry_Core$QuriSquiggleLang.FnDefinition.make("reverse", [{
                TAG: 2,
                _0: /* FRTypeAny */9,
                [Symbol.for("name")]: "FRTypeArray"
              }], (function (inputs, param, param$1) {
                if (inputs.length !== 1) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                var array = inputs[0];
                if (typeof array === "number" || array.TAG !== /* IEvArray */0) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                } else {
                  return {
                          TAG: 0,
                          _0: {
                            TAG: 0,
                            _0: Belt_Array.reverse(array._0),
                            [Symbol.for("name")]: "IEvArray"
                          },
                          [Symbol.for("name")]: "Ok"
                        };
                }
              }), undefined)], ["List.reverse([1,4,5])"], /* EvtArray */0, undefined, undefined, undefined),
  FunctionRegistry_Core$QuriSquiggleLang.$$Function.make("map", nameSpace, false, [FunctionRegistry_Core$QuriSquiggleLang.FnDefinition.make("map", [
              {
                TAG: 2,
                _0: /* FRTypeAny */9,
                [Symbol.for("name")]: "FRTypeArray"
              },
              /* FRTypeLambda */7
            ], (function (inputs, context, reducer) {
                if (inputs.length !== 2) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                var array = inputs[0];
                if (typeof array === "number") {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                if (array.TAG !== /* IEvArray */0) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                var lambda = inputs[1];
                if (typeof lambda === "number" || lambda.TAG !== /* IEvLambda */5) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                } else {
                  return {
                          TAG: 0,
                          _0: map(array._0, lambda._0, context, reducer),
                          [Symbol.for("name")]: "Ok"
                        };
                }
              }), undefined)], ["List.map([1,4,5], {|x| x+1})"], /* EvtArray */0, undefined, undefined, undefined),
  FunctionRegistry_Core$QuriSquiggleLang.$$Function.make("reduce", nameSpace, false, [FunctionRegistry_Core$QuriSquiggleLang.FnDefinition.make("reduce", [
              {
                TAG: 2,
                _0: /* FRTypeAny */9,
                [Symbol.for("name")]: "FRTypeArray"
              },
              /* FRTypeAny */9,
              /* FRTypeLambda */7
            ], (function (inputs, context, reducer) {
                if (inputs.length !== 3) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                var array = inputs[0];
                if (typeof array === "number") {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                if (array.TAG !== /* IEvArray */0) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                var initialValue = inputs[1];
                var lambda = inputs[2];
                if (typeof lambda === "number" || lambda.TAG !== /* IEvLambda */5) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                } else {
                  return {
                          TAG: 0,
                          _0: reduce(array._0, initialValue, lambda._0, context, reducer),
                          [Symbol.for("name")]: "Ok"
                        };
                }
              }), undefined)], ["List.reduce([1,4,5], 2, {|acc, el| acc+el})"], undefined, undefined, undefined, undefined),
  FunctionRegistry_Core$QuriSquiggleLang.$$Function.make("reduceReverse", nameSpace, false, [FunctionRegistry_Core$QuriSquiggleLang.FnDefinition.make("reduceReverse", [
              {
                TAG: 2,
                _0: /* FRTypeAny */9,
                [Symbol.for("name")]: "FRTypeArray"
              },
              /* FRTypeAny */9,
              /* FRTypeLambda */7
            ], (function (inputs, context, reducer) {
                if (inputs.length !== 3) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                var array = inputs[0];
                if (typeof array === "number") {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                if (array.TAG !== /* IEvArray */0) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                var initialValue = inputs[1];
                var lambda = inputs[2];
                if (typeof lambda === "number" || lambda.TAG !== /* IEvLambda */5) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                } else {
                  return {
                          TAG: 0,
                          _0: reduceReverse(array._0, initialValue, lambda._0, context, reducer),
                          [Symbol.for("name")]: "Ok"
                        };
                }
              }), undefined)], ["List.reduceReverse([1,4,5], 2, {|acc, el| acc-el})"], undefined, undefined, undefined, undefined),
  FunctionRegistry_Core$QuriSquiggleLang.$$Function.make("filter", nameSpace, false, [FunctionRegistry_Core$QuriSquiggleLang.FnDefinition.make("filter", [
              {
                TAG: 2,
                _0: /* FRTypeAny */9,
                [Symbol.for("name")]: "FRTypeArray"
              },
              /* FRTypeLambda */7
            ], (function (inputs, context, reducer) {
                if (inputs.length !== 2) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                var array = inputs[0];
                if (typeof array === "number") {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                if (array.TAG !== /* IEvArray */0) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                }
                var lambda = inputs[1];
                if (typeof lambda === "number" || lambda.TAG !== /* IEvLambda */5) {
                  return {
                          TAG: 1,
                          _0: FunctionRegistry_Helpers$QuriSquiggleLang.impossibleError,
                          [Symbol.for("name")]: "Error"
                        };
                } else {
                  return {
                          TAG: 0,
                          _0: filter(array._0, lambda._0, context, reducer),
                          [Symbol.for("name")]: "Ok"
                        };
                }
              }), undefined)], ["List.filter([1,4,5], {|x| x>3})"], undefined, undefined, undefined, undefined)
];

var requiresNamespace = true;

exports.nameSpace = nameSpace;
exports.requiresNamespace = requiresNamespace;
exports.Internals = Internals;
exports.library = library;
/* library Not a pure module */
