// @quri/squiggle-lang/dist/scripts/conversion/generators.js
export const generateFloatRange = (min, max) => Math.random() * (max - min) + min;
export const generateIntRange = (min, max) => Math.floor(generateFloatRange(min, max));
export const generateIntMin = (min) => generateIntRange(min, 100);
export const generateInt = () => generateIntMin(-100);
const generatePositiveInt = () => generateIntMin(1);
export const generateNormal = () => `normal(${generateInt()}, ${generatePositiveInt()})`;
export const generateBeta = () => `beta(${generatePositiveInt()}, ${generatePositiveInt()})`;
export const generateLognormal = () => `lognormal(${generateInt()}, ${generatePositiveInt()})`;
export const generateExponential = () => `exponential(${generatePositiveInt()})`;
export const generateUniform = () => {
    const a = generateInt();
    const b = generateIntMin(a + 1);
    return `uniform(${a}, ${b})`;
};
export const generateTriangular = () => {
    const a = generateInt();
    const b = generateIntMin(a + 1);
    const c = generateIntMin(b + 1);
    return `triangular(${a}, ${b}, ${c})`;
};
export const distributions = {
    normal: generateNormal,
    beta: generateBeta,
    lognormal: generateLognormal,
    exponential: generateExponential,
    triangular: generateTriangular,
    uniform: generateUniform,
};
