// @quri/squiggle-lang/dist/utility/DateTime.js
import { Ok } from "./result.js";
import * as Result from "./result.js";
export const Duration = {
    minute: 60 * 1000,
    hour: 60 * 60 * 1000,
    day: 24 * 60 * 60 * 1000,
    year: 24 * 60 * 60 * 1000 * 365.25,
    fromFloat: (f) => f,
    toFloat: (d) => d,
    fromMinutes: (h) => h * Duration.minute,
    fromHours: (h) => h * Duration.hour,
    fromDays: (d) => d * Duration.day,
    fromYears: (y) => y * Duration.year,
    toMinutes: (t) => t / Duration.minute,
    toHours: (t) => t / Duration.hour,
    toDays: (t) => t / Duration.day,
    toYears: (t) => t / Duration.year,
    toString(t) {
        const shouldPluralize = (f) => f !== 1.0;
        const display = (f, s) => `${f.toPrecision(3)} ${s}${shouldPluralize(f) ? "s" : ""}`;
        const abs = Math.abs(t);
        if (abs >= Duration.year) {
            return display(t / Duration.year, "year");
        }
        else if (abs >= Duration.day) {
            return display(t / Duration.day, "day");
        }
        else if (abs >= Duration.hour) {
            return display(t / Duration.hour, "hour");
        }
        else if (abs >= Duration.minute) {
            return display(t / Duration.minute, "minute");
        }
        else {
            return t.toFixed() + "ms";
        }
    },
    add: (t1, t2) => t1 + t2,
    subtract: (t1, t2) => t1 - t2,
    multiply: (t1, t2) => t1 * t2,
    divide: (t1, t2) => t1 / t2,
};
export const DateModule = {
    toString(d) {
        return d.toDateString();
    },
    fmap(t, fn) {
        return new Date(fn(t.getTime()));
    },
    subtract(t1, t2) {
        const [f1, f2] = [t1.getTime(), t2.getTime()];
        const diff = f1 - f2;
        if (diff < 0) {
            return Result.Err("Cannot subtract a date by one that is in its future");
        }
        else {
            return Ok(Duration.fromFloat(diff));
        }
    },
    addDuration(t, duration) {
        return DateModule.fmap(t, (t) => t + duration);
    },
    subtractDuration(t, duration) {
        return DateModule.fmap(t, (t) => t - duration);
    },
    makeWithYearInt(y) {
        if (y < 100) {
            return Result.Err("Year must be over 100");
        }
        else if (y > 200000) {
            return Result.Err("Year must be less than 200000");
        }
        else {
            return Ok(new Date(y, 0));
        }
    },
    makeFromYear(year) {
        const floor = Math.floor(year);
        return Result.fmap(DateModule.makeWithYearInt(floor), (earlyDate) => {
            const diff = year - floor;
            return DateModule.addDuration(earlyDate, diff * Duration.year);
        });
    },
};
export { DateModule as Date };
