// @quri/squiggle-lang/dist/utility/E_A.js
import { Ok } from "./result.js";
export const zip = (xs, ys) => {
    const lenX = xs.length;
    const lenY = ys.length;
    const len = lenX < lenY ? lenX : lenY;
    const s = new Array(len);
    for (let i = 0; i < len; i++) {
        s[i] = [xs[i], ys[i]];
    }
    return s;
};
export const accumulate = (items, fn) => {
    const len = items.length;
    if (len === 0)
        return [];
    const result = new Array(len);
    result[0] = items[0];
    for (let i = 1; i < len; i++) {
        result[i] = fn(items[i], result[i - 1]);
    }
    return result;
};
export const accumulateWithError = (items, fn) => {
    const len = items.length;
    if (len === 0)
        return Ok([]);
    const results = new Array(len);
    results[0] = items[0];
    for (let i = 1; i < len; i++) {
        const r = fn(items[i], results[i - 1]);
        if (!r.ok)
            return r;
        results[i] = r.value;
    }
    return Ok(results);
};
export const unzip = (items) => {
    const len = items.length;
    const a1 = new Array(len);
    const a2 = new Array(len);
    for (let i = 0; i < len; i++) {
        const [v1, v2] = items[i];
        a1[i] = v1;
        a2[i] = v2;
    }
    return [a1, a2];
};
export const pairwise = (items, fn) => {
    const result = [];
    for (let i = 1; i < items.length; i++) {
        result.push(fn(items[i - 1], items[i]));
    }
    return result;
};
export const pairwiseWithError = (items, fn) => {
    const result = [];
    for (let i = 1; i < items.length; i++) {
        const r = fn(items[i - 1], items[i]);
        if (!r.ok)
            return r;
        result.push(r.value);
    }
    return Ok(result);
};
export const makeBy = (n, fn) => {
    const result = [];
    for (let i = 0; i < n; i++) {
        result.push(fn(i));
    }
    return result;
};
export function shuffle(array) {
    const shuffledArray = [...array];
    for (let i = shuffledArray.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [shuffledArray[i], shuffledArray[j]] = [shuffledArray[j], shuffledArray[i]];
    }
    return shuffledArray;
}
export function isEqual(arr1, arr2) {
    if (arr1.length !== arr2.length) {
        return false;
    }
    for (let i = 0; i < arr1.length; i++) {
        if (arr1[i] !== arr2[i]) {
            return false;
        }
    }
    return true;
}
