// @quri/squiggle-lang/dist/utility/result.js
export function Ok(value) {
    return { ok: true, value: value };
}
export function Err(value) {
    return { ok: false, value: value };
}
export function fmap(r, fn) {
    if (r.ok) {
        return Ok(fn(r.value));
    }
    else {
        return r;
    }
}
export function fmap2(r, fn1, fn2) {
    if (r.ok) {
        return Ok(fn1(r.value));
    }
    else {
        return Err(fn2(r.value));
    }
}
export function errMap(r, fn) {
    if (r.ok) {
        return r;
    }
    else {
        return Err(fn(r.value));
    }
}
export function bind(r, fn) {
    if (r.ok) {
        return fn(r.value);
    }
    else {
        return r;
    }
}
export function merge(a, b) {
    if (!a.ok) {
        return a;
    }
    if (!b.ok) {
        return b;
    }
    return Ok([a.value, b.value]);
}
export function getError(r) {
    if (!r.ok) {
        return r.value;
    }
    else {
        return undefined;
    }
}
export function getExt(r) {
    if (r.ok) {
        return r.value;
    }
    else {
        throw r.value;
    }
}
