// @quri/squiggle-lang/src/rescript/Distributions/SampleSetDist/SampleSetDist.gen.tsx
/* TypeScript file generated from SampleSetDist.res by genType. */
/* eslint-disable import/first */


import type {operationError as Operation_operationError} from '../../../../src/rescript/Utility/Operation.gen';

import type {xyShape as PointSetTypes_xyShape} from '../../../../src/rescript/Distributions/PointSetDist/PointSetTypes.gen';

// tslint:disable-next-line:interface-over-type-literal
export type JS_distJs = { readonly continuousDist?: PointSetTypes_xyShape; readonly discreteDist: PointSetTypes_xyShape };

// tslint:disable-next-line:interface-over-type-literal
export type Error_sampleSetError = 
    "TooFewSamples"
  | { tag: "NonNumericInput"; value: string }
  | { tag: "OperationError"; value: Operation_operationError };

// tslint:disable-next-line:interface-over-type-literal
export type Error_pointsetConversionError = "TooFewSamplesForConversionToPointSet";

// tslint:disable-next-line:interface-over-type-literal
export type sampleSetError = 
    "TooFewSamples"
  | { tag: "NonNumericInput"; value: string }
  | { tag: "OperationError"; value: Operation_operationError };

// tslint:disable-next-line:interface-over-type-literal
export type pointsetConversionError = "TooFewSamplesForConversionToPointSet";

// tslint:disable-next-line:interface-over-type-literal
export type T_t = number[];

// tslint:disable-next-line:interface-over-type-literal
export type t = number[];
