// @quri/squiggle-lang/dist/dist/DistError.js
import { XYShapeError } from "../XYShape.js";
export const tooFewSamplesForConversionToPointSet = () => {
    return {
        type: "TooFewSamplesForConversionToPointSet",
    };
};
export const distOperationError = (operationError) => {
    return {
        type: "OperationError",
        value: operationError,
    };
};
export const notYetImplemented = () => ({
    type: "NotYetImplemented",
});
export const unreachableError = () => ({
    type: "Unreachable",
});
export const distributionVerticalShiftIsInvalid = () => ({
    type: "DistributionVerticalShiftIsInvalid",
});
export const operationDistError = (e) => ({
    type: "OperationError",
    value: e,
});
export const sparklineError = (e) => ({
    type: "SparklineError",
    message: e,
});
export const logarithmOfDistributionError = (e) => ({
    type: "LogarithmOfDistributionError",
    message: e,
});
export const requestedStrategyInvalidError = (e) => ({
    type: "RequestedStrategyInvalidError",
    message: e,
});
export const otherError = (e) => ({
    type: "OtherError",
    message: e,
});
export const argumentError = (e) => ({
    type: "ArgumentError",
    message: e,
});
export const xyShapeDistError = (e) => ({
    type: "XYShapeError",
    value: e,
});
export const distErrorToString = (e) => {
    switch (e.type) {
        case "NotYetImplemented":
            return "Function not yet implemented";
        case "Unreachable":
            return "Unreachable";
        case "DistributionVerticalShiftIsInvalid":
            return "Distribution vertical shift is invalid";
        case "ArgumentError":
            return `Argument Error ${e.message}`;
        case "LogarithmOfDistributionError":
            return `Logarithm of input error: ${e.message}`;
        case "NonNumericInput":
            return `Found a non-number in input: ${e.message}`;
        case "OperationError":
            return e.value.toString();
        case "TooFewSamplesForConversionToPointSet":
            return "Too Few Samples to convert to point set";
        case "SparklineError":
            return e.message;
        case "RequestedStrategyInvalidError":
            return `Requested strategy invalid: ${e.message}`;
        case "XYShapeError":
            return `XY Shape Error: ${XYShapeError.toString(e.value)}`;
        case "OtherError":
            return e.message;
        case "TooFewSamples":
            return "Too few samples when constructing sample set";
        default:
            return `Unknown error ${JSON.stringify(e)}`;
    }
};
