// @quri/squiggle-lang/dist/dist/PointSetDist.js
import * as Continuous from "../PointSet/Continuous.js";
import * as Mixed from "../PointSet/Mixed.js";
import * as Result from "../utility/result.js";
import { createSparkline } from "../utility/sparklines.js";
import { BaseDist } from "./BaseDist.js";
import { sparklineError } from "./DistError.js";
export class PointSetDist extends BaseDist {
    constructor(pointSet) {
        super();
        this.pointSet = pointSet;
    }
    toString() {
        return "Point Set Distribution";
    }
    max() {
        return this.pointSet.maxX();
    }
    min() {
        return this.pointSet.minX();
    }
    mean() {
        return this.pointSet.mean();
    }
    variance() {
        return Result.Ok(this.pointSet.variance());
    }
    downsample(n) {
        return new PointSetDist(this.pointSet.downsample(n));
    }
    sample() {
        const randomItem = Math.random();
        return this.pointSet.integralYtoX(randomItem);
    }
    sampleN(n) {
        const items = new Array(n).fill(0);
        for (let i = 0; i < n; i++) {
            items[i] = this.sample();
        }
        return items;
    }
    _isEqual(other) {
        return this.pointSet.isEqual(other.pointSet);
    }
    truncate(left, right) {
        if (left === undefined && right === undefined) {
            return Result.Ok(this);
        }
        return Result.Ok(new PointSetDist(this.pointSet.truncate(left, right).normalize()));
    }
    normalize() {
        return new PointSetDist(this.pointSet.normalize());
    }
    integralSum() {
        return this.pointSet.integralSum();
    }
    pdf(f) {
        const mixedPoint = this.pointSet.xToY(f);
        return Result.Ok(mixedPoint.continuous + mixedPoint.discrete);
    }
    inv(f) {
        return this.pointSet.integralYtoX(f);
    }
    cdf(f) {
        return this.pointSet.integralXtoY(f);
    }
    toPointSetDist() {
        return Result.Ok(this);
    }
    toSparkline(bucketCount) {
        const continuous = this.pointSet.toContinuous();
        if (!continuous) {
            return Result.Err(sparklineError("Cannot find the sparkline of a discrete distribution"));
        }
        const downsampled = continuous.downsampleEquallyOverX(bucketCount);
        return Result.Ok(createSparkline(Continuous.getShape(downsampled).ys));
    }
    mapYResult(fn, integralSumCacheFn, integralCacheFn) {
        return Result.fmap(this.pointSet.mapYResult(fn, integralSumCacheFn, integralCacheFn), (pointSet) => new PointSetDist(pointSet));
    }
}
export const combineAlgebraically = (op, t1, t2) => {
    return new PointSetDist(Mixed.combineAlgebraically(op, t1.pointSet, t2.pointSet));
};
export const combinePointwise = (t1, t2, fn, integralSumCachesFn = () => undefined, integralCachesFn = () => undefined) => {
    return Result.fmap(Mixed.combinePointwise(t1.pointSet, t2.pointSet, fn, integralSumCachesFn, integralCachesFn), (pointSet) => new PointSetDist(pointSet));
};
export function expectedConvolutionCost(d) {
    return (d.pointSet.toContinuous().xyShape.xs.length +
        d.pointSet.toDiscrete().xyShape.xs.length);
}
