// @quri/squiggle-lang/dist/dist/SymbolicDist.js
import { BaseDist } from "./BaseDist.js";
import * as Result from "../utility/result.js";
import sum from "lodash/sum.js";
import jstat from "jstat";
import * as E_A_Floats from "../utility/E_A_Floats.js";
import * as XYShape from "../XYShape.js";
import * as magicNumbers from "../magicNumbers.js";
import * as Operation from "../operation.js";
import { PointSetDist } from "./PointSetDist.js";
import { Ok } from "../utility/result.js";
import { ContinuousShape } from "../PointSet/Continuous.js";
import { notYetImplemented, xyShapeDistError } from "./DistError.js";
import { DiscreteShape } from "../PointSet/Discrete.js";
const square = (n) => {
    return n * n;
};
export class SymbolicDist extends BaseDist {
    toSparkline(bucketCount, env) {
        return Result.bind(this.toPointSetDist({
            xyPointLength: bucketCount * 3,
            sampleCount: env.sampleCount,
        }, "Linear"), (r) => r.toSparkline(bucketCount));
    }
    normalize() {
        return this;
    }
    integralSum() {
        return 1;
    }
    pdf(f) {
        return Ok(this.simplePdf(f));
    }
    interpolateXs(opts) {
        const { xSelection, points } = opts;
        switch (xSelection) {
            case "Linear":
                return E_A_Floats.range(this.min(), this.max(), points);
            case "ByWeight": {
                const ys = E_A_Floats.range(SymbolicDist.minCdfValue, SymbolicDist.maxCdfValue, points);
                return ys.map((y) => this.inv(y));
            }
            default:
                throw new Error(`Unknown xSelection value ${xSelection}`);
        }
    }
    toPointSetDist(env, xSelection = "ByWeight") {
        const xs = this.interpolateXs({
            xSelection,
            points: env.xyPointLength,
            env,
        });
        const ys = xs.map((x) => this.simplePdf(x));
        const xyShapeR = XYShape.T.make(xs, ys);
        if (!xyShapeR.ok) {
            return Result.Err(xyShapeDistError(xyShapeR.value));
        }
        return Ok(new PointSetDist(new ContinuousShape({
            integralSumCache: 1.0,
            xyShape: xyShapeR.value,
        }).toMixed()));
    }
    truncate(left, right, opts) {
        if (!opts) {
            throw new Error("env is necessary for truncating a symbolic dist");
        }
        if (left === undefined && right === undefined) {
            return Result.Ok(this);
        }
        const pointSetDistR = this.toPointSetDist(opts.env);
        if (!pointSetDistR.ok) {
            return pointSetDistR;
        }
        return pointSetDistR.value.truncate(left, right);
    }
    min() {
        return this.inv(SymbolicDist.minCdfValue);
    }
    max() {
        return this.inv(SymbolicDist.maxCdfValue);
    }
    sampleN(n) {
        const result = new Array(n);
        for (let i = 0; i < n; i++) {
            result[i] = this.sample();
        }
        return result;
    }
    expectedConvolutionCost() {
        return magicNumbers.OpCost.symbolicCost;
    }
    isFloat() {
        return false;
    }
}
SymbolicDist.minCdfValue = 0.0001;
SymbolicDist.maxCdfValue = 0.9999;
export class Normal extends SymbolicDist {
    constructor({ mean, stdev }) {
        super();
        this._mean = mean;
        this._stdev = stdev;
    }
    static make({ mean, stdev, }) {
        if (stdev <= 0) {
            return Result.Err("Standard deviation of normal distribution must be larger than 0");
        }
        return Ok(new Normal({ mean, stdev }));
    }
    toString() {
        return `Normal(${this._mean},${this._stdev})`;
    }
    simplePdf(x) {
        return jstat.normal.pdf(x, this._mean, this._stdev);
    }
    cdf(x) {
        return jstat.normal.cdf(x, this._mean, this._stdev);
    }
    inv(x) {
        return jstat.normal.inv(x, this._mean, this._stdev);
    }
    sample() {
        return jstat.normal.sample(this._mean, this._stdev);
    }
    mean() {
        return jstat.normal.mean(this._mean, this._stdev);
    }
    stdev() {
        return Ok(this._stdev);
    }
    variance() {
        return Ok(this._stdev ** 2);
    }
    _isEqual(other) {
        return this._mean === other._mean && this._stdev === other._stdev;
    }
    static fromCredibleInterval({ low, high, probability, }) {
        if (low >= high) {
            return Result.Err("Low value must be less than high value");
        }
        if (probability <= 0 || probability >= 1) {
            return Result.Err("Probability must be in (0, 1) interval");
        }
        const normalizedSigmas = jstat.normal.inv(1 - (1 - probability) / 2, 0, 1);
        const mean = (low + high) / 2;
        const stdev = (high - low) / (2 * normalizedSigmas);
        return Normal.make({ mean, stdev });
    }
    static add(n1, n2) {
        const mean = n1._mean + n2._mean;
        const stdev = Math.sqrt(n1._stdev ** 2 + n2._stdev ** 2);
        return new Normal({ mean, stdev });
    }
    static subtract(n1, n2) {
        const mean = n1._mean - n2._mean;
        const stdev = Math.sqrt(n1._stdev ** 2 + n2._stdev ** 2);
        return new Normal({ mean, stdev });
    }
    static operate(operation, n1, n2) {
        if (operation === "Add") {
            return Normal.add(n1, n2);
        }
        else if (operation === "Subtract") {
            return Normal.subtract(n1, n2);
        }
        return undefined;
    }
    static operateFloatFirst(operation, n1, n2) {
        if (operation === "Add") {
            return new Normal({ mean: n1 + n2._mean, stdev: n2._stdev });
        }
        else if (operation === "Subtract") {
            return new Normal({ mean: n1 - n2._mean, stdev: n2._stdev });
        }
        else if (operation === "Multiply") {
            if (n1 === 0) {
                return new PointMass(0);
            }
            return new Normal({
                mean: n1 * n2._mean,
                stdev: Math.abs(n1) * n2._stdev,
            });
        }
        return undefined;
    }
    static operateFloatSecond(operation, n1, n2) {
        if (operation === "Add") {
            return new Normal({ mean: n1._mean + n2, stdev: n1._stdev });
        }
        else if (operation === "Subtract") {
            return new Normal({ mean: n1._mean - n2, stdev: n1._stdev });
        }
        else if (operation === "Multiply") {
            if (n2 === 0) {
                return new PointMass(0);
            }
            return new Normal({
                mean: n1._mean * n2,
                stdev: n1._stdev * Math.abs(n2),
            });
        }
        else if (operation === "Divide") {
            return new Normal({
                mean: n1._mean / n2,
                stdev: n1._stdev / Math.abs(n2),
            });
        }
        return undefined;
    }
}
export class Exponential extends SymbolicDist {
    constructor(rate) {
        super();
        this.rate = rate;
    }
    static make(rate) {
        if (rate <= 0) {
            return Result.Err("Exponential distributions rate must be larger than 0.");
        }
        return Ok(new Exponential(rate));
    }
    toString() {
        return `Exponential(${this.rate})`;
    }
    simplePdf(x) {
        return jstat.exponential.pdf(x, this.rate);
    }
    cdf(x) {
        return jstat.exponential.cdf(x, this.rate);
    }
    inv(p) {
        return jstat.exponential.inv(p, this.rate);
    }
    sample() {
        return jstat.exponential.sample(this.rate);
    }
    mean() {
        return jstat.exponential.mean(this.rate);
    }
    variance() {
        return Ok(jstat.exponential.variance(this.rate));
    }
    _isEqual(other) {
        return this.rate === other.rate;
    }
}
export class Cauchy extends SymbolicDist {
    constructor({ local, scale }) {
        super();
        this.local = local;
        this.scale = scale;
    }
    static make({ local, scale, }) {
        if (scale > 0) {
            return Ok(new Cauchy({ local, scale }));
        }
        else {
            return Result.Err("Cauchy distribution scale parameter must larger than 0.");
        }
    }
    toString() {
        return `Cauchy(${this.local}, ${this.scale})`;
    }
    simplePdf(x) {
        return jstat.cauchy.pdf(x, this.local, this.scale);
    }
    cdf(x) {
        return jstat.cauchy.cdf(x, this.local, this.scale);
    }
    inv(p) {
        return jstat.cauchy.inv(p, this.local, this.scale);
    }
    sample() {
        return jstat.cauchy.sample(this.local, this.scale);
    }
    mean() {
        return NaN;
    }
    stdev() {
        return Ok(NaN);
    }
    variance() {
        return Ok(NaN);
    }
    _isEqual(other) {
        return this.local === other.local && this.scale === other.scale;
    }
}
export class Triangular extends SymbolicDist {
    constructor({ low, medium, high, }) {
        super();
        this.low = low;
        this.medium = medium;
        this.high = high;
    }
    static make({ low, medium, high, }) {
        if (low < medium && medium < high) {
            return Ok(new Triangular({ low, medium, high }));
        }
        return Result.Err("Triangular values must be increasing order.");
    }
    toString() {
        return `Triangular(${this.low}, ${this.medium}, ${this.high})`;
    }
    simplePdf(x) {
        return jstat.triangular.pdf(x, this.low, this.high, this.medium);
    }
    cdf(x) {
        return jstat.triangular.cdf(x, this.low, this.high, this.medium);
    }
    inv(p) {
        return jstat.triangular.inv(p, this.low, this.high, this.medium);
    }
    sample() {
        return jstat.triangular.sample(this.low, this.high, this.medium);
    }
    mean() {
        return jstat.triangular.mean(this.low, this.high, this.medium);
    }
    variance() {
        return Ok(jstat.triangular.variance(this.low, this.high, this.medium));
    }
    _isEqual(other) {
        return (this.low === other.low &&
            this.medium === other.medium &&
            this.high === other.high);
    }
    min() {
        return this.low;
    }
    max() {
        return this.high;
    }
}
export class Beta extends SymbolicDist {
    constructor({ alpha, beta }) {
        super();
        this.alpha = alpha;
        this.beta = beta;
    }
    static make({ alpha, beta, }) {
        if (alpha > 0 && beta > 0) {
            return Ok(new Beta({ alpha, beta }));
        }
        else {
            return Result.Err("Beta distribution parameters must be positive");
        }
    }
    toString() {
        return `Beta(${this.alpha},${this.beta})`;
    }
    simplePdf(x) {
        return jstat.beta.pdf(x, this.alpha, this.beta);
    }
    cdf(x) {
        return jstat.beta.cdf(x, this.alpha, this.beta);
    }
    inv(x) {
        return jstat.beta.inv(x, this.alpha, this.beta);
    }
    sample() {
        return jstat.beta.sample(this.alpha, this.beta);
    }
    mean() {
        return jstat.beta.mean(this.alpha, this.beta);
    }
    variance() {
        return Ok(jstat.beta.variance(this.alpha, this.beta));
    }
    _isEqual(other) {
        return this.alpha === other.alpha && this.beta === other.beta;
    }
    static fromMeanAndSampleSize({ mean, sampleSize, }) {
        const alpha = mean * sampleSize;
        const beta = (1 - mean) * sampleSize;
        return Beta.make({ alpha, beta });
    }
    static fromMeanAndStdev({ mean, stdev, }) {
        if (!(0 < stdev && stdev <= 0.5)) {
            return Result.Err("Stdev must be in in between 0 and 0.5.");
        }
        else if (!(0 <= mean && mean <= 1)) {
            return Result.Err("Mean must be in between 0 and 1.0.");
        }
        else {
            const variance = stdev * stdev;
            const sampleSize = (mean * (1 - mean)) / variance - 1;
            return Beta.fromMeanAndSampleSize({ mean, sampleSize });
        }
    }
}
export class Lognormal extends SymbolicDist {
    constructor({ mu, sigma }) {
        super();
        this.mu = mu;
        this.sigma = sigma;
    }
    static make({ mu, sigma, }) {
        if (sigma <= 0) {
            return Result.Err("Lognormal standard deviation must be larger than 0");
        }
        return Ok(new Lognormal({ mu, sigma }));
    }
    toString() {
        return `Lognormal(${this.mu},${this.sigma})`;
    }
    simplePdf(x) {
        return jstat.lognormal.pdf(x, this.mu, this.sigma);
    }
    cdf(x) {
        return jstat.lognormal.cdf(x, this.mu, this.sigma);
    }
    inv(x) {
        return jstat.lognormal.inv(x, this.mu, this.sigma);
    }
    sample() {
        return jstat.lognormal.sample(this.mu, this.sigma);
    }
    mean() {
        return jstat.lognormal.mean(this.mu, this.sigma);
    }
    variance() {
        return Ok((Math.exp(this.sigma * this.sigma) - 1) *
            Math.exp(2 * this.mu + this.sigma * this.sigma));
    }
    _isEqual(other) {
        return this.mu === other.mu && this.sigma === other.sigma;
    }
    static fromCredibleInterval({ low, high, probability, }) {
        if (low >= high) {
            return Result.Err("Low value must be less than high value");
        }
        if (low <= 0) {
            return Result.Err("Low value must be above 0");
        }
        if (probability <= 0 || probability >= 1) {
            return Result.Err("Probability must be in (0, 1) interval");
        }
        const logLow = Math.log(low);
        const logHigh = Math.log(high);
        const normalizedSigmas = jstat.normal.inv(1 - (1 - probability) / 2, 0, 1);
        const mu = (logLow + logHigh) / 2;
        const sigma = (logHigh - logLow) / (2 * normalizedSigmas);
        return Lognormal.make({ mu, sigma });
    }
    static fromMeanAndStdev({ mean, stdev, }) {
        if (mean <= 0) {
            return Result.Err("Lognormal mean must be larger than 0");
        }
        else if (stdev <= 0) {
            return Result.Err("Lognormal standard deviation must be larger than 0");
        }
        else {
            const variance = stdev ** 2;
            const meanSquared = mean ** 2;
            const mu = 2 * Math.log(mean) - 0.5 * Math.log(variance + meanSquared);
            const sigma = Math.sqrt(Math.log(variance / meanSquared + 1));
            return Ok(new Lognormal({ mu, sigma }));
        }
    }
    static multiply(l1, l2) {
        const mu = l1.mu + l2.mu;
        const sigma = Math.sqrt(l1.sigma ** 2 + l2.sigma ** 2);
        return new Lognormal({ mu, sigma });
    }
    static divide(l1, l2) {
        const mu = l1.mu - l2.mu;
        const sigma = Math.sqrt(l1.sigma ** 2 + l2.sigma ** 2);
        return new Lognormal({ mu, sigma });
    }
    static operate(operation, n1, n2) {
        if (operation === "Multiply") {
            return Lognormal.multiply(n1, n2);
        }
        else if (operation === "Divide") {
            return Lognormal.divide(n1, n2);
        }
        return undefined;
    }
    static operateFloatFirst(operation, n1, n2) {
        if (operation === "Multiply") {
            return n1 > 0
                ? new Lognormal({ mu: Math.log(n1) + n2.mu, sigma: n2.sigma })
                : undefined;
        }
        else if (operation === "Divide") {
            return n1 > 0
                ? new Lognormal({ mu: Math.log(n1) - n2.mu, sigma: n2.sigma })
                : undefined;
        }
        return undefined;
    }
    static operateFloatSecond(operation, n1, n2) {
        if (operation === "Multiply") {
            return n2 > 0
                ? new Lognormal({ mu: n1.mu + Math.log(n2), sigma: n1.sigma })
                : undefined;
        }
        else if (operation === "Divide") {
            return n2 > 0
                ? new Lognormal({ mu: n1.mu - Math.log(n2), sigma: n1.sigma })
                : undefined;
        }
        return undefined;
    }
}
export class Uniform extends SymbolicDist {
    constructor({ low, high }) {
        super();
        this.high = high;
        this.low = low;
    }
    static make({ low, high, }) {
        if (high > low) {
            return Ok(new Uniform({ low, high }));
        }
        else {
            return Result.Err("High must be larger than low");
        }
    }
    interpolateXs(opts) {
        if (opts.xSelection === "ByWeight") {
            const distance = this.high - this.low;
            const dx = magicNumbers.Epsilon.ten * distance;
            return [
                this.low - dx,
                this.low,
                this.low + dx,
                this.high - dx,
                this.high,
                this.high + dx,
            ];
        }
        return super.interpolateXs(opts);
    }
    toString() {
        return `Uniform(${this.low},${this.high})`;
    }
    simplePdf(x) {
        return jstat.uniform.pdf(x, this.low, this.high);
    }
    cdf(x) {
        return jstat.uniform.cdf(x, this.low, this.high);
    }
    inv(x) {
        return jstat.uniform.inv(x, this.low, this.high);
    }
    sample() {
        return jstat.uniform.sample(this.low, this.high);
    }
    mean() {
        return jstat.uniform.mean(this.low, this.high);
    }
    variance() {
        return Ok(Math.pow(this.high - this.low, 2) / 12);
    }
    _isEqual(other) {
        return this.low === other.low && this.high === other.high;
    }
    min() {
        return this.low;
    }
    max() {
        return this.high;
    }
    truncate(left, right) {
        const newLow = Math.max(left ?? -Infinity, this.low);
        const newHigh = Math.min(right ?? Infinity, this.high);
        return Ok(new Uniform({ low: newLow, high: newHigh }));
    }
}
export class Logistic extends SymbolicDist {
    constructor({ location, scale, }) {
        super();
        this.location = location;
        this.scale = scale;
    }
    static make({ location, scale, }) {
        if (scale > 0) {
            return Ok(new Logistic({ location, scale }));
        }
        else {
            return Result.Err("Scale must be positive");
        }
    }
    toString() {
        return `Logistic(${this.location},${this.scale})`;
    }
    simplePdf(x) {
        if (this.scale === 0)
            return this.location === x ? +Infinity : 0;
        const exp_delta = Math.pow(Math.E, -((x - this.location) / this.scale));
        return exp_delta / (this.scale * square(1 + exp_delta));
    }
    cdf(x) {
        if (this.scale === 0)
            return this.location < x ? 0 : 1;
        const exp_delta = Math.pow(Math.E, -((x - this.location) / this.scale));
        return 1 / (1 + exp_delta);
    }
    inv(p) {
        if (this.scale == 0)
            return this.location;
        return this.location + this.scale * Math.log(p / (1 - p));
    }
    sample() {
        const s = Math.random();
        return this.inv(s);
    }
    mean() {
        return this.location;
    }
    variance() {
        return Result.Ok((square(this.scale) * square(Math.PI)) / 3);
    }
    _isEqual(other) {
        return this.location === other.location && this.scale === other.scale;
    }
}
export class Bernoulli extends SymbolicDist {
    constructor(p) {
        super();
        this.p = p;
    }
    static make(p) {
        if (p >= 0.0 && p <= 1.0) {
            return Ok(new Bernoulli(p));
        }
        else {
            return Result.Err("Bernoulli parameter must be between 0 and 1");
        }
    }
    toString() {
        return `Bernoulli(${this.p})`;
    }
    pmf(x) {
        return x === 0 ? 1 - this.p : this.p;
    }
    simplePdf(x) {
        return this.pmf(x);
    }
    cdf(x) {
        return x < 0 ? 0 : x >= 1 ? 1 : 1 - this.p;
    }
    inv(prob) {
        return prob <= 1 - this.p ? 0 : 1;
    }
    mean() {
        return this.p;
    }
    sample() {
        const s = Math.random();
        return this.inv(s);
    }
    _isEqual(other) {
        return this.p === other.p;
    }
    min() {
        return this.p === 1 ? 1 : 0;
    }
    max() {
        return this.p === 0 ? 0 : 1;
    }
    variance() {
        return Ok(this.p * (1 - this.p));
    }
    toPointSetDist() {
        return Ok(new PointSetDist(new DiscreteShape({
            integralSumCache: 1.0,
            xyShape: { xs: [0, 1], ys: [1 - this.p, this.p] },
        }).toMixed()));
    }
}
export class Gamma extends SymbolicDist {
    constructor({ shape, scale }) {
        super();
        this.shape = shape;
        this.scale = scale;
    }
    static make({ shape, scale, }) {
        if (shape <= 0) {
            return Result.Err("shape must be larger than 0");
        }
        if (scale <= 0) {
            return Result.Err("scale must be larger than 0");
        }
        return Ok(new Gamma({ shape, scale }));
    }
    toString() {
        return `(${this.shape}, ${this.scale})`;
    }
    simplePdf(x) {
        return jstat.gamma.pdf(x, this.shape, this.scale);
    }
    cdf(x) {
        return jstat.gamma.cdf(x, this.shape, this.scale);
    }
    inv(x) {
        return jstat.gamma.inv(x, this.shape, this.scale);
    }
    sample() {
        return jstat.gamma.sample(this.shape, this.scale);
    }
    mean() {
        return jstat.gamma.mean(this.shape, this.scale);
    }
    variance() {
        return Ok(this.shape * this.scale * this.scale);
    }
    _isEqual(other) {
        return this.shape === other.shape && this.scale === other.scale;
    }
}
export class PointMass extends SymbolicDist {
    constructor(t) {
        super();
        this.t = t;
    }
    toString() {
        return `PointMass(${this.t})`;
    }
    static make(t) {
        if (isFinite(t)) {
            return Ok(new PointMass(t));
        }
        else {
            return Result.Err("PointMass must be finite");
        }
    }
    simplePdf(x) {
        return x === this.t ? 1.0 : 0.0;
    }
    cdf(x) {
        return x >= this.t ? 1.0 : 0.0;
    }
    inv(p) {
        return this.t;
    }
    mean() {
        return this.t;
    }
    variance() {
        return Ok(0);
    }
    sample() {
        return this.t;
    }
    _isEqual(other) {
        return this.t === other.t;
    }
    min() {
        return this.t;
    }
    max() {
        return this.t;
    }
    expectedConvolutionCost() {
        return magicNumbers.OpCost.floatCost;
    }
    isFloat() {
        return true;
    }
    toPointSetDist() {
        return Ok(new PointSetDist(new DiscreteShape({
            integralSumCache: 1.0,
            xyShape: { xs: [this.t], ys: [1.0] },
        }).toMixed()));
    }
}
export class Binomial extends SymbolicDist {
    constructor(n, p) {
        super();
        this.n = n;
        this.p = p;
    }
    toString() {
        return `Binomial(${this.n},{${this.p}})`;
    }
    static make(n, p) {
        if (!Number.isInteger(n) || n < 0) {
            return Result.Err("The number of trials (n) must be a non-negative integer.");
        }
        if (p < 0 || p > 1) {
            return Result.Err("Binomial p must be between 0 and 1");
        }
        return Ok(new Binomial(n, p));
    }
    simplePdf(x) {
        return jstat.binomial.pdf(x, this.n, this.p);
    }
    cdf(k) {
        return jstat.binomial.cdf(k, this.n, this.p);
    }
    inv(p) {
        throw notYetImplemented();
    }
    mean() {
        return this.n * this.p;
    }
    variance() {
        return Ok(this.n * this.p * (1 - this.p));
    }
    sample() {
        const bernoulli = Bernoulli.make(this.p);
        if (bernoulli.ok) {
            return sum(Array.from({ length: this.n }, () => bernoulli.value.sample()));
        }
        else {
            throw new Error("Binomial sampling failed");
        }
    }
    _isEqual(other) {
        return this.n === other.n && this.p === other.p;
    }
    toPointSetDist() {
        return Result.Err(notYetImplemented());
    }
}
export class Poisson extends SymbolicDist {
    constructor(lambda) {
        super();
        this.lambda = lambda;
    }
    toString() {
        return `Poisson(${this.lambda}})`;
    }
    static make(lambda) {
        if (lambda <= 0) {
            throw new Error("Lambda must be a positive number for a Poisson distribution.");
        }
        return Ok(new Poisson(lambda));
    }
    simplePdf(x) {
        return jstat.poisson.pdf(x, this.lambda);
    }
    cdf(k) {
        return jstat.poisson.cdf(k, this.lambda);
    }
    inv(p) {
        throw new Error("Poisson inv not implemented");
    }
    mean() {
        return this.lambda;
    }
    variance() {
        return Ok(this.lambda);
    }
    sample() {
        return jstat.poisson.sample(this.lambda);
    }
    _isEqual(other) {
        return this.lambda === other.lambda;
    }
    toPointSetDist() {
        return Result.Err(notYetImplemented());
    }
}
export const tryAnalyticalSimplification = (d1, d2, op) => {
    if (d1 instanceof PointMass && d2 instanceof PointMass) {
        return Result.fmap(Operation.Algebraic.toFn(op)(d1.t, d2.t), (v) => new PointMass(v));
    }
    else if (d1 instanceof Normal && d2 instanceof Normal) {
        const out = Normal.operate(op, d1, d2);
        return out ? Ok(out) : undefined;
    }
    else if (d1 instanceof PointMass && d2 instanceof Normal) {
        const out = Normal.operateFloatFirst(op, d1.t, d2);
        return out ? Ok(out) : undefined;
    }
    else if (d1 instanceof Normal && d2 instanceof PointMass) {
        const out = Normal.operateFloatSecond(op, d1, d2.t);
        return out ? Ok(out) : undefined;
    }
    else if (d1 instanceof Lognormal && d2 instanceof Lognormal) {
        const out = Lognormal.operate(op, d1, d2);
        return out ? Ok(out) : undefined;
    }
    else if (d1 instanceof PointMass && d2 instanceof Lognormal) {
        const out = Lognormal.operateFloatFirst(op, d1.t, d2);
        return out ? Ok(out) : undefined;
    }
    else if (d1 instanceof Lognormal && d2 instanceof PointMass) {
        const out = Lognormal.operateFloatSecond(op, d1, d2.t);
        return out ? Ok(out) : undefined;
    }
    else {
        return undefined;
    }
};
