// @quri/squiggle-lang/src/rescript/ForTS/ForTS_Distribution/ForTS_Distribution_PointSetDistribution.bs.js
'use strict';

var ForTS_Distribution_PointSetDistribution_tag = require("./ForTS_Distribution_PointSetDistribution_tag");

var pstMixed_ = ForTS_Distribution_PointSetDistribution_tag.pointSetDistributionTag.Mixed;

var pstDiscrete_ = ForTS_Distribution_PointSetDistribution_tag.pointSetDistributionTag.Discrete;

var pstContinuous_ = ForTS_Distribution_PointSetDistribution_tag.pointSetDistributionTag.Continuous;

function getTag(variant) {
  switch (variant.TAG | 0) {
    case /* Mixed */0 :
        return pstMixed_;
    case /* Discrete */1 :
        return pstDiscrete_;
    case /* Continuous */2 :
        return pstContinuous_;
    
  }
}

function getMixed(variant) {
  switch (variant.TAG | 0) {
    case /* Mixed */0 :
        return variant._0;
    case /* Discrete */1 :
    case /* Continuous */2 :
        return ;
    
  }
}

function getDiscrete(variant) {
  switch (variant.TAG | 0) {
    case /* Discrete */1 :
        return variant._0;
    case /* Mixed */0 :
    case /* Continuous */2 :
        return ;
    
  }
}

function getContinues(variant) {
  switch (variant.TAG | 0) {
    case /* Mixed */0 :
    case /* Discrete */1 :
        return ;
    case /* Continuous */2 :
        return variant._0;
    
  }
}

function toDistribution(v) {
  return {
          TAG: 0,
          _0: v,
          [Symbol.for("name")]: "PointSet"
        };
}

exports.pstMixed_ = pstMixed_;
exports.pstDiscrete_ = pstDiscrete_;
exports.pstContinuous_ = pstContinuous_;
exports.getTag = getTag;
exports.getMixed = getMixed;
exports.getDiscrete = getDiscrete;
exports.getContinues = getContinues;
exports.toDistribution = toDistribution;
/* pstMixed_ Not a pure module */
