// @quri/squiggle-lang/dist/dist/distOperations/binaryOperations.js
import { accumulateWithError, pairwiseWithError } from "../../utility/E_A.js";
import { Ok, bind } from "../../utility/result.js";
import { PointMass } from "../SymbolicDist.js";
import { algebraicCombination } from "./algebraicCombination.js";
import { pointwiseCombination } from "./pointwiseCombination.js";
function algebraic(t1, t2, env, operation) {
    return algebraicCombination({ t1, t2, arithmeticOperation: operation, env });
}
function pointwise(t1, t2, env, algebraicOperation) {
    return pointwiseCombination({
        t1,
        t2,
        env,
        algebraicOperation,
    });
}
export const binaryOperations = {
    algebraicAdd: (t1, t2, { env }) => algebraic(t1, t2, env, "Add"),
    algebraicMultiply: (t1, t2, { env }) => algebraic(t1, t2, env, "Multiply"),
    algebraicDivide: (t1, t2, { env }) => algebraic(t1, t2, env, "Divide"),
    algebraicSubtract: (t1, t2, { env }) => algebraic(t1, t2, env, "Subtract"),
    algebraicLogarithm: (t1, t2, { env }) => algebraic(t1, t2, env, "Logarithm"),
    algebraicPower: (t1, t2, { env }) => algebraic(t1, t2, env, "Power"),
    pointwiseAdd: (t1, t2, { env }) => pointwise(t1, t2, env, "Add"),
    pointwiseMultiply: (t1, t2, { env }) => pointwise(t1, t2, env, "Multiply"),
    pointwiseDivide: (t1, t2, { env }) => pointwise(t1, t2, env, "Divide"),
    pointwiseSubtract: (t1, t2, { env }) => pointwise(t1, t2, env, "Subtract"),
    pointwiseLogarithm: (t1, t2, { env }) => pointwise(t1, t2, env, "Logarithm"),
    pointwisePower: (t1, t2, { env }) => pointwise(t1, t2, env, "Power"),
};
export const algebraicSum = (dists, env) => dists.reduce((accumulatedDist, currentDist) => bind(accumulatedDist, (aVal) => binaryOperations.algebraicAdd(aVal, currentDist, { env })), Ok(new PointMass(0)));
export const algebraicProduct = (dists, env) => dists.reduce((accumulatedDist, currentDist) => bind(accumulatedDist, (aVal) => binaryOperations.algebraicMultiply(aVal, currentDist, { env })), Ok(new PointMass(1)));
export const algebraicCumSum = (dists, env) => accumulateWithError(dists, (a, b) => binaryOperations.algebraicAdd(a, b, { env }));
export const algebraicCumProd = (dists, env) => accumulateWithError(dists, (a, b) => binaryOperations.algebraicMultiply(a, b, { env }));
export const algebraicDiff = (dists, env) => pairwiseWithError(dists, (a, b) => binaryOperations.algebraicSubtract(b, a, {
    env,
}));
