// @quri/squiggle-lang/src/rescript/SqError.gen.tsx
/* TypeScript file generated from SqError.res by genType. */
/* eslint-disable import/first */


// @ts-ignore: Implicit any on import
const SqErrorBS = require('./SqError.bs');

import type {frame as Reducer_T_frame} from '../../src/rescript/Reducer/Reducer_T.gen';

import type {t as Reducer_FrameStack_t} from '../../src/rescript/Reducer/Reducer_FrameStack.gen';

// tslint:disable-next-line:max-classes-per-file 
export abstract class Message_t { protected opaque!: any }; /* simulate opaque types */

// tslint:disable-next-line:max-classes-per-file 
// tslint:disable-next-line:class-name
export abstract class t { protected opaque!: any }; /* simulate opaque types */

export const fromMessage: (message:Message_t) => t = SqErrorBS.fromMessage;

export const getTopFrame: (t:t) => (null | undefined | Reducer_T_frame) = SqErrorBS.getTopFrame;

export const getFrameStack: (t:t) => Reducer_FrameStack_t = SqErrorBS.getFrameStack;

export const toString: (t:t) => string = SqErrorBS.toString;

export const createOtherError: (v:string) => t = SqErrorBS.createOtherError;

export const getFrameArray: (t:t) => Reducer_T_frame[] = SqErrorBS.getFrameArray;

export const toStringWithStackTrace: (t:t) => string = SqErrorBS.toStringWithStackTrace;
