// @quri/squiggle-lang/src/rescript/ForTS/ForTS_ReducerProject.bs.js
'use strict';

var ReducerProject$QuriSquiggleLang = require("../ReducerProject/ReducerProject.bs.js");
var Reducer_Namespace$QuriSquiggleLang = require("../Reducer/Reducer_Namespace.bs.js");

function createProject(param) {
  return ReducerProject$QuriSquiggleLang.createProject(undefined);
}

var getSourceIds = ReducerProject$QuriSquiggleLang.getSourceIds;

var setSource = ReducerProject$QuriSquiggleLang.setSource;

var removeSource = ReducerProject$QuriSquiggleLang.removeSource;

var getSource = ReducerProject$QuriSquiggleLang.getSource;

var touchSource = ReducerProject$QuriSquiggleLang.touchSource;

var clean = ReducerProject$QuriSquiggleLang.clean;

var cleanAll = ReducerProject$QuriSquiggleLang.cleanAll;

var cleanResults = ReducerProject$QuriSquiggleLang.cleanResults;

var cleanAllResults = ReducerProject$QuriSquiggleLang.cleanAllResults;

var getIncludes = ReducerProject$QuriSquiggleLang.getIncludes;

var getPastChain = ReducerProject$QuriSquiggleLang.getPastChain;

var getContinues = ReducerProject$QuriSquiggleLang.getContinues;

var setContinues = ReducerProject$QuriSquiggleLang.setContinues;

var getDependencies = ReducerProject$QuriSquiggleLang.getDependencies;

var getDependents = ReducerProject$QuriSquiggleLang.getDependents;

var getRunOrder = ReducerProject$QuriSquiggleLang.getRunOrder;

var getRunOrderFor = ReducerProject$QuriSquiggleLang.getRunOrderFor;

var parseIncludes = ReducerProject$QuriSquiggleLang.parseIncludes;

var rawParse = ReducerProject$QuriSquiggleLang.rawParse;

var run = ReducerProject$QuriSquiggleLang.run;

var runAll = ReducerProject$QuriSquiggleLang.runAll;

function getBindings(project, sourceId) {
  return Reducer_Namespace$QuriSquiggleLang.toMap(ReducerProject$QuriSquiggleLang.getBindings(project, sourceId));
}

var getResult = ReducerProject$QuriSquiggleLang.getResult;

var evaluate = ReducerProject$QuriSquiggleLang.evaluate;

var setEnvironment = ReducerProject$QuriSquiggleLang.setEnvironment;

var getEnvironment = ReducerProject$QuriSquiggleLang.getEnvironment;

var T;

var Private;

exports.T = T;
exports.Private = Private;
exports.createProject = createProject;
exports.getSourceIds = getSourceIds;
exports.setSource = setSource;
exports.removeSource = removeSource;
exports.getSource = getSource;
exports.touchSource = touchSource;
exports.clean = clean;
exports.cleanAll = cleanAll;
exports.cleanResults = cleanResults;
exports.cleanAllResults = cleanAllResults;
exports.getIncludes = getIncludes;
exports.getPastChain = getPastChain;
exports.getContinues = getContinues;
exports.setContinues = setContinues;
exports.getDependencies = getDependencies;
exports.getDependents = getDependents;
exports.getRunOrder = getRunOrder;
exports.getRunOrderFor = getRunOrderFor;
exports.parseIncludes = parseIncludes;
exports.rawParse = rawParse;
exports.run = run;
exports.runAll = runAll;
exports.getBindings = getBindings;
exports.getResult = getResult;
exports.evaluate = evaluate;
exports.setEnvironment = setEnvironment;
exports.getEnvironment = getEnvironment;
/* ReducerProject-QuriSquiggleLang Not a pure module */
