// @quri/squiggle-lang/src/rescript/ForTS/ForTS__Types.gen.tsx
/* TypeScript file generated from ForTS__Types.res by genType. */
/* eslint-disable import/first */


import type {distributionError as ForTS_Distribution_Error_distributionError} from '../../../src/rescript/ForTS/ForTS_Distribution/ForTS_Distribution_Error.gen';

import type {distribution as ForTS_Distribution_distribution} from '../../../src/rescript/ForTS/ForTS_Distribution/ForTS_Distribution.gen';

import type {environment as ForTS_Distribution_Environment_environment} from '../../../src/rescript/ForTS/ForTS_Distribution/ForTS_Distribution_Environment.gen';

import type {pointSetDistribution as ForTS_Distribution_PointSetDistribution_pointSetDistribution} from '../../../src/rescript/ForTS/ForTS_Distribution/ForTS_Distribution_PointSetDistribution.gen';

import type {reducerProject as ForTS_ReducerProject_reducerProject} from './ForTS_ReducerProject.gen';

import type {sampleSetDistribution as ForTS_Distribution_SampleSetDistribution_sampleSetDistribution} from '../../../src/rescript/ForTS/ForTS_Distribution/ForTS_Distribution_SampleSetDistribution.gen';

import type {squiggleValue_Array as ForTS_SquiggleValue_Array_squiggleValue_Array} from '../../../src/rescript/ForTS/ForTS_SquiggleValue/ForTS_SquiggleValue_Array.gen';

import type {squiggleValue_Declaration as ForTS_SquiggleValue_Declaration_squiggleValue_Declaration} from '../../../src/rescript/ForTS/ForTS_SquiggleValue/ForTS_SquiggleValue_Declaration.gen';

import type {squiggleValue_Lambda as ForTS_SquiggleValue_Lambda_squiggleValue_Lambda} from '../../../src/rescript/ForTS/ForTS_SquiggleValue/ForTS_SquiggleValue_Lambda.gen';

import type {squiggleValue_Record as ForTS_SquiggleValue_Record_squiggleValue_Record} from '../../../src/rescript/ForTS/ForTS_SquiggleValue/ForTS_SquiggleValue_Record.gen';

import type {squiggleValue as ForTS_SquiggleValue_squiggleValue} from '../../../src/rescript/ForTS/ForTS_SquiggleValue/ForTS_SquiggleValue.gen';

import type {symbolicDistribution as ForTS_Distribution_SymbolicDistribution_symbolicDistribution} from '../../../src/rescript/ForTS/ForTS_Distribution/ForTS_Distribution_SymbolicDistribution.gen';

// tslint:disable-next-line:interface-over-type-literal
export type reducerProject = ForTS_ReducerProject_reducerProject;

// tslint:disable-next-line:interface-over-type-literal
export type squiggleValue = ForTS_SquiggleValue_squiggleValue;

// tslint:disable-next-line:interface-over-type-literal
export type squiggleValue_Array = ForTS_SquiggleValue_Array_squiggleValue_Array;

// tslint:disable-next-line:interface-over-type-literal
export type squiggleValue_Declaration = ForTS_SquiggleValue_Declaration_squiggleValue_Declaration;

// tslint:disable-next-line:interface-over-type-literal
export type squiggleValue_Lambda = ForTS_SquiggleValue_Lambda_squiggleValue_Lambda;

// tslint:disable-next-line:interface-over-type-literal
export type squiggleValue_Record = ForTS_SquiggleValue_Record_squiggleValue_Record;

// tslint:disable-next-line:interface-over-type-literal
export type squiggleValue_Distribution = ForTS_Distribution_distribution;

// tslint:disable-next-line:interface-over-type-literal
export type distribution = squiggleValue_Distribution;

// tslint:disable-next-line:interface-over-type-literal
export type distributionError = ForTS_Distribution_Error_distributionError;

// tslint:disable-next-line:interface-over-type-literal
export type environment = ForTS_Distribution_Environment_environment;

// tslint:disable-next-line:interface-over-type-literal
export type pointSetDistribution = ForTS_Distribution_PointSetDistribution_pointSetDistribution;

// tslint:disable-next-line:interface-over-type-literal
export type sampleSetDistribution = ForTS_Distribution_SampleSetDistribution_sampleSetDistribution;

// tslint:disable-next-line:interface-over-type-literal
export type symbolicDistribution = ForTS_Distribution_SymbolicDistribution_symbolicDistribution;
