// @quri/squiggle-lang/dist/library/registry/core.js
import { BuiltinLambda } from "../../reducer/lambda.js";
export class Registry {
    constructor(functions, fnNameDict) {
        this.functions = functions;
        this.fnNameDict = fnNameDict;
    }
    static make(fns) {
        const dict = new Map();
        for (const fn of fns) {
            for (const def of fn.definitions) {
                const names = [
                    ...(fn.nameSpace == "" ? [] : [`${fn.nameSpace}.${fn.name}`]),
                    ...(fn.requiresNamespace ? [] : [fn.name]),
                ];
                for (const name of names) {
                    if (dict.has(name)) {
                        dict.get(name)?.push(def);
                    }
                    else {
                        dict.set(name, [def]);
                    }
                }
            }
        }
        return new Registry(fns, dict);
    }
    allExamplesWithFns() {
        return this.functions
            .map((fn) => fn.examples?.map((example) => ({
            fn,
            example,
        })) ?? [])
            .flat();
    }
    allNames() {
        return [...this.fnNameDict.keys()];
    }
    makeLambda(fnName) {
        const definitions = this.fnNameDict.get(fnName);
        if (definitions === undefined) {
            throw new Error(`Function ${fnName} has no definitions`);
        }
        else {
            return new BuiltinLambda(fnName, definitions);
        }
    }
}
