// @quri/squiggle-lang/dist/library/registry/frTypes.js
import { ImmutableMap } from "../../utility/immutableMap.js";
import { vArray, vBool, vDate, vDist, vLambda, vNumber, vDict, vScale, vString, vTimeDuration, vInput, } from "../../value/index.js";
export const frNumber = {
    unpack: (v) => (v.type === "Number" ? v.value : undefined),
    pack: (v) => vNumber(v),
    getName: () => "number",
};
export const frString = {
    unpack: (v) => (v.type === "String" ? v.value : undefined),
    pack: (v) => vString(v),
    getName: () => "string",
};
export const frBool = {
    unpack: (v) => (v.type === "Bool" ? v.value : undefined),
    pack: (v) => vBool(v),
    getName: () => "bool",
};
export const frDate = {
    unpack: (v) => (v.type === "Date" ? v.value : undefined),
    pack: (v) => vDate(v),
    getName: () => "date",
};
export const frTimeDuration = {
    unpack: (v) => (v.type === "TimeDuration" ? v.value : undefined),
    pack: (v) => vTimeDuration(v),
    getName: () => "duration",
};
export const frDistOrNumber = {
    unpack: (v) => v.type === "Dist" ? v.value : v.type === "Number" ? v.value : undefined,
    pack: (v) => (typeof v === "number" ? vNumber(v) : vDist(v)),
    getName: () => "distribution|number",
};
export const frNumberOrString = {
    unpack: (v) => v.type === "String" ? v.value : v.type === "Number" ? v.value : undefined,
    pack: (v) => (typeof v === "number" ? vNumber(v) : vString(v)),
    getName: () => "number|string",
};
export const frDist = {
    unpack: (v) => (v.type === "Dist" ? v.value : undefined),
    pack: (v) => vDist(v),
    getName: () => "distribution",
};
export const frLambda = {
    unpack: (v) => (v.type === "Lambda" ? v.value : undefined),
    pack: (v) => vLambda(v),
    getName: () => "lambda",
};
export const frLambdaN = (paramLength) => {
    return {
        unpack: (v) => {
            return v.type === "Lambda" &&
                v.value.parameterCounts().includes(paramLength)
                ? v.value
                : undefined;
        },
        pack: (v) => vLambda(v),
        getName: () => `lambda(${paramLength})`,
    };
};
export const frLambdaNand = (paramLengths) => {
    return {
        unpack: (v) => {
            const counts = v.type === "Lambda" && v.value.parameterCounts();
            return counts && paramLengths.every((p) => counts.includes(p))
                ? v.value
                : undefined;
        },
        pack: (v) => vLambda(v),
        getName: () => `lambda(${paramLengths.join(",")})`,
    };
};
export const frScale = {
    unpack: (v) => (v.type === "Scale" ? v.value : undefined),
    pack: (v) => vScale(v),
    getName: () => "scale",
};
export const frInput = {
    unpack: (v) => (v.type === "Input" ? v.value : undefined),
    pack: (v) => vInput(v),
    getName: () => "input",
};
export const frArray = (itemType) => {
    return {
        unpack: (v) => {
            if (v.type !== "Array") {
                return undefined;
            }
            if (itemType.getName() === "any") {
                return v.value;
            }
            const unpackedArray = [];
            for (const item of v.value) {
                const unpackedItem = itemType.unpack(item);
                if (unpackedItem === undefined) {
                    return undefined;
                }
                unpackedArray.push(unpackedItem);
            }
            return unpackedArray;
        },
        pack: (v) => vArray(v.map(itemType.pack)),
        getName: () => `list(${itemType.getName()})`,
    };
};
export function frTuple(...types) {
    const numTypes = types.length;
    return {
        unpack: (v) => {
            if (v.type !== "Array" || v.value.length !== numTypes) {
                return undefined;
            }
            const items = types.map((type, index) => type.unpack(v.value[index]));
            if (items.some((item) => item === undefined)) {
                return undefined;
            }
            return items;
        },
        pack: (values) => {
            return vArray(values.map((val, index) => types[index].pack(val)));
        },
        getName: () => `tuple(${types.map((type) => type.getName()).join(", ")})`,
    };
}
export const frDictWithArbitraryKeys = (itemType) => {
    return {
        unpack: (v) => {
            if (v.type !== "Dict") {
                return undefined;
            }
            let unpackedMap = ImmutableMap();
            for (const [key, value] of v.value.entries()) {
                const unpackedItem = itemType.unpack(value);
                if (unpackedItem === undefined) {
                    return undefined;
                }
                unpackedMap = unpackedMap.set(key, unpackedItem);
            }
            return unpackedMap;
        },
        pack: (v) => vDict(ImmutableMap([...v.entries()].map(([k, v]) => [k, itemType.pack(v)]))),
        getName: () => `dict(${itemType.getName()})`,
    };
};
export const frAny = {
    unpack: (v) => v,
    pack: (v) => v,
    getName: () => "any",
};
export function frDict(...allKvs) {
    return {
        unpack: (v) => {
            if (v.type !== "Dict") {
                return undefined;
            }
            const r = v.value;
            const result = {};
            for (const [key, valueShape] of allKvs) {
                const subvalue = r.get(key);
                if (subvalue === undefined) {
                    if ("isOptional" in valueShape) {
                        continue;
                    }
                    return undefined;
                }
                const unpackedSubvalue = valueShape.unpack(subvalue);
                if (unpackedSubvalue === undefined) {
                    return undefined;
                }
                result[key] = unpackedSubvalue;
            }
            return result;
        },
        pack: (v) => vDict(ImmutableMap(allKvs
            .filter(([key, valueShape]) => !("isOptional" in valueShape) || v[key] !== null)
            .map(([key, valueShape]) => [key, valueShape.pack(v[key])]))),
        getName: () => "{" +
            allKvs
                .map(([name, frType]) => `${name}: ${frType.getName()}`)
                .join(", ") +
            "}",
    };
}
export const frOptional = (itemType) => {
    return {
        unpack: (v) => {
            return itemType.unpack(v);
        },
        pack: (v) => {
            if (v === null) {
                throw new Error("Unable to pack null value");
            }
            return itemType.pack(v);
        },
        getName: () => `optional(${itemType.getName()})`,
        isOptional: true,
    };
};
