// @quri/squiggle-lang/src/rescript/Distributions/PointSetDist/PointSetTypes.gen.tsx
/* TypeScript file generated from PointSetTypes.res by genType. */
/* eslint-disable import/first */


import type {interpolationStrategy as XYShape_interpolationStrategy} from '../../../../src/rescript/Utility/XYShape.gen';

import type {xyShape as XYShape_xyShape} from '../../../../src/rescript/Utility/XYShape.gen';

// tslint:disable-next-line:interface-over-type-literal
export type xyShape = XYShape_xyShape;

// tslint:disable-next-line:interface-over-type-literal
export type interpolationStrategy = XYShape_interpolationStrategy;

// tslint:disable-next-line:interface-over-type-literal
export type continuousShape = {
  readonly xyShape: xyShape; 
  readonly interpolation: interpolationStrategy; 
  readonly integralSumCache?: number; 
  readonly integralCache?: continuousShape
};

// tslint:disable-next-line:interface-over-type-literal
export type discreteShape = {
  readonly xyShape: xyShape; 
  readonly integralSumCache?: number; 
  readonly integralCache?: continuousShape
};

// tslint:disable-next-line:interface-over-type-literal
export type mixedShape = {
  readonly continuous: continuousShape; 
  readonly discrete: discreteShape; 
  readonly integralSumCache?: number; 
  readonly integralCache?: continuousShape
};

// tslint:disable-next-line:max-classes-per-file 
// tslint:disable-next-line:class-name
export abstract class pointSetDist { protected opaque!: any }; /* simulate opaque types */

// tslint:disable-next-line:interface-over-type-literal
export type distPlus = {
  readonly pointSetDist: pointSetDist; 
  readonly integralCache: continuousShape; 
  readonly squiggleString?: string
};

// tslint:disable-next-line:interface-over-type-literal
export type sparklineError = "CannotSparklineDiscrete";
