// @quri/squiggle-lang/dist/expression/index.js
import { vVoid } from "../value/index.js";
export const eArray = (anArray) => ({
    type: "Array",
    value: anArray,
});
export const eValue = (value) => ({
    type: "Value",
    value,
});
export const eCall = (fn, args) => ({
    type: "Call",
    value: {
        fn,
        args,
    },
});
export const eLambda = (name, parameters, body) => ({
    type: "Lambda",
    value: {
        name,
        parameters,
        body,
    },
});
export const eDict = (aMap) => ({
    type: "Dict",
    value: aMap,
});
export const eResolvedSymbol = (name, offset) => ({
    type: "ResolvedSymbol",
    value: { name, offset },
});
export const eBlock = (exprs) => ({
    type: "Block",
    value: exprs,
});
export const eProgram = (statements, exports) => ({
    type: "Program",
    value: {
        statements,
        exports,
    },
});
export const eLetStatement = (left, right) => ({
    type: "Assign",
    value: {
        left,
        right,
    },
});
export const eTernary = (condition, ifTrue, ifFalse) => ({
    type: "Ternary",
    value: {
        condition,
        ifTrue,
        ifFalse,
    },
});
export const eVoid = () => ({
    type: "Value",
    value: vVoid(),
});
export function expressionToString(expression) {
    switch (expression.type) {
        case "Block":
            return `{${expression.value.map(expressionToString).join("; ")}}`;
        case "Program": {
            const exports = new Set(expression.value.exports);
            return expression.value.statements
                .map((statement) => {
                const statementString = expressionToString(statement);
                return statement.type === "Assign" &&
                    exports.has(statement.value.left)
                    ? `export ${statementString}`
                    : statementString;
            })
                .join("; ");
        }
        case "Array":
            return `[${expression.value.map(expressionToString).join(", ")}]`;
        case "Dict":
            return `{${expression.value
                .map(([key, value]) => `${expressionToString(key)}: ${expressionToString(value)}`)
                .join(", ")}}`;
        case "ResolvedSymbol":
            return expression.value.name;
        case "Ternary":
            return `${expressionToString(expression.value.condition)} ? (${expressionToString(expression.value.ifTrue)}) : (${expressionToString(expression.value.ifFalse)})`;
        case "Assign":
            return `${expression.value.left} = ${expressionToString(expression.value.right)}`;
        case "Call":
            return `(${expressionToString(expression.value.fn)})(${expression.value.args.map(expressionToString).join(", ")})`;
        case "Lambda":
            return `{|${expression.value.parameters
                .map((parameter) => parameter.name)
                .join(", ")}| ${expressionToString(expression.value.body)}}`;
        case "Value":
            return expression.value.toString();
        default:
            return `Unknown expression ${expression}`;
    }
}
