// @quri/squiggle-lang/dist/ast/peggyHelpers.js
export const infixFunctions = {
    "+": "add",
    "-": "subtract",
    "!=": "unequal",
    ".-": "dotSubtract",
    ".*": "dotMultiply",
    "./": "dotDivide",
    ".^": "dotPow",
    ".+": "dotAdd",
    "*": "multiply",
    "/": "divide",
    "&&": "and",
    "^": "pow",
    "<": "smaller",
    "<=": "smallerEq",
    "==": "equal",
    ">": "larger",
    ">=": "largerEq",
    "||": "or",
    to: "to",
};
export const unaryFunctions = {
    "-": "unaryMinus",
    "!": "not",
    ".-": "unaryDotMinus",
};
export function nodeCall(fn, args, location) {
    return { type: "Call", fn, args, location };
}
export function makeInfixChain(head, tail, location) {
    return tail.reduce((result, [operator, right]) => {
        return nodeInfixCall(operator, result, right, location);
    }, head);
}
export function nodeInfixCall(op, arg1, arg2, location) {
    return {
        type: "InfixCall",
        op,
        args: [arg1, arg2],
        location,
    };
}
export function nodeUnaryCall(op, arg, location) {
    return { type: "UnaryCall", op, arg, location };
}
export function nodePipe(leftArg, fn, rightArgs, location) {
    return { type: "Pipe", leftArg, fn, rightArgs, location };
}
export function nodeDotLookup(arg, key, location) {
    return { type: "DotLookup", arg, key, location };
}
export function nodeBracketLookup(arg, key, location) {
    return { type: "BracketLookup", arg, key, location };
}
export function nodeArray(elements, location) {
    return { type: "Array", elements, location };
}
export function nodeDict(elements, location) {
    const symbols = {};
    for (const element of elements) {
        if (element.type === "KeyValue" && element.key.type === "String") {
            symbols[element.key.value] = element;
        }
        else if (element.type === "Identifier") {
            symbols[element.value] = element;
        }
    }
    return { type: "Dict", elements, symbols, location };
}
export function nodeUnitValue(value, unit, location) {
    return { type: "UnitValue", value, unit, location };
}
export function nodeBlock(statements, location) {
    return { type: "Block", statements, location };
}
export function nodeProgram(imports, statements, location) {
    const symbols = {};
    for (const statement of statements) {
        if (statement.type === "LetStatement" ||
            statement.type === "DefunStatement") {
            symbols[statement.variable.value] = statement;
        }
    }
    return { type: "Program", imports, statements, symbols, location };
}
export function nodeBoolean(value, location) {
    return { type: "Boolean", value, location };
}
export function nodeFloat(args, location) {
    return { type: "Float", ...args, location };
}
export function nodeIdentifier(value, location) {
    return { type: "Identifier", value, location };
}
export function nodeIdentifierWithAnnotation(variable, annotation, location) {
    return { type: "IdentifierWithAnnotation", variable, annotation, location };
}
export function nodeKeyValue(key, value, location) {
    if (key.type === "Identifier") {
        key = {
            ...key,
            type: "String",
        };
    }
    return { type: "KeyValue", key, value, location };
}
export function nodeLambda(args, body, location, name) {
    return { type: "Lambda", args, body, location, name: name?.value };
}
export function nodeLetStatement(variable, value, exported, location) {
    const patchedValue = value.type === "Lambda" ? { ...value, name: variable.value } : value;
    return {
        type: "LetStatement",
        variable,
        value: patchedValue,
        exported,
        location,
    };
}
export function nodeDefunStatement(variable, value, exported, location) {
    return { type: "DefunStatement", variable, value, exported, location };
}
export function nodeString(value, location) {
    return { type: "String", value, location };
}
export function nodeTernary(condition, trueExpression, falseExpression, kind, location) {
    return {
        type: "Ternary",
        condition,
        trueExpression,
        falseExpression,
        kind,
        location,
    };
}
export function nodeVoid(location) {
    return { type: "Void", location };
}
export function lineComment(text, location) {
    return {
        type: "lineComment",
        value: text,
        location,
    };
}
export function blockComment(text, location) {
    return {
        type: "blockComment",
        value: text,
        location,
    };
}
export function parseEscapeSequence(char, location, error) {
    if (char[0] == "'") {
        return "'";
    }
    else {
        try {
            return JSON.parse(`"\\${char.join("")}"`);
        }
        catch (e) {
            error(`Incorrect escape sequence: ${char.join("")}`, location);
        }
    }
}
