// @quri/squiggle-lang/dist/public/SqValue/index.js
import { vLambda, vNumber, vString } from "../../value/index.js";
import { SqArray } from "./SqArray.js";
import { SqCalculator } from "./SqCalculator.js";
import { SqDict } from "./SqDict.js";
import { wrapDistribution } from "./SqDistribution/index.js";
import { wrapDomain } from "./SqDomain.js";
import { wrapInput } from "./SqInput.js";
import { SqLambda } from "./SqLambda.js";
import { wrapPlot } from "./SqPlot.js";
import { wrapScale } from "./SqScale.js";
import { SqTableChart } from "./SqTableChart.js";
export function wrapValue(value, context) {
    switch (value.type) {
        case "Array":
            return new SqArrayValue(value, context);
        case "Bool":
            return new SqBoolValue(value, context);
        case "Date":
            return new SqDateValue(value, context);
        case "Dist":
            return new SqDistributionValue(value, context);
        case "Lambda":
            return new SqLambdaValue(value, context);
        case "Number":
            return new SqNumberValue(value, context);
        case "Dict":
            return new SqDictValue(value, context);
        case "String":
            return new SqStringValue(value, context);
        case "Plot":
            return new SqPlotValue(value, context);
        case "TableChart":
            return new SqTableChartValue(value, context);
        case "Calculator":
            return new SqCalculatorValue(value, context);
        case "Scale":
            return new SqScaleValue(value, context);
        case "TimeDuration":
            return new SqTimeDurationValue(value, context);
        case "Void":
            return new SqVoidValue(value, context);
        case "Domain":
            return new SqDomainValue(value, context);
        case "Input":
            return new SqInputValue(value, context);
        default:
            throw new Error(`Unknown value ${JSON.stringify(value)}`);
    }
}
export class SqAbstractValue {
    constructor(_value, context) {
        this._value = _value;
        this.context = context;
    }
    toString() {
        return this._value.toString();
    }
    publicName() {
        return this._value.publicName;
    }
}
export class SqArrayValue extends SqAbstractValue {
    constructor() {
        super(...arguments);
        this.tag = "Array";
    }
    get value() {
        return new SqArray(this._value.value, this.context);
    }
    asJS() {
        return this.value.getValues().map((value) => value.asJS());
    }
}
export class SqBoolValue extends SqAbstractValue {
    constructor() {
        super(...arguments);
        this.tag = "Bool";
    }
    get value() {
        return this._value.value;
    }
    asJS() {
        return this.value;
    }
}
export class SqDateValue extends SqAbstractValue {
    constructor() {
        super(...arguments);
        this.tag = "Date";
    }
    get value() {
        return this._value.value;
    }
    asJS() {
        return this.value;
    }
}
export class SqDistributionValue extends SqAbstractValue {
    constructor() {
        super(...arguments);
        this.tag = "Dist";
    }
    get value() {
        return wrapDistribution(this._value.value);
    }
    asJS() {
        return this.value;
    }
}
export class SqLambdaValue extends SqAbstractValue {
    constructor() {
        super(...arguments);
        this.tag = "Lambda";
    }
    static create(value) {
        return new SqLambdaValue(vLambda(value._value));
    }
    get value() {
        return new SqLambda(this._value.value, this.context);
    }
    asJS() {
        return this.value;
    }
}
export class SqNumberValue extends SqAbstractValue {
    constructor() {
        super(...arguments);
        this.tag = "Number";
    }
    static create(value) {
        return new SqNumberValue(vNumber(value));
    }
    get value() {
        return this._value.value;
    }
    asJS() {
        return this.value;
    }
}
export class SqDictValue extends SqAbstractValue {
    constructor() {
        super(...arguments);
        this.tag = "Dict";
    }
    get value() {
        return new SqDict(this._value.value, this.context);
    }
    asJS() {
        return new Map(this.value.entries().map(([k, v]) => [k, v.asJS()]));
    }
}
export class SqStringValue extends SqAbstractValue {
    constructor() {
        super(...arguments);
        this.tag = "String";
    }
    static create(value) {
        return new SqStringValue(vString(value));
    }
    get value() {
        return this._value.value;
    }
    asJS() {
        return this.value;
    }
}
export class SqTimeDurationValue extends SqAbstractValue {
    constructor() {
        super(...arguments);
        this.tag = "TimeDuration";
    }
    get value() {
        return this._value.value;
    }
    asJS() {
        return this._value.value;
    }
}
export class SqPlotValue extends SqAbstractValue {
    constructor() {
        super(...arguments);
        this.tag = "Plot";
    }
    get value() {
        return wrapPlot(this._value.value, this.context);
    }
    asJS() {
        return this.value;
    }
}
export class SqTableChartValue extends SqAbstractValue {
    constructor() {
        super(...arguments);
        this.tag = "TableChart";
    }
    get value() {
        return new SqTableChart(this._value.value, this.context);
    }
    asJS() {
        return this.value;
    }
}
export class SqCalculatorValue extends SqAbstractValue {
    constructor() {
        super(...arguments);
        this.tag = "Calculator";
    }
    get value() {
        return new SqCalculator(this._value.value, this.context);
    }
    asJS() {
        return this.value;
    }
}
export class SqScaleValue extends SqAbstractValue {
    constructor() {
        super(...arguments);
        this.tag = "Scale";
    }
    get value() {
        return wrapScale(this._value.value);
    }
    asJS() {
        return this.value;
    }
}
export class SqInputValue extends SqAbstractValue {
    constructor() {
        super(...arguments);
        this.tag = "Input";
    }
    get value() {
        return wrapInput(this._value.value);
    }
    asJS() {
        return this.value;
    }
}
export class SqVoidValue extends SqAbstractValue {
    constructor() {
        super(...arguments);
        this.tag = "Void";
    }
    get value() {
        return null;
    }
    asJS() {
        return null;
    }
}
export class SqDomainValue extends SqAbstractValue {
    constructor() {
        super(...arguments);
        this.tag = "Domain";
    }
    get value() {
        return wrapDomain(this._value.value);
    }
    asJS() {
        return this.value;
    }
}
export function toStringResult(result) {
    return `${result.ok ? "Ok" : "Error"}(${result.value.toString()})`;
}
