// @quri/squiggle-lang/dist/public/SqValue/SqCalculator.js
import * as Result from "../../utility/result.js";
import { SqOtherError } from "../SqError.js";
import { SqLambda } from "./SqLambda.js";
import { wrapValue } from "./index.js";
import { wrapInput } from "./SqInput.js";
export class SqCalculator {
    constructor(_value, context) {
        this._value = _value;
        this.context = context;
    }
    run(_arguments, env) {
        const sqLambda = new SqLambda(this._value.fn, undefined);
        const response = sqLambda.call(_arguments, env);
        const newContext = this.context && this.context.extend({ type: "calculator" });
        if (!newContext) {
            return Result.Err(new SqOtherError("Context creation for calculator failed."));
        }
        else if (!response.ok) {
            return response;
        }
        else {
            return Result.Ok(wrapValue(response.value._value, newContext));
        }
    }
    get title() {
        return this._value.title;
    }
    get description() {
        return this._value.description;
    }
    get autorun() {
        return this._value.autorun;
    }
    get sampleCount() {
        return this._value.sampleCount;
    }
    get hashString() {
        const rowData = JSON.stringify(this._value.inputs);
        const paramData = this._value.fn.toString() || "";
        return (rowData +
            paramData +
            this._value.description +
            this._value.title +
            this._value.autorun +
            this._value.sampleCount);
    }
    get inputs() {
        return this._value.inputs.map(wrapInput);
    }
}
