// @quri/squiggle-lang/dist/public/SqValue/SqInput.js
import { vInput } from "../../value/index.js";
export const wrapInput = (value) => {
    switch (value.type) {
        case "text":
            return new SqTextInput(value);
        case "textArea":
            return new SqTextAreaInput(value);
        case "checkbox":
            return new SqCheckboxInput(value);
        case "select":
            return new SqSelectInput(value);
    }
};
class SqAbstractInput {
    constructor(_value) {
        this._value = _value;
    }
    toString() {
        return vInput(this._value).toString();
    }
    get name() {
        return this._value.name;
    }
    get description() {
        return this._value.description;
    }
    get default() {
        return this._value.default;
    }
}
export class SqTextInput extends SqAbstractInput {
    constructor() {
        super(...arguments);
        this.tag = "text";
    }
}
export class SqTextAreaInput extends SqAbstractInput {
    constructor() {
        super(...arguments);
        this.tag = "textArea";
    }
}
export class SqCheckboxInput extends SqAbstractInput {
    constructor() {
        super(...arguments);
        this.tag = "checkbox";
    }
}
export class SqSelectInput extends SqAbstractInput {
    constructor() {
        super(...arguments);
        this.tag = "select";
    }
    get options() {
        return this._value.options;
    }
}
