// @quri/squiggle-lang/dist/public/SqValue/SqLambda.js
import { IRuntimeError } from "../../errors/IError.js";
import { getStdLib } from "../../library/index.js";
import { createContext } from "../../reducer/context.js";
import * as Result from "../../utility/result.js";
import { SqOtherError, SqRuntimeError } from "../SqError.js";
import { wrapDomain } from "./SqDomain.js";
import { wrapValue } from "./index.js";
function lambdaToSqLambdaSignatures(lambda) {
    switch (lambda.type) {
        case "UserDefinedLambda":
            return [
                lambda.parameters.map((param) => {
                    return {
                        name: param.name,
                        domain: param.domain ? wrapDomain(param.domain.value) : undefined,
                    };
                }),
            ];
        case "BuiltinLambda":
            return lambda.signatures().map((def) => def.map((p, index) => ({
                name: index.toString(),
                domain: undefined,
                typeName: p.getName(),
            })));
    }
}
export class SqLambda {
    constructor(_value, context) {
        this._value = _value;
        this.context = context;
    }
    static createFromStdlibName(name) {
        const value = getStdLib().get(name);
        if (!value) {
            throw new Error(`Name ${name} not found in stdlib`);
        }
        if (value.type !== "Lambda") {
            throw new Error(`Stdlib value ${name} is not a function`);
        }
        return new SqLambda(value.value);
    }
    parameterCounts() {
        return this._value.parameterCounts();
    }
    signatures() {
        return lambdaToSqLambdaSignatures(this._value);
    }
    call(args, env) {
        if (!env) {
            if (!this.context) {
                return Result.Err(new SqOtherError("Programmatically constructed lambda call requires env argument"));
            }
            env = this.context.project.getEnvironment();
        }
        const rawArgs = args.map((arg) => arg._value);
        try {
            const value = this._value.call(rawArgs, createContext(env));
            return Result.Ok(wrapValue(value));
        }
        catch (e) {
            return Result.Err(new SqRuntimeError(IRuntimeError.fromException(e)));
        }
    }
    toString() {
        return this._value.toString();
    }
    parameterString() {
        return this._value.parameterString();
    }
}
