// @quri/squiggle-lang/dist/public/SqValue/SqScale.js
import { vScale, SCALE_SYMLOG_DEFAULT_CONSTANT, SCALE_POWER_DEFAULT_CONSTANT, } from "../../value/index.js";
export const wrapScale = (value) => {
    switch (value.type) {
        case "linear":
            return SqLinearScale.create(value);
        case "log":
            return SqLogScale.create(value);
        case "symlog":
            return SqSymlogScale.create(value);
        case "power":
            return SqPowerScale.create(value);
    }
};
class SqAbstractScale {
    constructor(_value) {
        this._value = _value;
    }
    toString() {
        return vScale(this._value).toString();
    }
    get min() {
        return this._value.min;
    }
    get max() {
        return this._value.max;
    }
    get tickFormat() {
        return this._value.tickFormat;
    }
    get title() {
        return this._value.title;
    }
}
export class SqLinearScale extends SqAbstractScale {
    constructor() {
        super(...arguments);
        this.tag = "linear";
    }
    static create(args = {}) {
        return new SqLinearScale({ type: "linear", ...args });
    }
}
export class SqLogScale extends SqAbstractScale {
    constructor() {
        super(...arguments);
        this.tag = "log";
    }
    static create(args = {}) {
        return new SqLogScale({ type: "log", ...args });
    }
}
export class SqSymlogScale extends SqAbstractScale {
    constructor(args) {
        super({
            type: "symlog",
            ...args,
        });
        this.tag = "symlog";
        this._constant = args.constant ?? SCALE_SYMLOG_DEFAULT_CONSTANT;
    }
    static create(args) {
        return new SqSymlogScale(args);
    }
    get constant() {
        return this._constant;
    }
}
export class SqPowerScale extends SqAbstractScale {
    constructor(args) {
        super({
            type: "power",
            ...args,
        });
        this.tag = "power";
        this._exponent = args.exponent ?? SCALE_POWER_DEFAULT_CONSTANT;
    }
    static create(args) {
        return new SqPowerScale(args);
    }
    get exponent() {
        return this._exponent;
    }
}
