// @quri/squiggle-lang/dist/public/SqValue/SqTableChart.js
import * as Result from "../../utility/result.js";
import { SqOtherError } from "../SqError.js";
import { SqLambda } from "./SqLambda.js";
import { wrapValue } from "./index.js";
const wrapFn = ({ fn }) => {
    return new SqLambda(fn, undefined);
};
const getItem = (row, column, element, fn, env, context) => {
    const response = fn.call([element], env);
    const newContext = context && context.extend({ type: "cellAddress", value: { row, column } });
    if (response.ok && context) {
        return Result.Ok(wrapValue(response.value._value, newContext));
    }
    else if (response.ok) {
        return Result.Err(new SqOtherError("Context creation for table failed."));
    }
    else {
        return response;
    }
};
export class SqTableChart {
    constructor(_value, context) {
        this._value = _value;
        this.context = context;
    }
    item(rowI, columnI, env) {
        return getItem(rowI, columnI, wrapValue(this._value.data[rowI], this.context), wrapFn(this._value.columns[columnI]), env, this.context);
    }
    items(env) {
        const wrappedDataItems = this._value.data.map((r) => wrapValue(r, this.context));
        const wrappedFns = this._value.columns.map(wrapFn);
        return wrappedDataItems.map((item, rowI) => wrappedFns.map((fn, columnI) => getItem(rowI, columnI, item, fn, env, this.context)));
    }
    get rowCount() {
        return this._value.data.length;
    }
    get columnCount() {
        return this._value.columns.length;
    }
    get columnNames() {
        return this._value.columns.map(({ name }) => name);
    }
    get title() {
        return this._value.title;
    }
}
