// @quri/squiggle-lang/dist/errors/IError.js
import { FrameStack } from "../reducer/frameStack.js";
import { ErrorMessage, REJavaScriptExn, REOther } from "./messages.js";
export class IRuntimeError extends Error {
    constructor(m, frameStack) {
        super();
        this.m = m;
        this.frameStack = frameStack;
    }
    static fromMessage(message) {
        return new IRuntimeError(message, FrameStack.make());
    }
    static fromMessageWithFrameStack(message, frameStack) {
        return new IRuntimeError(message, frameStack);
    }
    static fromException(exn) {
        if (exn instanceof IRuntimeError) {
            return exn;
        }
        else if (exn instanceof ErrorMessage) {
            return IRuntimeError.fromMessage(exn);
        }
        else if (exn instanceof Error) {
            return IRuntimeError.fromMessage(new REJavaScriptExn(exn.message, exn.name));
        }
        else {
            return IRuntimeError.fromMessage(new REOther("Unknown exception"));
        }
    }
    toString() {
        return this.m.toString();
    }
    toStringWithDetails() {
        return (this.toString() +
            (this.frameStack.isEmpty()
                ? ""
                : "\nStack trace:\n" + this.frameStack.toString()));
    }
    getTopFrame() {
        return this.frameStack.getTopFrame();
    }
    getFrameArray() {
        return this.frameStack.toFrameArray();
    }
}
export function rethrowWithFrameStack(err, frameStack) {
    if (err instanceof IRuntimeError) {
        throw err;
    }
    else if (err instanceof ErrorMessage) {
        throw IRuntimeError.fromMessageWithFrameStack(err, frameStack);
    }
    else if (err instanceof Error) {
        throw IRuntimeError.fromMessageWithFrameStack(new REJavaScriptExn(err.message, err.name), frameStack);
    }
    else {
        throw IRuntimeError.fromMessageWithFrameStack(new REOther("Unknown exception"), frameStack);
    }
}
export class ICompileError extends Error {
    constructor(message, location) {
        super();
        this.message = message;
        this.location = location;
    }
    toString() {
        return this.message;
    }
    toStringWithDetails() {
        return (this.toString() +
            "\nLocation:\n  " +
            `at line ${this.location.start.line}, column ${this.location.start.column}, file ${this.location.source}`);
    }
}
