// @quri/squiggle-lang/dist/errors/messages.js
import { distErrorToString } from "../dist/DistError.js";
export class ErrorMessage extends Error {
}
export class REArityError extends ErrorMessage {
    constructor(fn, arity, usedArity) {
        super();
        this.fn = fn;
        this.arity = arity;
        this.usedArity = usedArity;
    }
    toString() {
        return `${this.arity} arguments expected. Instead ${this.usedArity} argument(s) were passed.`;
    }
}
export class REArrayIndexNotFound extends ErrorMessage {
    constructor(msg, index) {
        super();
        this.msg = msg;
        this.index = index;
    }
    toString() {
        return `${this.msg}: ${this.index}`;
    }
}
export class REDistributionError extends ErrorMessage {
    constructor(err) {
        super();
        this.err = err;
    }
    toString() {
        return `Distribution Math Error: ${distErrorToString(this.err)}`;
    }
}
export class REExpectedType extends ErrorMessage {
    constructor(typeName, valueString) {
        super();
        this.typeName = typeName;
        this.valueString = valueString;
    }
    toString() {
        return `Expected type: ${this.typeName} but got: ${this.valueString}`;
    }
}
export class RENotAFunction extends ErrorMessage {
    constructor(value) {
        super();
        this.value = value;
    }
    toString() {
        return `${this.value} is not a function`;
    }
}
export class REOperationError extends ErrorMessage {
    constructor(err) {
        super();
        this.err = err;
    }
    toString() {
        return `Math Error: ${this.err.toString()}`;
    }
}
export class REDictPropertyNotFound extends ErrorMessage {
    constructor(msg, index) {
        super();
        this.msg = msg;
        this.index = index;
    }
    toString() {
        return `${this.msg}: ${this.index}`;
    }
}
export class RESymbolNotFound extends ErrorMessage {
    constructor(symbolName) {
        super();
        this.symbolName = symbolName;
    }
    toString() {
        return `${this.symbolName} is not defined`;
    }
}
export class RESyntaxError extends ErrorMessage {
    constructor(desc) {
        super();
        this.desc = desc;
    }
    toString() {
        return `Syntax Error: ${this.desc}`;
    }
}
export class RETodo extends ErrorMessage {
    constructor(msg) {
        super();
        this.msg = msg;
    }
    toString() {
        return `TODO: ${this.msg}`;
    }
}
export class REDomainError extends ErrorMessage {
    constructor(value, domain) {
        super();
        this.value = value;
        this.domain = domain;
    }
    toString() {
        return `Domain Error: Parameter ${this.value} must be in domain ${this.domain}`;
    }
}
export class REJavaScriptExn extends ErrorMessage {
    constructor(msg, name) {
        super();
        this.msg = msg;
        this.name = name;
    }
    toString() {
        let answer = `JS Exception: ${this.name}`;
        if (this.msg.length)
            answer += `: ${this.msg}`;
        return answer;
    }
}
export class REArgumentError extends ErrorMessage {
    constructor(msg) {
        super(msg);
        this.msg = msg;
    }
    toString() {
        return `Argument Error: ${this.message}`;
    }
}
export class REOther extends ErrorMessage {
    constructor(msg) {
        super(msg);
        this.msg = msg;
    }
    toString() {
        return `Error: ${this.message}`;
    }
}
export class REAmbiguous extends ErrorMessage {
    constructor(msg) {
        super(msg);
        this.msg = msg;
    }
    toString() {
        return `Ambiguous Error: ${this.message}`;
    }
}
