// @quri/squiggle-lang/dist/reducer/frameStack.js
export const topFrameName = "<top>";
export class Frame {
    isFrame() {
        return 1;
    }
    constructor(name, location) {
        this.name = name;
        this.location = location;
    }
    toString() {
        return (this.name +
            (this.location
                ? ` at line ${this.location.start.line}, column ${this.location.start.column}, file ${this.location.source}`
                : ""));
    }
}
export class FrameStack {
    constructor(frame, parent) {
        this.frame = frame;
        this.parent = parent;
    }
    static make() {
        return new FrameStack(new Frame("<root>"));
    }
    extend(name, location) {
        return new FrameStack(new Frame(name, location), this);
    }
    toString() {
        return this.toFrameArray()
            .map((f) => "  " + f.toString())
            .join("\n");
    }
    toFrameArray() {
        const result = [];
        let t = this;
        while (t && t.frame) {
            if (!t.parent)
                break;
            result.push(t.frame);
            t = t.parent;
        }
        return result;
    }
    getTopFrame() {
        return this.parent ? this.frame : undefined;
    }
    isEmpty() {
        return this.parent === undefined;
    }
}
