// @quri/squiggle-lang/dist/reducer/lambda.js
import * as IError from "../errors/IError.js";
import { REArityError, REDomainError, REOther } from "../errors/messages.js";
import * as Context from "./context.js";
import { fnDefinitionToString, tryCallFnDefinition, } from "../library/registry/fnDefinition.js";
import uniq from "lodash/uniq.js";
import { sort } from "../utility/E_A_Floats.js";
export class BaseLambda {
    constructor(body) {
        this.body = body;
    }
    callFrom(args, context, ast) {
        const newContext = {
            stack: context.stack,
            environment: context.environment,
            frameStack: context.frameStack.extend(Context.currentFunctionName(context), ast?.location),
            evaluate: context.evaluate,
            inFunction: this,
        };
        try {
            return this.body(args, newContext);
        }
        catch (e) {
            IError.rethrowWithFrameStack(e, newContext.frameStack);
        }
    }
    call(args, context) {
        return this.callFrom(args, context, undefined);
    }
}
export class UserDefinedLambda extends BaseLambda {
    constructor(name, parameters, stack, body, location) {
        const lambda = (args, context) => {
            const argsLength = args.length;
            const parametersLength = parameters.length;
            if (argsLength !== parametersLength) {
                throw new REArityError(undefined, parametersLength, argsLength);
            }
            let localStack = stack;
            for (let i = 0; i < parametersLength; i++) {
                const parameter = parameters[i];
                localStack = localStack.push(parameter.name, args[i]);
                if (parameter.domain && !parameter.domain.value.includes(args[i])) {
                    throw new REDomainError(args[i], parameter.domain);
                }
            }
            const lambdaContext = {
                stack: localStack,
                environment: context.environment,
                frameStack: context.frameStack,
                evaluate: context.evaluate,
                inFunction: context.inFunction,
            };
            const [value] = context.evaluate(body, lambdaContext);
            return value;
        };
        super(lambda);
        this.type = "UserDefinedLambda";
        this.name = name;
        this.parameters = parameters;
        this.location = location;
    }
    getName() {
        return this.name || "<anonymous>";
    }
    _getParameterNames() {
        return this.parameters.map((parameter) => parameter.name);
    }
    parameterString() {
        return this._getParameterNames().join(",");
    }
    toString() {
        return `lambda(${this._getParameterNames().join(",")}=>internal code)`;
    }
    parameterCounts() {
        return [this.parameters.length];
    }
    parameterCountString() {
        return this.parameters.length.toString();
    }
}
export class BuiltinLambda extends BaseLambda {
    constructor(name, signatures) {
        super((args, context) => this._call(args, context));
        this.name = name;
        this.type = "BuiltinLambda";
        this._definitions = signatures;
    }
    getName() {
        return this.name;
    }
    toString() {
        return this.name;
    }
    parameterString() {
        return this._definitions.map(fnDefinitionToString).join(" | ");
    }
    parameterCounts() {
        return sort(uniq(this._definitions.map((d) => d.inputs.length)));
    }
    parameterCountString() {
        return `[${this.parameterCounts().join(",")}]`;
    }
    signatures() {
        return this._definitions.map((d) => d.inputs);
    }
    _call(args, context) {
        const signatures = this._definitions;
        const showNameMatchDefinitions = () => {
            const defsString = signatures
                .map(fnDefinitionToString)
                .map((def) => `  ${this.name}${def}\n`)
                .join("");
            return `There are function matches for ${this.name}(), but with different arguments:\n${defsString}`;
        };
        for (const signature of signatures) {
            const callResult = tryCallFnDefinition(signature, args, context);
            if (callResult !== undefined) {
                return callResult;
            }
        }
        throw new REOther(showNameMatchDefinitions());
    }
}
