// @quri/squiggle-lang/dist/public/SqValue/SqDistribution/index.js
import { SampleSetDist } from "../../../dist/SampleSetDist/index.js";
import * as Result from "../../../utility/result.js";
import { Ok } from "../../../utility/result.js";
import { SymbolicDist } from "../../../dist/SymbolicDist.js";
import { PointSetDist } from "../../../dist/PointSetDist.js";
import { SqDistributionError } from "./SqDistributionError.js";
import { wrapPointSet } from "./SqPointSet.js";
export var SqDistributionTag;
(function (SqDistributionTag) {
    SqDistributionTag["PointSet"] = "PointSet";
    SqDistributionTag["SampleSet"] = "SampleSet";
    SqDistributionTag["Symbolic"] = "Symbolic";
})(SqDistributionTag || (SqDistributionTag = {}));
export function wrapDistribution(value) {
    if (value instanceof SymbolicDist) {
        return new SqSymbolicDistribution(value);
    }
    else if (value instanceof SampleSetDist) {
        return new SqSampleSetDistribution(value);
    }
    else if (value instanceof PointSetDist) {
        return new SqPointSetDistribution(value);
    }
    throw new Error(`Unknown value ${value}`);
}
export class SqAbstractDistribution {
    constructor(_value) {
        this._value = _value;
    }
    pointSet(env) {
        const innerResult = this._value.toPointSetDist(env);
        return Result.fmap2(innerResult, (dist) => wrapPointSet(dist.pointSet), (e) => new SqDistributionError(e));
    }
    asSampleSetDist(env) {
        const innerResult = SampleSetDist.fromDist(this._value, env);
        return Result.fmap2(innerResult, (dist) => new SqSampleSetDistribution(dist), (e) => new SqDistributionError(e));
    }
    toString() {
        return this._value.toString();
    }
    toSparkline(env) {
        return this._value.toSparkline(20, env);
    }
    mean(env) {
        return this._value.mean();
    }
    integralSum() {
        return this._value.integralSum();
    }
    isNormalized() {
        return this._value.isNormalized();
    }
    pdf(env, n) {
        return Result.fmap2(this._value.pdf(n, { env }), (v) => v, (e) => new SqDistributionError(e));
    }
    cdf(env, n) {
        return Ok(this._value.cdf(n));
    }
    inv(env, n) {
        return Ok(this._value.inv(n));
    }
    stdev(env) {
        return Result.fmap2(this._value.stdev(), (v) => v, (e) => new SqDistributionError(e));
    }
}
export class SqPointSetDistribution extends SqAbstractDistribution {
    constructor() {
        super(...arguments);
        this.tag = SqDistributionTag.PointSet;
    }
}
export class SqSampleSetDistribution extends SqAbstractDistribution {
    constructor() {
        super(...arguments);
        this.tag = SqDistributionTag.SampleSet;
    }
    getSamples() {
        return this._value.samples;
    }
}
export class SqSymbolicDistribution extends SqAbstractDistribution {
    constructor() {
        super(...arguments);
        this.tag = SqDistributionTag.Symbolic;
    }
}
