// @quri/squiggle-lang/dist/public/SqValue/SqDistribution/SqPointSet.js
import zipWith from "lodash/zipWith.js";
import { PointSetDist } from "../../../dist/PointSetDist.js";
import { ContinuousShape } from "../../../PointSet/Continuous.js";
import { DiscreteShape } from "../../../PointSet/Discrete.js";
import { MixedShape } from "../../../PointSet/Mixed.js";
import { SqPointSetDistribution } from "./index.js";
var Tag;
(function (Tag) {
    Tag["Mixed"] = "Mixed";
    Tag["Discrete"] = "Discrete";
    Tag["Continuous"] = "Continuous";
})(Tag || (Tag = {}));
function shapePoints(x) {
    const xs = x.xyShape.xs;
    const ys = x.xyShape.ys;
    return zipWith(xs, ys, (x, y) => ({ x, y }));
}
export function wrapPointSet(value) {
    if (value instanceof ContinuousShape) {
        return new SqContinuousPointSet(value);
    }
    else if (value instanceof DiscreteShape) {
        return new SqDiscretePointSet(value);
    }
    else if (value instanceof MixedShape) {
        return new SqMixedPointSet(value);
    }
    throw new Error(`Unknown PointSet shape ${value}`);
}
class SqAbstractPointSet {
    constructor(_value) { }
}
export class SqMixedPointSet {
    constructor(_value) {
        this._value = _value;
        this.tag = Tag.Mixed;
    }
    get value() {
        return this._value;
    }
    asShape() {
        const v = this.value;
        return {
            discrete: shapePoints(v.discrete),
            continuous: shapePoints(v.continuous),
        };
    }
    asDistribution() {
        return new SqPointSetDistribution(new PointSetDist(this.value));
    }
}
export class SqDiscretePointSet {
    constructor(_value) {
        this._value = _value;
        this.tag = Tag.Discrete;
    }
    get value() {
        return this._value;
    }
    asShape() {
        const v = this.value;
        return {
            discrete: shapePoints(v),
            continuous: [],
        };
    }
    asDistribution() {
        return new SqPointSetDistribution(new PointSetDist(this.value.toMixed()));
    }
}
export class SqContinuousPointSet {
    constructor(_value) {
        this._value = _value;
        this.tag = Tag.Continuous;
    }
    get value() {
        return this._value;
    }
    asShape() {
        const v = this.value;
        return {
            discrete: [],
            continuous: shapePoints(v),
        };
    }
    asDistribution() {
        return new SqPointSetDistribution(new PointSetDist(this.value.toMixed()));
    }
}
