// @quri/squiggle-lang/dist/src/rescript/Reducer/Reducer_FrameStack.bs.js
'use strict';
var Belt_List = require("@rescript/std/lib/js/belt_List.js");
var Reducer_T$QuriSquiggleLang = require("./Reducer_T.bs.js");
function toString(param) {
    var $$location = param.location;
    return param.name + ($$location !== undefined ? " at line " + $$location.start.line.toString() + ", column " + $$location.start.column.toString() + "" : "");
}
function getLocation(t) {
    return t.location;
}
function getName(t) {
    return t.name;
}
var Frame = {
    toString: toString,
    getLocation: getLocation,
    getName: getName
};
function make(param) {
    return 0;
}
function extend(t, name, $$location) {
    return Belt_List.add(t, {
        name: name,
        location: $$location
    });
}
function makeSingleFrameStack($$location) {
    return extend(0, Reducer_T$QuriSquiggleLang.topFrameName, $$location);
}
function toString$1(t) {
    return Belt_List.toArray(Belt_List.map(t, (function (s) {
        return "  " + toString(s);
    }))).join("\n");
}
var toFrameArray = Belt_List.toArray;
var getTopFrame = Belt_List.head;
function isEmpty(t) {
    return Belt_List.head(t) !== undefined;
}
exports.Frame = Frame;
exports.make = make;
exports.extend = extend;
exports.makeSingleFrameStack = makeSingleFrameStack;
exports.toString = toString$1;
exports.toFrameArray = toFrameArray;
exports.getTopFrame = getTopFrame;
exports.isEmpty = isEmpty;
