// @quri/squiggle-lang/dist/src/rescript/Reducer/Reducer_Namespace.bs.js
'use strict';
var Belt_MapString = require("@rescript/std/lib/js/belt_MapString.js");
var E_A$QuriSquiggleLang = require("../Utility/E/E_A.bs.js");
var Reducer_Value$QuriSquiggleLang = require("./Reducer_Value.bs.js");
function make(param) {
}
var get = Belt_MapString.get;
var set = Belt_MapString.set;
function mergeFrom(from, to) {
    return Belt_MapString.reduce(to, from, Belt_MapString.set);
}
function mergeMany(namespaces) {
    return E_A$QuriSquiggleLang.reduce(namespaces, undefined, mergeFrom);
}
function toString(namespace) {
    return E_A$QuriSquiggleLang.fmap(Belt_MapString.toArray(namespace), (function (param) {
        return "" + param[0] + ": " + Reducer_Value$QuriSquiggleLang.toString(param[1]) + "";
    })).toString();
}
var fromArray = Belt_MapString.fromArray;
function toMap(namespace) {
    return namespace;
}
function toRecord(namespace) {
    return {
        TAG: 7,
        _0: namespace,
        [Symbol.for("name")]: "IEvRecord"
    };
}
exports.make = make;
exports.get = get;
exports.set = set;
exports.mergeFrom = mergeFrom;
exports.mergeMany = mergeMany;
exports.toString = toString;
exports.fromArray = fromArray;
exports.toMap = toMap;
exports.toRecord = toRecord;
