// @quri/squiggle-lang/dist/fr/danger.js
import jstat from "jstat";
import { scaleLog, scaleLogWithThreshold, } from "../dist/distOperations/index.js";
import { scaleMultiply, scalePower, } from "../dist/distOperations/scaleOperations.js";
import { REArgumentError, REOther } from "../errors/messages.js";
import { makeDefinition } from "../library/registry/fnDefinition.js";
import { frAny, frArray, frDist, frLambda, frNumber, } from "../library/registry/frTypes.js";
import { FnFactory, unpackDistResult, distResultToValue, makeTwoArgsDist, makeOneArgDist, } from "../library/registry/helpers.js";
import * as E_A from "../utility/E_A.js";
import { vArray, vNumber } from "../value/index.js";
import * as SymbolicDist from "../dist/SymbolicDist.js";
const { factorial } = jstat;
const maker = new FnFactory({
    nameSpace: "Danger",
    requiresNamespace: true,
});
function combinations(arr, k) {
    if (k === 0)
        return [[]];
    if (k === arr.length)
        return [arr];
    const withoutFirst = combinations(arr.slice(1), k);
    const withFirst = combinations(arr.slice(1), k - 1).map((comb) => [
        arr[0],
        ...comb,
    ]);
    return withFirst.concat(withoutFirst);
}
function allCombinations(arr) {
    let allCombs = [];
    for (let k = 1; k <= arr.length; k++) {
        allCombs = allCombs.concat(combinations(arr, k));
    }
    return allCombs;
}
const choose = (n, k) => factorial(n) / (factorial(n - k) * factorial(k));
const combinatoricsLibrary = [
    maker.nn2n({
        name: "laplace",
        examples: [`Danger.laplace(1, 20)`],
        fn: (successes, trials) => (successes + 1) / (trials + 2),
    }),
    maker.n2n({
        name: "factorial",
        examples: [`Danger.factorial(20)`],
        fn: factorial,
    }),
    maker.nn2n({
        name: "choose",
        examples: [`Danger.choose(1, 20)`],
        fn: choose,
    }),
    maker.make({
        name: "binomial",
        output: "Number",
        examples: [`Danger.binomial(1, 20, 0.5)`],
        definitions: [
            makeDefinition([frNumber, frNumber, frNumber], ([n, k, p]) => vNumber(choose(n, k) * Math.pow(p, k) * Math.pow(1 - p, n - k))),
        ],
    }),
];
const integrateFunctionBetweenWithNumIntegrationPoints = (lambda, min, max, numIntegrationPoints, context) => {
    const applyFunctionAtFloatToFloatOption = (point) => {
        const result = lambda.call([vNumber(point)], context);
        if (result.type === "Number") {
            return result.value;
        }
        throw new REOther("Error 1 in Danger.integrate. It's possible that your function doesn't return a number, try definining auxiliaryFunction(x) = mean(yourFunction(x)) and integrate auxiliaryFunction instead");
    };
    const numTotalPoints = numIntegrationPoints | 0;
    const numInnerPoints = numTotalPoints - 2;
    const numOuterPoints = 2;
    const totalWeight = max - min;
    const weightForAnInnerPoint = totalWeight / (numTotalPoints - 1);
    const weightForAnOuterPoint = totalWeight / (numTotalPoints - 1) / 2;
    const innerPointIncrement = (max - min) / (numTotalPoints - 1);
    const innerXs = E_A.makeBy(numInnerPoints, (i) => min + (i + 1) * innerPointIncrement);
    const ys = innerXs.map((x) => applyFunctionAtFloatToFloatOption(x));
    const verbose = false;
    if (verbose) {
        console.log("numTotalPoints", numTotalPoints);
        console.log("numInnerPoints", numInnerPoints);
        console.log("numOuterPoints", numOuterPoints);
        console.log("totalWeight", totalWeight);
        console.log("weightForAnInnerPoint", weightForAnInnerPoint);
        console.log("weightForAnOuterPoint", weightForAnOuterPoint);
        console.log("weightForAnInnerPoint * numInnerPoints + weightForAnOuterPoint * numOuterPoints", weightForAnInnerPoint * numInnerPoints +
            weightForAnOuterPoint * numOuterPoints);
        console.log("sum of weights == totalWeight", weightForAnInnerPoint * numInnerPoints +
            weightForAnOuterPoint * numOuterPoints ===
            totalWeight);
        console.log("innerPointIncrement", innerPointIncrement);
        console.log("innerXs", innerXs);
        console.log("ys", ys);
    }
    const innerPointsSum = ys.reduce((a, b) => a + b, 0);
    const yMin = applyFunctionAtFloatToFloatOption(min);
    const yMax = applyFunctionAtFloatToFloatOption(max);
    const result = (yMin + yMax) * weightForAnOuterPoint +
        innerPointsSum * weightForAnInnerPoint;
    return vNumber(result);
};
const integrationLibrary = [
    maker.make({
        name: "integrateFunctionBetweenWithNumIntegrationPoints",
        requiresNamespace: false,
        output: "Number",
        examples: [
            `Danger.integrateFunctionBetweenWithNumIntegrationPoints({|x| x+1}, 1, 10, 10)`,
        ],
        definitions: [
            makeDefinition([frLambda, frNumber, frNumber, frNumber], ([lambda, min, max, numIntegrationPoints], context) => {
                if (numIntegrationPoints === 0) {
                    throw new REOther("Integration error 4 in Danger.integrate: Increment can't be 0.");
                }
                return integrateFunctionBetweenWithNumIntegrationPoints(lambda, min, max, numIntegrationPoints, context);
            }),
        ],
    }),
    maker.make({
        name: "integrateFunctionBetweenWithEpsilon",
        requiresNamespace: false,
        output: "Number",
        examples: [
            `Danger.integrateFunctionBetweenWithEpsilon({|x| x+1}, 1, 10, 0.1)`,
        ],
        definitions: [
            makeDefinition([frLambda, frNumber, frNumber, frNumber], ([lambda, min, max, epsilon], context) => {
                if (epsilon === 0) {
                    throw new REOther("Integration error in Danger.integrate: Increment can't be 0.");
                }
                return integrateFunctionBetweenWithNumIntegrationPoints(lambda, min, max, (max - min) / epsilon, context);
            }),
        ],
    }),
];
const findBiggestElementIndex = (xs) => xs.reduce((acc, newElement, index) => {
    if (newElement > xs[acc]) {
        return index;
    }
    else {
        return acc;
    }
}, 0);
const diminishingReturnsLibrary = [
    maker.make({
        name: "optimalAllocationGivenDiminishingMarginalReturnsForManyFunctions",
        output: "Array",
        requiresNamespace: false,
        examples: [
            `Danger.optimalAllocationGivenDiminishingMarginalReturnsForManyFunctions([{|x| x+1}, {|y| 10}], 100, 0.01)`,
        ],
        definitions: [
            makeDefinition([frArray(frLambda), frNumber, frNumber], ([lambdas, funds, approximateIncrement], context) => {
                if (lambdas.length <= 1) {
                    throw new REOther("Error in Danger.optimalAllocationGivenDiminishingMarginalReturnsForManyFunctions, number of functions should be greater than 1.");
                }
                if (funds <= 0) {
                    throw new REOther("Error in Danger.optimalAllocationGivenDiminishingMarginalReturnsForManyFunctions, funds should be greater than 0.");
                }
                if (approximateIncrement <= 0) {
                    throw new REOther("Error in Danger.optimalAllocationGivenDiminishingMarginalReturnsForManyFunctions, approximateIncrement should be greater than 0.");
                }
                if (approximateIncrement >= funds) {
                    throw new REOther("Error in Danger.optimalAllocationGivenDiminishingMarginalReturnsForManyFunctions, approximateIncrement should be smaller than funds amount.");
                }
                const applyFunctionAtPoint = (lambda, point) => {
                    const lambdaResult = lambda.call([vNumber(point)], context);
                    if (lambdaResult.type === "Number") {
                        return lambdaResult.value;
                    }
                    throw new REOther("Error 1 in Danger.optimalAllocationGivenDiminishingMarginalReturnsForManyFunctions. It's possible that your function doesn't return a number, try definining auxiliaryFunction(x) = mean(yourFunction(x)) and integrate auxiliaryFunction instead");
                };
                const numDivisions = Math.round(funds / approximateIncrement);
                const increment = funds / numDivisions;
                const arrayOfIncrements = new Array(numDivisions).fill(increment);
                const initAccumulator = {
                    optimalAllocations: new Array(lambdas.length).fill(0),
                    currentMarginalReturns: lambdas.map((lambda) => applyFunctionAtPoint(lambda, 0)),
                };
                const optimalAllocationEndAccumulator = arrayOfIncrements.reduce((acc, newIncrement) => {
                    const oldMarginalReturns = acc.currentMarginalReturns;
                    const indexOfBiggestDMR = findBiggestElementIndex(oldMarginalReturns);
                    const newOptimalAllocations = [...acc.optimalAllocations];
                    const newOptimalAllocationsi = newOptimalAllocations[indexOfBiggestDMR] + newIncrement;
                    newOptimalAllocations[indexOfBiggestDMR] = newOptimalAllocationsi;
                    const lambdai = lambdas[indexOfBiggestDMR];
                    const newMarginalResultsLambdai = applyFunctionAtPoint(lambdai, newOptimalAllocationsi);
                    const newCurrentMarginalReturns = [...oldMarginalReturns];
                    newCurrentMarginalReturns[indexOfBiggestDMR] =
                        newMarginalResultsLambdai;
                    const newAcc = {
                        optimalAllocations: newOptimalAllocations,
                        currentMarginalReturns: newCurrentMarginalReturns,
                    };
                    return newAcc;
                }, initAccumulator);
                return vArray(optimalAllocationEndAccumulator.optimalAllocations.map(vNumber));
            }),
        ],
    }),
];
const mapYLibrary = [
    maker.d2d({
        name: "mapYLog",
        fn: (dist, env) => unpackDistResult(scaleLog(dist, Math.E, { env })),
    }),
    maker.d2d({
        name: "mapYLog10",
        fn: (dist, env) => unpackDistResult(scaleLog(dist, 10, { env })),
    }),
    maker.dn2d({
        name: "mapYLog",
        fn: (dist, x, env) => unpackDistResult(scaleLog(dist, x, { env })),
    }),
    maker.make({
        name: "mapYLogWithThreshold",
        definitions: [
            makeDefinition([frDist, frNumber, frNumber], ([dist, base, eps], { environment }) => distResultToValue(scaleLogWithThreshold(dist, {
                env: environment,
                eps,
                base,
            }))),
        ],
    }),
    maker.make({
        name: "combinations",
        definitions: [
            makeDefinition([frArray(frAny), frNumber], ([elements, n]) => {
                if (n > elements.length) {
                    throw new REArgumentError(`Combinations of length ${n} were requested, but full list is only ${elements.length} long.`);
                }
                return vArray(combinations(elements, n).map((v) => vArray(v)));
            }),
        ],
    }),
    maker.make({
        name: "allCombinations",
        definitions: [
            makeDefinition([frArray(frAny)], ([elements]) => {
                return vArray(allCombinations(elements).map((v) => vArray(v)));
            }),
        ],
    }),
    maker.dn2d({
        name: "mapYMultiply",
        fn: (dist, f, env) => unpackDistResult(scaleMultiply(dist, f, { env })),
    }),
    maker.dn2d({
        name: "mapYPow",
        fn: (dist, f, env) => unpackDistResult(scalePower(dist, f, { env })),
    }),
    maker.d2d({
        name: "mapYExp",
        fn: (dist, env) => unpackDistResult(scalePower(dist, Math.E, { env })),
    }),
    maker.make({
        name: "binomialDist",
        examples: ["Danger.binomialDist(8, 0.5)"],
        definitions: [makeTwoArgsDist((n, p) => SymbolicDist.Binomial.make(n, p))],
    }),
    maker.make({
        name: "poissonDist",
        examples: ["Danger.poissonDist(10)"],
        definitions: [
            makeOneArgDist((lambda) => SymbolicDist.Poisson.make(lambda)),
        ],
    }),
];
export const library = [
    ...combinatoricsLibrary,
    ...integrationLibrary,
    ...diminishingReturnsLibrary,
    ...mapYLibrary,
];
