// @quri/squiggle-lang/dist/public/SqError.js
class SqAbstractError {
}
export class SqFrame {
    constructor(_value) {
        this._value = _value;
    }
    name() {
        return this._value.name;
    }
    location() {
        return this._value.location;
    }
}
export class SqRuntimeError extends SqAbstractError {
    constructor(_value) {
        super();
        this._value = _value;
        this.tag = "runtime";
    }
    toString() {
        return this._value.toString();
    }
    toStringWithDetails() {
        return this._value.toStringWithDetails();
    }
    getTopFrame() {
        const frame = this._value.getTopFrame();
        return frame ? new SqFrame(frame) : undefined;
    }
    getFrameArray() {
        const frames = this._value.getFrameArray();
        return frames.map((frame) => new SqFrame(frame));
    }
    location() {
        return this.getTopFrame()?.location();
    }
}
export class SqCompileError extends SqAbstractError {
    constructor(_value) {
        super();
        this._value = _value;
        this.tag = "compile";
    }
    toString() {
        return this._value.toString();
    }
    toStringWithDetails() {
        return this._value.toStringWithDetails();
    }
    location() {
        return this._value.location;
    }
}
export class SqOtherError extends SqAbstractError {
    constructor(_value) {
        super();
        this._value = _value;
        this.tag = "other";
    }
    toString() {
        return this._value;
    }
    toStringWithDetails() {
        return this._value;
    }
}
