// @quri/squiggle-lang/dist/public/SqValueContext.js
import { isBindingStatement } from "../ast/utils.js";
export class SqValueContext {
    constructor(props) {
        this.project = props.project;
        this.sourceId = props.sourceId;
        this.source = props.source;
        this.ast = props.ast;
        this.valueAst = props.valueAst;
        this.valueAstIsPrecise = props.valueAstIsPrecise;
        this.path = props.path;
    }
    extend(item) {
        let ast = this.valueAst;
        let newAst;
        const itemisNotTableIndexOrCalculator = item.type !== "cellAddress" && item.type !== "calculator";
        if (this.valueAstIsPrecise && itemisNotTableIndexOrCalculator) {
            while (true) {
                if (ast.type === "Block") {
                    ast = ast.statements[ast.statements.length - 1];
                }
                else if (ast.type === "KeyValue") {
                    ast = ast.value;
                }
                else if (isBindingStatement(ast)) {
                    ast = ast.value;
                }
                else {
                    break;
                }
            }
            switch (ast.type) {
                case "Program": {
                    if (this.path.root === "bindings") {
                        newAst = ast.symbols[item.value];
                        break;
                    }
                    break;
                }
                case "Dict":
                    newAst = ast.symbols[item.value];
                    break;
                case "Array":
                    if (typeof item === "number") {
                        const element = ast.elements[item];
                        if (element) {
                            newAst = element;
                        }
                    }
                    break;
            }
        }
        return new SqValueContext({
            project: this.project,
            sourceId: this.sourceId,
            source: this.source,
            ast: this.ast,
            valueAst: newAst ?? this.valueAst,
            valueAstIsPrecise: Boolean(newAst),
            path: this.path.extend(item),
        });
    }
    findLocation() {
        return this.valueAst.location;
    }
    docstring() {
        if (!this.valueAstIsPrecise) {
            return;
        }
        if (!this.ast.comments.length) {
            return;
        }
        if (this.path.root === "bindings" && this.path.isRoot()) {
            return;
        }
        const valueStarts = this.valueAst.location.start.offset;
        let a = 0, b = this.ast.comments.length - 1;
        while (a < b) {
            const m = Math.floor((a + b) / 2);
            const commentToCheck = this.ast.comments[m + 1];
            if (commentToCheck.location.end.offset > valueStarts) {
                b = m;
            }
            else {
                a = m + 1;
            }
        }
        const comment = this.ast.comments[a];
        const commentEnds = comment.location.end.offset;
        if (commentEnds > valueStarts) {
            return;
        }
        if (comment.type !== "blockComment") {
            return;
        }
        const match = comment.value.match(/^\*(?!\*)(.*)/s);
        if (!match) {
            return;
        }
        const ok = this.source.substring(commentEnds, valueStarts).match(/^\s*$/);
        if (ok) {
            return match[1].trim();
        }
    }
}
