// @quri/squiggle-lang/dist/src/rescript/ForTS/ForTS_SquiggleValue/ForTS_SquiggleValue.gen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTimeDuration = exports.getString = exports.getRecord = exports.getNumber = exports.getLambda = exports.getDistribution = exports.getDeclaration = exports.getDate = exports.getBool = exports.getArray = exports.toStringResult = exports.toString = exports.getTag = void 0;
const ForTS_SquiggleValueBS = require('./ForTS_SquiggleValue.bs');
exports.getTag = ForTS_SquiggleValueBS.getTag;
exports.toString = ForTS_SquiggleValueBS.toString;
const toStringResult = function (Arg1) {
    const result = ForTS_SquiggleValueBS.toStringResult(Arg1.tag === "Ok"
        ? { TAG: 0, _0: Arg1.value }
        : { TAG: 1, _0: Arg1.value });
    return result;
};
exports.toStringResult = toStringResult;
exports.getArray = ForTS_SquiggleValueBS.getArray;
exports.getBool = ForTS_SquiggleValueBS.getBool;
exports.getDate = ForTS_SquiggleValueBS.getDate;
exports.getDeclaration = ForTS_SquiggleValueBS.getDeclaration;
exports.getDistribution = ForTS_SquiggleValueBS.getDistribution;
exports.getLambda = ForTS_SquiggleValueBS.getLambda;
exports.getNumber = ForTS_SquiggleValueBS.getNumber;
exports.getRecord = ForTS_SquiggleValueBS.getRecord;
exports.getString = ForTS_SquiggleValueBS.getString;
exports.getTimeDuration = ForTS_SquiggleValueBS.getTimeDuration;
