// @quri/squiggle-lang/dist/value/domain.js
import { REArgumentError } from "../errors/messages.js";
class BaseDomain {
}
export class NumericRangeDomain extends BaseDomain {
    constructor(min, max) {
        super();
        this.min = min;
        this.max = max;
        this.type = "NumericRange";
    }
    toString() {
        return `Number.rangeDomain({ min: ${this.min}, max: ${this.max} })`;
    }
    includes(value) {
        return (value.type === "Number" &&
            value.value >= this.min &&
            value.value <= this.max);
    }
    isEqual(other) {
        return this.min === other.min && this.max === other.max;
    }
}
export function annotationToDomain(value) {
    if (value.type === "Domain") {
        return value.value;
    }
    if (value.type !== "Array") {
        throw new REArgumentError("Only array domains are supported");
    }
    if (value.value.length !== 2) {
        throw new REArgumentError("Expected two-value array");
    }
    const [min, max] = value.value;
    if (min.type !== "Number") {
        throw new REArgumentError("Min value is not a number");
    }
    if (max.type !== "Number") {
        throw new REArgumentError("Max value is not a number");
    }
    if (min.value >= max.value) {
        throw new REArgumentError(`The range minimum (${min.value}) must be lower than the range maximum (${max.value})`);
    }
    return new NumericRangeDomain(min.value, max.value);
}
