// @quri/squiggle-lang/dist/value/index.js
import isInteger from "lodash/isInteger.js";
import { REArrayIndexNotFound, REDictPropertyNotFound, REOther, } from "../errors/messages.js";
import * as DateTime from "../utility/DateTime.js";
import { ImmutableMap } from "../utility/immutableMap.js";
import { shuffle } from "../utility/E_A.js";
import lodashIsEqual from "lodash/isEqual.js";
class BaseValue {
    clone() {
        return Object.assign(Object.create(Object.getPrototypeOf(this)), this);
    }
}
class VArray extends BaseValue {
    constructor(value) {
        super();
        this.value = value;
        this.type = "Array";
        this.publicName = "List";
    }
    toString() {
        return "[" + this.value.map((v) => v.toString()).join(",") + "]";
    }
    get(key) {
        if (key.type === "Number") {
            if (!isInteger(key.value)) {
                throw new REArrayIndexNotFound("Array index must be an integer", key.value);
            }
            const index = key.value | 0;
            if (index >= 0 && index < this.value.length) {
                return this.value[index];
            }
            else {
                throw new REArrayIndexNotFound("Array index not found", index);
            }
        }
        throw new REOther("Can't access non-numerical key on an array");
    }
    flatten() {
        return new VArray(this.value.reduce((acc, v) => acc.concat(v.type === "Array" ? v.value : [v]), []));
    }
    shuffle() {
        return new VArray(shuffle(this.value));
    }
    isEqual(other) {
        if (this.value.length !== other.value.length) {
            return false;
        }
        for (let i = 0; i < this.value.length; i++) {
            isEqual(this.value[i], other.value[i]);
        }
        return true;
    }
}
export const vArray = (v) => new VArray(v);
class VBool extends BaseValue {
    constructor(value) {
        super();
        this.value = value;
        this.type = "Bool";
        this.publicName = "Boolean";
    }
    toString() {
        return String(this.value);
    }
    isEqual(other) {
        return this.value === other.value;
    }
}
export const vBool = (v) => new VBool(v);
class VDate extends BaseValue {
    constructor(value) {
        super();
        this.value = value;
        this.type = "Date";
        this.publicName = "Date";
    }
    toString() {
        return DateTime.Date.toString(this.value);
    }
    isEqual(other) {
        return this.value === other.value;
    }
}
export const vDate = (v) => new VDate(v);
class VDist extends BaseValue {
    constructor(value) {
        super();
        this.value = value;
        this.type = "Dist";
        this.publicName = "Distribution";
    }
    toString() {
        return this.value.toString();
    }
    isEqual(other) {
        return this.value.isEqual(other.value);
    }
}
export const vDist = (v) => new VDist(v);
class VLambda extends BaseValue {
    constructor(value) {
        super();
        this.value = value;
        this.type = "Lambda";
        this.publicName = "Function";
    }
    toString() {
        return this.value.toString();
    }
    get(key) {
        if (key.type === "String" && key.value === "parameters") {
            switch (this.value.type) {
                case "UserDefinedLambda":
                    return vArray(this.value.parameters.map((parameter) => {
                        const fields = [
                            ["name", vString(parameter.name)],
                        ];
                        if (parameter.domain) {
                            fields.push(["domain", parameter.domain]);
                        }
                        return vDict(ImmutableMap(fields));
                    }));
                case "BuiltinLambda":
                    throw new REOther("Can't access parameters on built in functions");
            }
        }
        throw new REOther("No such field");
    }
}
export const vLambda = (v) => new VLambda(v);
class VNumber extends BaseValue {
    constructor(value) {
        super();
        this.value = value;
        this.type = "Number";
        this.publicName = "Number";
    }
    toString() {
        return String(this.value);
    }
    isEqual(other) {
        return this.value === other.value;
    }
}
export const vNumber = (v) => new VNumber(v);
class VString extends BaseValue {
    constructor(value) {
        super();
        this.value = value;
        this.type = "String";
        this.publicName = "String";
    }
    toString() {
        return JSON.stringify(this.value);
    }
    isEqual(other) {
        return this.value === other.value;
    }
}
export const vString = (v) => new VString(v);
class VDict extends BaseValue {
    constructor(value) {
        super();
        this.value = value;
        this.type = "Dict";
        this.publicName = "Dictionary";
    }
    toString() {
        return ("{" +
            [...this.value.entries()]
                .map(([k, v]) => `${k}: ${v.toString()}`)
                .join(",") +
            "}");
    }
    get(key) {
        if (key.type === "String") {
            const result = this.value.get(key.value);
            if (!result) {
                throw new REDictPropertyNotFound("Dict property not found", key.value);
            }
            return result;
        }
        else {
            throw new REOther("Can't access non-string key on a dict");
        }
    }
    isEqual(other) {
        if (this.value.size !== other.value.size) {
            return false;
        }
        for (const [key, valueA] of this.value.entries()) {
            const valueB = other.value.get(key);
            if (!valueB) {
                return false;
            }
            if (!isEqual(valueA, valueB)) {
                return false;
            }
        }
        return true;
    }
}
export const vDict = (v) => new VDict(v);
class VTimeDuration extends BaseValue {
    constructor(value) {
        super();
        this.value = value;
        this.type = "TimeDuration";
        this.publicName = "Time Duration";
    }
    toString() {
        return DateTime.Duration.toString(this.value);
    }
    isEqual(other) {
        return this.value === other.value;
    }
}
export const vTimeDuration = (v) => new VTimeDuration(v);
function scaleIsEqual(valueA, valueB) {
    if (valueA.type !== valueB.type ||
        valueA.min !== valueB.min ||
        valueA.max !== valueB.max ||
        valueA.tickFormat !== valueB.tickFormat) {
        return false;
    }
    switch (valueA.type) {
        case "symlog":
            return (valueA.constant ===
                valueB.constant);
        case "power":
            return (valueA.exponent ===
                valueB.exponent);
        default:
            return true;
    }
}
export const SCALE_SYMLOG_DEFAULT_CONSTANT = 0.0001;
export const SCALE_POWER_DEFAULT_CONSTANT = 0.1;
class VScale extends BaseValue {
    constructor(value) {
        super();
        this.value = value;
        this.type = "Scale";
        this.publicName = "Scale";
    }
    toString() {
        switch (this.value.type) {
            case "linear":
                return "Linear scale";
            case "log":
                return "Logarithmic scale";
            case "symlog":
                return `Symlog scale ({constant: ${this.value.constant || SCALE_SYMLOG_DEFAULT_CONSTANT}})`;
            case "power":
                return `Power scale ({exponent: ${this.value.exponent || SCALE_POWER_DEFAULT_CONSTANT}})`;
        }
    }
    isEqual(other) {
        return scaleIsEqual(this.value, other.value);
    }
}
export const vScale = (scale) => new VScale(scale);
class VInput extends BaseValue {
    constructor(value) {
        super();
        this.value = value;
        this.type = "Input";
        this.publicName = "Input";
    }
    toString() {
        switch (this.value.type) {
            case "text":
                return "Text input";
            case "textArea":
                return "Text area input";
            case "checkbox":
                return "Check box input";
            case "select":
                return `Select input (${this.value.options.join(", ")})`;
        }
    }
    isEqual(other) {
        return lodashIsEqual(this.value, other.value);
    }
}
export const vInput = (input) => new VInput(input);
class VTableChart extends BaseValue {
    constructor(value) {
        super();
        this.value = value;
        this.type = "TableChart";
        this.publicName = "Table Chart";
    }
    toString() {
        return `Table with ${this.value.columns.length}x${this.value.data.length} elements`;
    }
}
export const vTableChart = (v) => new VTableChart(v);
class VCalculator extends BaseValue {
    constructor(value) {
        super();
        this.value = value;
        this.type = "Calculator";
        this.publicName = "Calculator";
        this.error = null;
        if (!value.fn.parameterCounts().includes(value.inputs.length)) {
            this.setError(`Calculator function needs ${value.fn.parameterCountString()} parameters, but ${value.inputs.length} fields were provided.`);
        }
        if (value.inputs.some((x) => x.name === "")) {
            this.setError(`Calculator field names can't be empty.`);
        }
        const fieldNames = value.inputs.map((f) => f.name);
        const uniqueNames = new Set(fieldNames);
        if (fieldNames.length !== uniqueNames.size) {
            this.setError(`Duplicate calculator field names found.`);
        }
    }
    setError(message) {
        this.error = new REOther(message);
    }
    getError() {
        return this.error;
    }
    toString() {
        return `Calculator`;
    }
}
export const vCalculator = (v) => new VCalculator(v);
class VPlot extends BaseValue {
    constructor(value) {
        super();
        this.value = value;
        this.type = "Plot";
        this.publicName = "Plot";
    }
    toString() {
        switch (this.value.type) {
            case "distributions":
                return `Plot containing ${this.value.distributions
                    .map((x) => x.name)
                    .join(", ")}`;
            case "numericFn":
                return `Plot for numeric function ${this.value.fn}`;
            case "distFn":
                return `Plot for dist function ${this.value.fn}`;
            case "scatter":
                return `Scatter plot for distributions ${this.value.xDist} and ${this.value.yDist}`;
            case "relativeValues":
                return `Plot for relative values ${this.value.ids.join(", ")}`;
        }
    }
    get(key) {
        if (key.type === "String" &&
            key.value === "fn" &&
            (this.value.type === "numericFn" ||
                this.value.type === "distFn" ||
                this.value.type === "relativeValues")) {
            return vLambda(this.value.fn);
        }
        throw new REOther("Trying to access non-existent field");
    }
}
export const vPlot = (plot) => new VPlot(plot);
export class VDomain extends BaseValue {
    constructor(value) {
        super();
        this.value = value;
        this.type = "Domain";
        this.publicName = "Domain";
    }
    toString() {
        return this.value.toString();
    }
    get(key) {
        if (key.type === "String" && this.value.type === "NumericRange") {
            if (key.value === "min") {
                return vNumber(this.value.min);
            }
            if (key.value === "max") {
                return vNumber(this.value.max);
            }
        }
        throw new REOther("Trying to access non-existent field");
    }
    isEqual(other) {
        return this.value.isEqual(other.value);
    }
}
export const vDomain = (domain) => new VDomain(domain);
class VVoid extends BaseValue {
    constructor() {
        super();
        this.type = "Void";
        this.publicName = "Void";
    }
    toString() {
        return "()";
    }
}
export const vVoid = () => new VVoid();
export function isEqual(a, b) {
    if (a.type !== b.type) {
        return false;
    }
    switch (a.type) {
        case "Bool":
        case "Number":
        case "String":
        case "Dist":
        case "Date":
        case "TimeDuration":
        case "Scale":
        case "Domain":
        case "Array":
        case "Dict":
            return a.isEqual(b);
        case "Void":
            return true;
    }
    if (a.toString() !== b.toString()) {
        return false;
    }
    throw new REOther("Equal not implemented for these inputs");
}
const _isUniqableType = (t) => "isEqual" in t;
export function uniq(array) {
    const uniqueArray = [];
    for (const item of array) {
        if (!_isUniqableType(item)) {
            throw new REOther(`Can't apply uniq() to element with type ${item.type}`);
        }
        if (!uniqueArray.some((existingItem) => isEqual(existingItem, item))) {
            uniqueArray.push(item);
        }
    }
    return uniqueArray;
}
export function uniqBy(array, fn) {
    const seen = [];
    const uniqueArray = [];
    for (const item of array) {
        const computed = fn(item);
        if (!_isUniqableType(computed)) {
            throw new REOther(`Can't apply uniq() to element with type ${computed.type}`);
        }
        if (!seen.some((existingItem) => isEqual(existingItem, computed))) {
            seen.push(computed);
            uniqueArray.push(item);
        }
    }
    return uniqueArray;
}
