// @quri/squiggle-lang/dist/dist/SampleSetDist/index.js
import * as Discrete from "../../PointSet/Discrete.js";
import * as XYShape from "../../XYShape.js";
import * as E_A_Floats from "../../utility/E_A_Floats.js";
import * as E_A_Sorted from "../../utility/E_A_Sorted.js";
import * as Result from "../../utility/result.js";
import { ContinuousShape } from "../../PointSet/Continuous.js";
import { DiscreteShape } from "../../PointSet/Discrete.js";
import { buildMixedShape } from "../../PointSet/Mixed.js";
import { BaseDist } from "../BaseDist.js";
import { distOperationError, otherError, tooFewSamplesForConversionToPointSet, } from "../DistError.js";
import { PointSetDist } from "../PointSetDist.js";
import { samplesToPointSetDist } from "./samplesToPointSetDist.js";
import { isEqual } from "../../utility/E_A.js";
export class SampleSetDist extends BaseDist {
    constructor(samples) {
        super();
        this.samples = samples;
    }
    static make(a) {
        if (a.length > 5) {
            return Result.Ok(new SampleSetDist(a));
        }
        else {
            return Result.Err({ type: "TooFewSamples" });
        }
    }
    static fromFn(fn, env) {
        const samples = [];
        for (let i = 0; i < env.sampleCount; i++) {
            samples.push(fn(i));
        }
        return SampleSetDist.make(samples);
    }
    toString() {
        return "Sample Set Distribution";
    }
    _isEqual(other) {
        if (this.samples === other.samples) {
            return true;
        }
        else {
            return isEqual(this.samples, other.samples);
        }
    }
    toSparkline(bucketCount, env) {
        return Result.bind(this.toPointSetDist({
            xyPointLength: bucketCount * 3,
            sampleCount: env.sampleCount,
        }), (r) => r.toSparkline(bucketCount));
    }
    static fromDist(d, env) {
        return SampleSetDist.make(d.sampleN(env.sampleCount));
    }
    integralSum() {
        return 1;
    }
    normalize() {
        return this;
    }
    min() {
        return Math.min(...this.samples);
    }
    max() {
        return Math.max(...this.samples);
    }
    mean() {
        return E_A_Floats.mean(this.samples);
    }
    abs() {
        return new SampleSetDist(this.samples.map(Math.abs));
    }
    truncate(leftCutoff, rightCutoff) {
        let truncated = this.samples;
        if (leftCutoff !== undefined) {
            truncated = truncated.filter((x) => x >= leftCutoff);
        }
        if (rightCutoff !== undefined) {
            truncated = truncated.filter((x) => x <= rightCutoff);
        }
        return Result.bind(SampleSetDist.make(truncated), (dist) => SampleSetDist.make(dist.sampleN(this.samples.length)));
    }
    sample() {
        const index = Math.floor(Math.random() * this.samples.length);
        return this.samples[index];
    }
    sampleN(n) {
        if (n <= this.samples.length) {
            return this.samples.slice(0, n);
        }
        else {
            const result = [];
            for (let i = 1; i <= n; i++) {
                result.push(this.sample());
            }
            return result;
        }
    }
    cdf(f) {
        const countBelowF = this.samples.reduce((acc, x) => acc + (x <= f ? 1 : 0), 0);
        return countBelowF / this.samples.length;
    }
    inv(f) {
        const sorted = E_A_Floats.sort(this.samples);
        return E_A_Sorted.percentile(sorted, f);
    }
    pdf(f, opts) {
        const pointSetDistR = this.toPointSetDist(opts.env);
        if (!pointSetDistR.ok) {
            return pointSetDistR;
        }
        return pointSetDistR.value.pdf(f);
    }
    variance() {
        return Result.Ok(E_A_Floats.variance(this.samples));
    }
    mode() {
        return Result.Err(otherError("Not implemented, https://github.com/quantified-uncertainty/squiggle/issues/1392"));
    }
    range(pWidth, absolute = true) {
        if (pWidth < 0 || pWidth > 1) {
            return Result.Err(otherError("pWidth must be between 0 and 1"));
        }
        const dist = absolute ? this.abs() : this;
        return Result.Ok({
            low: dist.inv(0.5 - pWidth / 2),
            high: dist.inv(0.5 + pWidth / 2),
        });
    }
    toPointSetDist(env) {
        const dists = samplesToPointSetDist({
            samples: this.samples,
            continuousOutputLength: env.xyPointLength,
            kernelWidth: undefined,
        });
        const result = buildMixedShape({
            continuous: dists.continuousDist
                ? new ContinuousShape({ xyShape: dists.continuousDist })
                : undefined,
            discrete: new DiscreteShape({ xyShape: dists.discreteDist }),
        });
        if (!result) {
            return Result.Err(tooFewSamplesForConversionToPointSet());
        }
        return Result.Ok(new PointSetDist(result));
    }
    samplesMap(fn) {
        return buildSampleSetFromFn(this.samples.length, (i) => fn(this.samples[i]));
    }
}
const buildSampleSetFromFn = (n, fn) => {
    const samples = [];
    for (let i = 0; i < n; i++) {
        const result = fn(i);
        if (!result.ok) {
            return Result.Err(distOperationError(result.value));
        }
        samples.push(result.value);
    }
    return SampleSetDist.make(samples);
};
export const map2 = ({ fn, t1, t2, }) => {
    const length = Math.min(t1.samples.length, t2.samples.length);
    return buildSampleSetFromFn(length, (i) => fn(t1.samples[i], t2.samples[i]));
};
export const map3 = ({ fn, t1, t2, t3, }) => {
    const length = Math.min(t1.samples.length, t2.samples.length, t3.samples.length);
    return buildSampleSetFromFn(length, (i) => fn(t1.samples[i], t2.samples[i], t3.samples[i]));
};
export const mapN = ({ fn, t1, }) => {
    const length = Math.max(...t1.map((t) => t.samples.length));
    return buildSampleSetFromFn(length, (i) => fn(t1
        .map((t) => (i < t.samples.length ? t.samples[i] : undefined))
        .filter((v) => v !== undefined)));
};
export const mixture = (values, intendedLength) => {
    const dists = values.map((pair) => pair[0]);
    const totalWeight = values.reduce((acc, v) => acc + v[1], 0);
    const discreteSamples = Discrete.sampleN(new DiscreteShape({
        xyShape: XYShape.T.fromZippedArray(values.map(([, weight], i) => [i, weight / totalWeight])),
    }), intendedLength);
    const samples = discreteSamples.map((distIndexToChoose, index) => {
        const chosenDist = dists[distIndexToChoose];
        if (chosenDist.samples.length < index) {
            throw new Error("Mixture unreachable error");
        }
        return chosenDist.samples[index];
    });
    return SampleSetDist.make(samples);
};
export const minOfTwo = (t1, t2) => {
    return map2({ fn: (a, b) => Result.Ok(Math.min(a, b)), t1, t2 });
};
export const maxOfTwo = (t1, t2) => {
    return map2({ fn: (a, b) => Result.Ok(Math.max(a, b)), t1, t2 });
};
export const minOfFloat = (t, f) => {
    return t.samplesMap((a) => Result.Ok(Math.min(a, f)));
};
export const maxOfFloat = (t, f) => {
    return t.samplesMap((a) => Result.Ok(Math.max(a, f)));
};
