// @quri/squiggle-lang/dist/dist/SampleSetDist/logKde.js
import { kde } from "./kde.js";
const logSamples = (s) => s.map(Math.log);
const reverseNegativeSamples = (s) => s.map(Math.abs).slice().reverse();
const unLogShape = ({ xs, ys }) => {
    const adjustedXs = xs.map(Math.exp);
    const adjustedYs = ys
        .map((y, index) => y / adjustedXs[index])
        .map((r) => (r < 0 ? 0 : r));
    return { xs: adjustedXs, ys: adjustedYs };
};
const reverseNegativeShape = ({ xs, ys }) => ({
    xs: xs
        .slice()
        .reverse()
        .map((x) => -x),
    ys: ys.slice().reverse(),
});
const logKdeSingleSide = ({ samples, weight, outputLength, kernelWidth, sign, }) => {
    let adjustedSamples = samples;
    if (sign === "negative") {
        adjustedSamples = reverseNegativeSamples(adjustedSamples);
    }
    adjustedSamples = logSamples(adjustedSamples);
    const { xs, ys } = kde({
        samples: adjustedSamples,
        outputLength: Math.round(outputLength),
        weight,
        kernelWidth,
    });
    let result = unLogShape({ xs, ys });
    if (sign === "negative") {
        result = reverseNegativeShape(result);
    }
    return result;
};
export const logKde = ({ samples, outputLength, weight, kernelWidth, }) => {
    const singleSideKde = (sign) => {
        let sideSamples = sign === "positive"
            ? samples.filter((v) => v > 0)
            : samples.filter((v) => v < 0);
        if (sideSamples.length < 2) {
            sideSamples = [];
        }
        const proportion = sideSamples.length / samples.length;
        return logKdeSingleSide({
            samples: sideSamples,
            weight: weight * proportion,
            outputLength: Math.round(outputLength * proportion),
            kernelWidth,
            sign,
        });
    };
    const [positivePart, negativePart] = [
        singleSideKde("positive"),
        singleSideKde("negative"),
    ];
    const result = {
        xs: [...negativePart.xs, ...positivePart.xs],
        ys: [...negativePart.ys, ...positivePart.ys],
    };
    return result;
};
