// @quri/squiggle-lang/dist/dist/SampleSetDist/samplesToPointSetDist.js
import { kde } from "./kde.js";
import { splitContinuousAndDiscrete, continuousAreSameFilter, minContinuousSamplesFilter, } from "./splitContinuousAndDiscrete.js";
import * as E_A_Floats from "../../utility/E_A_Floats.js";
import { logKde } from "./logKde.js";
import { SampleSetDist } from "./index.js";
import sum from "lodash/sum.js";
const minDiscreteToKeep = (samples) => Math.max(20, Math.round(samples.length / 50));
const MIN_SAMPLES_FOR_KDE = 5;
const shouldBeLogHeuristicFn = ({ p10, p90 }) => {
    return Math.log(p90 / p10) > 3;
};
const checkIfShouldBeLog = (samples) => {
    const dist = SampleSetDist.make(samples);
    if (dist.ok) {
        const range = dist.value.range(0.8, true);
        if (range.ok) {
            const { low, high } = range.value;
            return shouldBeLogHeuristicFn({ p10: low, p90: high });
        }
    }
    return false;
};
export const samplesToPointSetDist = ({ samples: unsortedSamples, continuousOutputLength, kernelWidth, logScale, }) => {
    const samples = E_A_Floats.sort(unsortedSamples);
    const { continuousSamples, discreteShape } = minContinuousSamplesFilter(MIN_SAMPLES_FOR_KDE, continuousAreSameFilter(splitContinuousAndDiscrete(samples, minDiscreteToKeep(samples))));
    const relevantSampleLength = continuousSamples.length + sum(discreteShape.ys);
    const pointWeight = 1 / relevantSampleLength;
    let continuousDist;
    if (continuousSamples.length > 0) {
        const kdeParams = {
            samples: continuousSamples,
            outputLength: continuousOutputLength,
            weight: pointWeight,
            kernelWidth,
        };
        if (logScale === undefined) {
            logScale = checkIfShouldBeLog(continuousSamples);
        }
        continuousDist = logScale ? logKde(kdeParams) : kde(kdeParams);
    }
    return {
        continuousDist,
        discreteDist: {
            xs: discreteShape.xs,
            ys: discreteShape.ys.map((y) => y * pointWeight),
        },
    };
};
