// @quri/squiggle-lang/dist/dist/SampleSetDist/splitContinuousAndDiscrete.js
export const splitContinuousAndDiscrete = (sortedArray, minDiscreteWeight) => {
    const continuous = [];
    const discreteCount = [];
    const discreteValue = [];
    if (!Number.isInteger(minDiscreteWeight)) {
        throw new Error(`Minimum discrete weight must be an integer. Got ${minDiscreteWeight}`);
    }
    if (minDiscreteWeight < 2) {
        throw new Error("Minimum discrete weight must be at least 2");
    }
    const minDistance = minDiscreteWeight - 1;
    const len = sortedArray.length;
    let i = 0;
    while (i < len - minDistance) {
        const value = sortedArray[i];
        if (value !== sortedArray[i + minDistance]) {
            continuous.push(value);
            i++;
        }
        else {
            const iOrig = i;
            let base = minDistance;
            const isEqualAt = (ind) => ind < len && sortedArray[ind] === value;
            while (isEqualAt(iOrig + base * 2)) {
                base *= 2;
            }
            let lo = iOrig + base;
            i = Math.min(lo + base, len);
            while (i - lo > 1) {
                const mid = lo + Math.floor((i - lo) / 2);
                if (sortedArray[mid] === value) {
                    lo = mid;
                }
                else {
                    i = mid;
                }
            }
            discreteValue.push(value);
            discreteCount.push(i - iOrig);
        }
    }
    continuous.push(...sortedArray.slice(i));
    return {
        continuousSamples: continuous,
        discreteShape: { xs: discreteValue, ys: discreteCount },
    };
};
export const continuousAreSameFilter = ({ continuousSamples, discreteShape: { xs, ys }, }) => {
    const allContinuousAreSame = continuousSamples.length > 0 &&
        continuousSamples[0] === continuousSamples[continuousSamples.length - 1];
    if (allContinuousAreSame) {
        return {
            continuousSamples: [],
            discreteShape: {
                xs: [...xs, continuousSamples[0]],
                ys: [...ys, continuousSamples.length],
            },
        };
    }
    return { continuousSamples, discreteShape: { xs, ys } };
};
export const minContinuousSamplesFilter = (minSampleCount, { continuousSamples, discreteShape }) => {
    if (continuousSamples.length < minSampleCount) {
        return { continuousSamples: [], discreteShape };
    }
    return { continuousSamples, discreteShape };
};
