// @quri/squiggle-lang/dist/src/js/SqError.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SqFrame = exports.SqError = void 0;
const RSError = __importStar(require("../rescript/SqError.gen"));
const RSFrameStack = __importStar(require("../rescript/Reducer/Reducer_FrameStack.gen"));
class SqError {
    constructor(_value) {
        this._value = _value;
    }
    toString() {
        return RSError.toString(this._value);
    }
    toStringWithStackTrace() {
        return RSError.toStringWithStackTrace(this._value);
    }
    static createOtherError(v) {
        return new SqError(RSError.createOtherError(v));
    }
    getTopFrame() {
        const frame = RSFrameStack.getTopFrame(RSError.getFrameStack(this._value));
        return frame ? new SqFrame(frame) : undefined;
    }
    getFrameArray() {
        const frames = RSError.getFrameArray(this._value);
        return frames.map((frame) => new SqFrame(frame));
    }
    location() {
        var _a;
        return (_a = this.getTopFrame()) === null || _a === void 0 ? void 0 : _a.location();
    }
}
exports.SqError = SqError;
class SqFrame {
    constructor(_value) {
        this._value = _value;
    }
    name() {
        return RSFrameStack.Frame.getName(this._value);
    }
    location() {
        return RSFrameStack.Frame.getLocation(this._value);
    }
}
exports.SqFrame = SqFrame;
