// @quri/squiggle-lang/dist/src/js/SqProject.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SqProject = void 0;
const RSProject = __importStar(require("../rescript/ForTS/ForTS_ReducerProject.gen"));
const SqError_1 = require("./SqError");
const SqRecord_1 = require("./SqRecord");
const SqValue_1 = require("./SqValue");
const types_1 = require("./types");
const SqValueLocation_1 = require("./SqValueLocation");
class SqProject {
    constructor(_value) {
        this._value = _value;
    }
    static create() {
        return new SqProject(RSProject.createProject());
    }
    getSourceIds() {
        return RSProject.getSourceIds(this._value);
    }
    setSource(sourceId, value) {
        return RSProject.setSource(this._value, sourceId, value);
    }
    removeSource(sourceId) {
        RSProject.removeSource(this._value, sourceId);
    }
    getSource(sourceId) {
        return RSProject.getSource(this._value, sourceId);
    }
    touchSource(sourceId) {
        return RSProject.touchSource(this._value, sourceId);
    }
    clean(sourceId) {
        return RSProject.clean(this._value, sourceId);
    }
    cleanAll() {
        return RSProject.cleanAll(this._value);
    }
    cleanResults(sourceId) {
        return RSProject.cleanResults(this._value, sourceId);
    }
    cleanAllResults() {
        return RSProject.cleanAllResults(this._value);
    }
    getIncludes(sourceId) {
        return (0, types_1.resultMap2)(RSProject.getIncludes(this._value, sourceId), (a) => a, (v) => new SqError_1.SqError(v));
    }
    getContinues(sourceId) {
        return RSProject.getContinues(this._value, sourceId);
    }
    setContinues(sourceId, continues) {
        return RSProject.setContinues(this._value, sourceId, continues);
    }
    getRunOrder() {
        return RSProject.getRunOrder(this._value);
    }
    getRunOrderFor(sourceId) {
        return RSProject.getRunOrderFor(this._value, sourceId);
    }
    parseIncludes(sourceId) {
        return RSProject.parseIncludes(this._value, sourceId);
    }
    run(sourceId) {
        return RSProject.run(this._value, sourceId);
    }
    runAll() {
        return RSProject.runAll(this._value);
    }
    getBindings(sourceId) {
        return new SqRecord_1.SqRecord(RSProject.getBindings(this._value, sourceId), new SqValueLocation_1.SqValueLocation(this, sourceId, {
            root: "bindings",
            items: [],
        }));
    }
    getResult(sourceId) {
        const innerResult = RSProject.getResult(this._value, sourceId);
        return (0, types_1.resultMap2)(innerResult, (v) => (0, SqValue_1.wrapValue)(v, new SqValueLocation_1.SqValueLocation(this, sourceId, {
            root: "result",
            items: [],
        })), (v) => new SqError_1.SqError(v));
    }
    setEnvironment(environment) {
        RSProject.setEnvironment(this._value, environment);
    }
    getEnvironment() {
        return RSProject.getEnvironment(this._value);
    }
}
exports.SqProject = SqProject;
