// @quri/squiggle-lang/src/rescript/Reducer/Reducer_Context.bs.js
'use strict';

var E_O$QuriSquiggleLang = require("../Utility/E/E_O.bs.js");
var Reducer_T$QuriSquiggleLang = require("./Reducer_T.bs.js");
var Reducer_Bindings$QuriSquiggleLang = require("./Reducer_Bindings.bs.js");
var Reducer_Lambda_T$QuriSquiggleLang = require("./Reducer_Lambda_T.bs.js");
var DistributionOperation$QuriSquiggleLang = require("../Distributions/DistributionOperation.bs.js");

function createContext(stdLib, environment) {
  return {
          bindings: Reducer_Bindings$QuriSquiggleLang.extend(Reducer_Bindings$QuriSquiggleLang.fromNamespace(stdLib)),
          environment: environment,
          frameStack: /* [] */0,
          inFunction: undefined
        };
}

function currentFunctionName(t) {
  return E_O$QuriSquiggleLang.$$default(E_O$QuriSquiggleLang.fmap(t.inFunction, Reducer_Lambda_T$QuriSquiggleLang.name), Reducer_T$QuriSquiggleLang.topFrameName);
}

var defaultEnvironment = DistributionOperation$QuriSquiggleLang.defaultEnv;

exports.defaultEnvironment = defaultEnvironment;
exports.createContext = createContext;
exports.currentFunctionName = currentFunctionName;
/* Reducer_Bindings-QuriSquiggleLang Not a pure module */
