// @quri/squiggle-lang/src/rescript/Reducer/Reducer_Bindings.bs.js
'use strict';

var Belt_MapString = require("@rescript/std/lib/js/belt_MapString.js");
var Reducer_Namespace$QuriSquiggleLang = require("./Reducer_Namespace.bs.js");

function get(_param, id) {
  while(true) {
    var param = _param;
    var parent = param.parent;
    var v = Reducer_Namespace$QuriSquiggleLang.get(param.namespace, id);
    if (v !== undefined) {
      return v;
    }
    if (parent === undefined) {
      return ;
    }
    _param = parent;
    continue ;
  };
}

function set(bindings, id, value) {
  return {
          namespace: Reducer_Namespace$QuriSquiggleLang.set(bindings.namespace, id, value),
          parent: bindings.parent
        };
}

function toString(param) {
  var parent = param.parent;
  var pairs = Reducer_Namespace$QuriSquiggleLang.toString(param.namespace);
  if (parent !== undefined) {
    return "{" + pairs + "} / " + toString(parent) + "";
  } else {
    return "{" + pairs + "}";
  }
}

function extend(bindings) {
  return {
          namespace: Reducer_Namespace$QuriSquiggleLang.make(undefined),
          parent: bindings
        };
}

function make(param) {
  return {
          namespace: Reducer_Namespace$QuriSquiggleLang.make(undefined),
          parent: undefined
        };
}

function removeResult(bindings) {
  return {
          namespace: Belt_MapString.remove(bindings.namespace, "__result__"),
          parent: bindings.parent
        };
}

function locals(param) {
  return param.namespace;
}

function fromNamespace(namespace) {
  return {
          namespace: namespace,
          parent: undefined
        };
}

exports.get = get;
exports.set = set;
exports.toString = toString;
exports.extend = extend;
exports.make = make;
exports.removeResult = removeResult;
exports.locals = locals;
exports.fromNamespace = fromNamespace;
/* Reducer_Namespace-QuriSquiggleLang Not a pure module */
