// @quri/squiggle-lang/src/rescript/Reducer/Reducer_T.gen.tsx
/* TypeScript file generated from Reducer_T.res by genType. */
/* eslint-disable import/first */


// tslint:disable-next-line:max-classes-per-file 
// tslint:disable-next-line:class-name
export abstract class value { protected opaque!: any }; /* simulate opaque types */

// tslint:disable-next-line:max-classes-per-file 
// tslint:disable-next-line:class-name
export abstract class arrayValue { protected opaque!: any }; /* simulate opaque types */

// tslint:disable-next-line:max-classes-per-file 
// tslint:disable-next-line:class-name
export abstract class map { protected opaque!: any }; /* simulate opaque types */

// tslint:disable-next-line:max-classes-per-file 
// tslint:disable-next-line:class-name
export abstract class lambdaValue { protected opaque!: any }; /* simulate opaque types */

// tslint:disable-next-line:max-classes-per-file 
// tslint:disable-next-line:class-name
export abstract class lambdaDeclaration { protected opaque!: any }; /* simulate opaque types */

// tslint:disable-next-line:max-classes-per-file 
// tslint:disable-next-line:class-name
export abstract class frame { protected opaque!: any }; /* simulate opaque types */

// tslint:disable-next-line:max-classes-per-file 
// tslint:disable-next-line:class-name
export abstract class frameStack { protected opaque!: any }; /* simulate opaque types */
