// @quri/squiggle-lang/src/rescript/Reducer/Reducer_FrameStack.gen.tsx
/* TypeScript file generated from Reducer_FrameStack.res by genType. */
/* eslint-disable import/first */


// @ts-ignore: Implicit any on import
const Reducer_FrameStackBS = require('./Reducer_FrameStack.bs');

import type {frameStack as Reducer_T_frameStack} from './Reducer_T.gen';

import type {frame as Reducer_T_frame} from './Reducer_T.gen';

import type {location as Reducer_Peggy_Parse_location} from '../../../src/rescript/Reducer/Reducer_Peggy/Reducer_Peggy_Parse.gen';

// tslint:disable-next-line:interface-over-type-literal
export type t = Reducer_T_frameStack;

export const Frame_getLocation: (t:Reducer_T_frame) => (null | undefined | Reducer_Peggy_Parse_location) = Reducer_FrameStackBS.Frame.getLocation;

export const Frame_getName: (t:Reducer_T_frame) => string = Reducer_FrameStackBS.Frame.getName;

export const toFrameArray: (t:t) => Reducer_T_frame[] = Reducer_FrameStackBS.toFrameArray;

export const getTopFrame: (t:t) => (null | undefined | Reducer_T_frame) = Reducer_FrameStackBS.getTopFrame;

export const Frame: { getLocation: (t:Reducer_T_frame) => (null | undefined | Reducer_Peggy_Parse_location); getName: (t:Reducer_T_frame) => string } = Reducer_FrameStackBS.Frame
