// @quri/squiggle-lang/dist/public/SqProject/index.js
import { isBindingStatement } from "../../ast/utils.js";
import { defaultEnv } from "../../dist/env.js";
import * as Library from "../../library/index.js";
import { createContext } from "../../reducer/context.js";
import { ImmutableMap } from "../../utility/immutableMap.js";
import * as Result from "../../utility/result.js";
import { vDict } from "../../value/index.js";
import { SqOtherError } from "../SqError.js";
import { SqDict } from "../SqValue/SqDict.js";
import { wrapValue } from "../SqValue/index.js";
import { SqValueContext } from "../SqValueContext.js";
import { SqValuePath } from "../SqValuePath.js";
import { ProjectItem } from "./ProjectItem.js";
function getNeedToRunError() {
    return new SqOtherError("Need to run");
}
export class SqProject {
    constructor(options) {
        this.inverseGraph = new Map();
        this.items = new Map();
        this.stdLib = options?.stdLib ?? Library.getStdLib();
        this.environment = options?.environment ?? defaultEnv;
        this.linker = options?.linker;
    }
    static create(options) {
        return new SqProject(options);
    }
    getEnvironment() {
        return this.environment;
    }
    setEnvironment(environment) {
        this.environment = environment;
    }
    getStdLib() {
        return this.stdLib;
    }
    getSourceIds() {
        return Array.from(this.items.keys());
    }
    getItem(sourceId) {
        const item = this.items.get(sourceId);
        if (!item) {
            throw new Error(`Source ${sourceId} not found`);
        }
        return item;
    }
    cleanDependents(initialSourceId) {
        const visited = new Set();
        const inner = (currentSourceId) => {
            visited.add(currentSourceId);
            if (currentSourceId !== initialSourceId) {
                this.clean(currentSourceId);
            }
            for (const sourceId of this.getDependents(currentSourceId)) {
                if (visited.has(sourceId)) {
                    continue;
                }
                inner(sourceId);
            }
        };
        inner(initialSourceId);
    }
    getDependents(sourceId) {
        return [...(this.inverseGraph.get(sourceId)?.values() ?? [])];
    }
    getDependencies(sourceId) {
        this.parseImports(sourceId);
        return this.getItem(sourceId).getDependencies();
    }
    removeImportEdges(fromSourceId) {
        const item = this.getItem(fromSourceId);
        if (item.imports?.ok) {
            for (const importData of item.imports.value) {
                this.inverseGraph.get(importData.sourceId)?.delete(fromSourceId);
            }
        }
    }
    touchSource(sourceId) {
        this.removeImportEdges(sourceId);
        this.getItem(sourceId).touchSource();
        this.cleanDependents(sourceId);
    }
    setSource(sourceId, value) {
        if (this.items.has(sourceId)) {
            this.removeImportEdges(sourceId);
            this.getItem(sourceId).setSource(value);
            this.cleanDependents(sourceId);
        }
        else {
            this.items.set(sourceId, new ProjectItem({ sourceId, source: value }));
        }
    }
    removeSource(sourceId) {
        if (!this.items.has(sourceId)) {
            return;
        }
        this.cleanDependents(sourceId);
        this.removeImportEdges(sourceId);
        this.setContinues(sourceId, []);
        this.items.delete(sourceId);
    }
    getSource(sourceId) {
        return this.items.get(sourceId)?.source;
    }
    clean(sourceId) {
        this.getItem(sourceId).clean();
    }
    cleanAll() {
        this.getSourceIds().forEach((id) => this.clean(id));
    }
    getImportIds(sourceId) {
        const imports = this.getImports(sourceId);
        if (!imports) {
            return Result.Err(getNeedToRunError());
        }
        return Result.fmap(imports, (imports) => imports.map((i) => i.sourceId));
    }
    getImports(sourceId) {
        return this.getItem(sourceId).imports;
    }
    getContinues(sourceId) {
        return this.getItem(sourceId).continues;
    }
    setContinues(sourceId, continues) {
        for (const continueId of this.getContinues(sourceId)) {
            this.inverseGraph.get(continueId)?.delete(sourceId);
        }
        for (const continueId of continues) {
            if (!this.inverseGraph.has(continueId)) {
                this.inverseGraph.set(continueId, new Set());
            }
            this.inverseGraph.get(continueId)?.add(sourceId);
        }
        this.getItem(sourceId).setContinues(continues);
        this.cleanDependents(sourceId);
    }
    getInternalOutput(sourceId) {
        return this.getItem(sourceId).output ?? Result.Err(getNeedToRunError());
    }
    parseImports(sourceId) {
        const item = this.getItem(sourceId);
        if (item.imports) {
            return;
        }
        item.parseImports(this.linker);
        for (const dependencyId of item.getDependencies()) {
            if (!this.inverseGraph.has(dependencyId)) {
                this.inverseGraph.set(dependencyId, new Set());
            }
            this.inverseGraph.get(dependencyId)?.add(sourceId);
        }
    }
    getOutput(sourceId) {
        const internalOutputR = this.getInternalOutput(sourceId);
        if (!internalOutputR.ok) {
            return internalOutputR;
        }
        const source = this.getSource(sourceId);
        if (source === undefined) {
            throw new Error("Internal error: source not found");
        }
        const astR = this.getItem(sourceId).ast;
        if (!astR) {
            throw new Error("Internal error: AST is missing when result is ok");
        }
        if (!astR.ok) {
            return astR;
        }
        const ast = astR.value;
        const lastStatement = ast.statements.at(-1);
        const hasEndExpression = !!lastStatement && !isBindingStatement(lastStatement);
        const result = wrapValue(internalOutputR.value.result, new SqValueContext({
            project: this,
            sourceId,
            source,
            ast,
            valueAst: hasEndExpression ? lastStatement : ast,
            valueAstIsPrecise: hasEndExpression,
            path: new SqValuePath({
                root: "result",
                items: [],
            }),
        }));
        const [bindings, exports] = ["bindings", "exports"].map((field) => new SqDict(internalOutputR.value[field], new SqValueContext({
            project: this,
            sourceId,
            source,
            ast: astR.value,
            valueAst: astR.value,
            valueAstIsPrecise: true,
            path: new SqValuePath({
                root: "bindings",
                items: [],
            }),
        })));
        return Result.Ok({ result, bindings, exports });
    }
    getResult(sourceId) {
        return Result.fmap(this.getOutput(sourceId), ({ result }) => result);
    }
    getBindings(sourceId) {
        return Result.fmap(this.getOutput(sourceId), ({ bindings }) => bindings);
    }
    async buildExternals(sourceId, pendingIds) {
        this.parseImports(sourceId);
        const rImports = this.getImports(sourceId);
        if (!rImports) {
            throw new Error("Internal logic error");
        }
        if (!rImports.ok) {
            return rImports;
        }
        let externals = ImmutableMap().merge(this.getStdLib());
        for (const importBinding of [
            ...this.getItem(sourceId).getImplicitImports(),
            ...rImports.value,
        ]) {
            if (!this.items.has(importBinding.sourceId)) {
                if (!this.linker) {
                    throw new Error(`Can't load source for ${importBinding.sourceId}, linker is missing`);
                }
                let newSource;
                try {
                    newSource = await this.linker.loadSource(importBinding.sourceId);
                }
                catch (e) {
                    return Result.Err(new SqOtherError(`Failed to load import ${importBinding.sourceId}`));
                }
                this.setSource(importBinding.sourceId, newSource);
            }
            if (pendingIds.has(importBinding.sourceId)) {
                return Result.Err(new SqOtherError(`Cyclic import ${importBinding.sourceId}`));
            }
            await this.innerRun(importBinding.sourceId, pendingIds);
            const outputR = this.getInternalOutput(importBinding.sourceId);
            if (!outputR.ok) {
                return outputR;
            }
            switch (importBinding.type) {
                case "flat":
                    externals = externals.merge(outputR.value.bindings);
                    break;
                case "named":
                    externals = externals.set(importBinding.variable, vDict(outputR.value.exports));
                    break;
                default:
                    throw new Error(`Internal error, ${importBinding}`);
            }
        }
        return Result.Ok(externals);
    }
    async innerRun(sourceId, pendingIds) {
        pendingIds.add(sourceId);
        const cachedOutput = this.getItem(sourceId).output;
        if (!cachedOutput) {
            const rExternals = await this.buildExternals(sourceId, pendingIds);
            if (!rExternals.ok) {
                this.getItem(sourceId).failRun(rExternals.value);
            }
            else {
                const context = createContext(this.getEnvironment());
                await this.getItem(sourceId).run(context, rExternals.value);
            }
        }
        pendingIds.delete(sourceId);
    }
    async run(sourceId) {
        await this.innerRun(sourceId, new Set());
    }
    findValuePathByOffset(sourceId, offset) {
        const { ast } = this.getItem(sourceId);
        if (!ast) {
            return Result.Err(new SqOtherError("Not parsed"));
        }
        if (!ast.ok) {
            return ast;
        }
        const found = SqValuePath.findByOffset({
            ast: ast.value,
            offset,
        });
        if (!found) {
            return Result.Err(new SqOtherError("Not found"));
        }
        return Result.Ok(found);
    }
}
