// @quri/squiggle-lang/dist/public/SqProject/ProjectItem.js
import { parse } from "../../ast/parse.js";
import { ICompileError, IRuntimeError } from "../../errors/IError.js";
import { compileAst } from "../../expression/compile.js";
import { evaluate } from "../../reducer/index.js";
import * as Result from "../../utility/result.js";
import { Ok } from "../../utility/result.js";
import { SqCompileError, SqRuntimeError } from "../SqError.js";
export class ProjectItem {
    constructor(props) {
        this.sourceId = props.sourceId;
        this.source = props.source;
        this.continues = [];
    }
    touchSource() {
        this.ast = undefined;
        this.imports = undefined;
        this.output = undefined;
    }
    setSource(source) {
        this.source = source;
        this.touchSource();
    }
    setAst(ast) {
        this.ast = ast;
        this.imports = undefined;
        this.output = undefined;
    }
    setImports(imports) {
        this.imports = imports;
        this.output = undefined;
    }
    clean() {
        this.output = undefined;
    }
    getDependencies() {
        if (!this.imports?.ok) {
            return this.continues;
        }
        return [...this.continues, ...this.imports.value.map((i) => i.sourceId)];
    }
    getImplicitImports() {
        const implicitImports = [];
        for (const continueId of this.continues) {
            implicitImports.push({
                type: "flat",
                sourceId: continueId,
            });
        }
        return implicitImports;
    }
    setContinues(continues) {
        this.continues = continues;
        this.clean();
    }
    parseImports(linker) {
        if (this.imports) {
            return;
        }
        this.buildAst();
        if (!this.ast) {
            throw new Error("Internal logic error");
        }
        if (!this.ast.ok) {
            this.setImports(this.ast);
            return;
        }
        const program = this.ast.value;
        const resolvedImports = [];
        for (const [file, variable] of program.imports) {
            const location = file.location;
            if (!linker) {
                this.setImports(Result.Err(new SqCompileError(new ICompileError("Can't use imports when linker is not configured", location))));
                return;
            }
            try {
                const sourceId = linker.resolve(file.value, this.sourceId);
                resolvedImports.push({
                    type: "named",
                    variable: variable.value,
                    sourceId,
                });
            }
            catch (e) {
                this.setImports(Result.Err(new SqCompileError(new ICompileError(`SqLinker.resolve has failed to resolve ${file.value}`, location))));
                return;
            }
        }
        this.setImports(Ok(resolvedImports));
    }
    buildAst() {
        if (this.ast) {
            return;
        }
        const ast = Result.errMap(parse(this.source, this.sourceId), (e) => new SqCompileError(e));
        this.setAst(ast);
    }
    failRun(e) {
        this.output = Result.Err(e);
    }
    async run(context, externals) {
        if (this.output) {
            return;
        }
        this.buildAst();
        if (!this.ast) {
            throw new Error("Internal logic error");
        }
        if (!this.ast.ok) {
            this.failRun(this.ast.value);
            return;
        }
        const expression = Result.errMap(compileAst(this.ast.value, externals), (e) => new SqCompileError(e));
        if (!expression.ok) {
            this.failRun(expression.value);
            return;
        }
        try {
            const wrappedEvaluate = context.evaluate;
            const asyncEvaluate = (expression, context) => {
                return wrappedEvaluate(expression, context);
            };
            if (expression.value.type !== "Program") {
                throw new Error("Expected Program expression");
            }
            const [result, contextAfterEvaluation] = evaluate(expression.value, {
                ...context,
                evaluate: asyncEvaluate,
            });
            const bindings = contextAfterEvaluation.stack.asBindings();
            const exportNames = new Set(expression.value.value.exports);
            const exports = bindings.filter((_, key) => exportNames.has(key));
            this.output = Ok({
                result,
                bindings,
                exports,
            });
        }
        catch (e) {
            this.failRun(new SqRuntimeError(IRuntimeError.fromException(e)));
        }
    }
}
