// @quri/squiggle-lang/dist/PointSet/AlgebraicShapeCombination.js
import * as XYShape from "../XYShape.js";
import { convolutionOperationToFn } from "./PointSet.js";
import * as E_A_Floats from "../utility/E_A_Floats.js";
const toDiscretePointMassesFromTriangulars = (s, { inverse } = { inverse: false }) => {
    let n = XYShape.T.length(s);
    const { xs, ys } = s;
    xs.unshift(xs[0]);
    ys.unshift(ys[0]);
    xs.push(xs[n - 1]);
    ys.push(ys[n - 1]);
    n = xs.length;
    const xsSq = new Array(n);
    const xsProdN1 = new Array(n - 1);
    const xsProdN2 = new Array(n - 2);
    for (let i = 0; i <= n - 1; i++) {
        xsSq[i] = xs[i] * xs[i];
    }
    for (let i = 0; i <= n - 2; i++) {
        xsProdN1[i] = xs[i] * xs[i + 1];
    }
    for (let i = 0; i <= n - 3; i++) {
        xsProdN2[i] = xs[i] * xs[i + 2];
    }
    const masses = new Array(n - 2);
    const means = new Array(n - 2);
    const variances = new Array(n - 2);
    if (inverse) {
        for (let i = 1; i <= n - 2; i++) {
            masses[i - 1] = ((xs[i + 1] - xs[i - 1]) * ys[i]) / 2;
            const a = xs[i - 1];
            const c = xs[i];
            const b = xs[i + 1];
            const inverseMean = (2 *
                ((a * Math.log(a / c)) / (a - c) + (b * Math.log(c / b)) / (b - c))) /
                (a - b);
            const inverseVar = (2 * (Math.log(c / a) / (a - c) + (b * Math.log(b / c)) / (b - c))) /
                (a - b) -
                inverseMean ** 2;
            means[i - 1] = inverseMean;
            variances[i - 1] = inverseVar;
        }
        return { n: n - 2, masses, means, variances };
    }
    else {
        for (let i = 1; i <= n - 2; i++) {
            masses[i - 1] = ((xs[i + 1] - xs[i - 1]) * ys[i]) / 2;
            means[i - 1] = (xs[i - 1] + xs[i] + xs[i + 1]) / 3;
            variances[i - 1] =
                (xsSq[i - 1] +
                    xsSq[i] +
                    xsSq[i + 1] -
                    xsProdN1[i - 1] -
                    xsProdN1[i] -
                    xsProdN2[i - 1]) /
                    18;
        }
        return { n: n - 2, masses, means, variances };
    }
};
export const combineShapesContinuousContinuous = (op, s1, s2) => {
    const t1m = toDiscretePointMassesFromTriangulars(s1);
    const t2m = toDiscretePointMassesFromTriangulars(s2, { inverse: false });
    const combineMeansFn = {
        Add: (m1, m2) => m1 + m2,
        Subtract: (m1, m2) => m1 - m2,
        Multiply: (m1, m2) => m1 * m2,
    }[op];
    const combineVariancesFn = {
        Add: (v1, v2) => v1 + v2,
        Subtract: (v1, v2) => v1 + v2,
        Multiply: (v1, v2, m1, m2) => v1 * v2 + v1 * m2 ** 2 + v2 * m1 ** 2,
    }[op];
    let outputMinX = Infinity;
    let outputMaxX = -Infinity;
    const masses = new Array(t1m.n * t2m.n);
    const means = new Array(t1m.n * t2m.n);
    const variances = new Array(t1m.n * t2m.n);
    for (let i = 0; i < t1m.n; i++) {
        for (let j = 0; j < t2m.n; j++) {
            const k = i * t2m.n + j;
            masses[k] = t1m.masses[i] * t2m.masses[j];
            const mean = combineMeansFn(t1m.means[i], t2m.means[j]);
            const variance = combineVariancesFn(t1m.variances[i], t2m.variances[j], t1m.means[i], t2m.means[j]);
            means[k] = mean;
            variances[k] = variance;
            const minX = mean - 2 * Math.sqrt(variance) * 1.644854;
            const maxX = mean + 2 * Math.sqrt(variance) * 1.644854;
            if (minX < outputMinX) {
                outputMinX = minX;
            }
            if (maxX > outputMaxX) {
                outputMaxX = maxX;
            }
        }
    }
    const nOut = 300;
    const outputXs = E_A_Floats.range(outputMinX, outputMaxX, nOut);
    const outputYs = new Array(nOut).fill(0);
    for (let j = 0; j < masses.length; j++) {
        if (variances[j] > 0 &&
            masses[j] > 0) {
            for (let i = 0; i < outputXs.length; i++) {
                const dx = outputXs[i] - means[j];
                const contribution = (masses[j] * Math.exp(-(dx ** 2) / (2 * variances[j]))) /
                    Math.sqrt(2 * 3.14159276 * variances[j]);
                outputYs[i] = outputYs[i] + contribution;
            }
        }
    }
    return { xs: outputXs, ys: outputYs };
};
export const combineShapesContinuousDiscrete = (op, continuousShape, discreteShape, opts) => {
    const t1n = XYShape.T.length(continuousShape);
    const t2n = XYShape.T.length(discreteShape);
    const opFunc = convolutionOperationToFn(op);
    const fn = opts.discretePosition === "First"
        ? (a, b) => opFunc(b, a)
        : opFunc;
    const outXYShapes = new Array(t2n);
    switch (op) {
        case "Add":
        case "Subtract":
            for (let j = 0; j <= t2n - 1; j++) {
                const dxyShape = new Array(t1n);
                for (let i = 0; i <= t1n - 1; i++) {
                    const index = opts.discretePosition === "First" && op === "Subtract"
                        ? t1n - 1 - i
                        : i;
                    dxyShape[index] = [
                        fn(continuousShape.xs[i], discreteShape.xs[j]),
                        continuousShape.ys[i] * discreteShape.ys[j],
                    ];
                }
                outXYShapes[j] = dxyShape;
            }
            break;
        case "Multiply":
            for (let j = 0; j <= t2n - 1; j++) {
                const dxyShape = new Array(t1n);
                for (let i = 0; i <= t1n - 1; i++) {
                    const index = discreteShape.xs[j] > 0 ? i : t1n - 1 - i;
                    dxyShape[index] = [
                        fn(continuousShape.xs[i], discreteShape.xs[j]),
                        (continuousShape.ys[i] * discreteShape.ys[j]) /
                            Math.abs(discreteShape.xs[j]),
                    ];
                }
                outXYShapes[j] = dxyShape;
            }
            break;
        default:
            throw new Error(`Unknown operation ${op}`);
    }
    return outXYShapes
        .map(XYShape.T.fromZippedArray)
        .reduce((acc, x) => XYShape.PointwiseCombination.addCombine(XYShape.XtoY.continuousInterpolator("Linear", "UseZero"), acc, x), XYShape.T.empty);
};
export const isOrdered = (a) => {
    return E_A_Floats.isSorted(a.xs);
};
