// @quri/squiggle-lang/dist/PointSet/Mixed.js
import * as XYShape from "../XYShape.js";
import * as Continuous from "./Continuous.js";
import * as Discrete from "./Discrete.js";
import * as MixedPoint from "./MixedPoint.js";
import * as Result from "../utility/result.js";
import * as Common from "./Common.js";
import { ContinuousShape } from "./Continuous.js";
import { DiscreteShape } from "./Discrete.js";
export class MixedShape {
    constructor(args) {
        this.continuous = args.continuous;
        this.discrete = args.discrete;
        this._integralSumCache = args.integralSumCache;
        this._integralCache = args.integralCache;
    }
    get integralCache() {
        return this._integralCache;
    }
    get integralSumCache() {
        return this._integralSumCache;
    }
    withAdjustedIntegralSum(integralSumCache) {
        return new MixedShape({
            continuous: this.continuous,
            discrete: this.discrete,
            integralSumCache,
            integralCache: this.integralCache,
        });
    }
    minX() {
        return Math.min(this.continuous.minX(), this.discrete.minX());
    }
    maxX() {
        return Math.max(this.continuous.maxX(), this.discrete.maxX());
    }
    isEmpty() {
        return this.continuous.isEmpty() && this.discrete.isEmpty();
    }
    toContinuous() {
        return this.continuous;
    }
    toDiscrete() {
        return this.discrete;
    }
    toMixed() {
        return this;
    }
    isEqual(other) {
        return (this.continuous.isEqual(other.continuous) &&
            this.discrete.isEqual(other.discrete));
    }
    truncate(leftCutoff, rightCutoff) {
        return new MixedShape({
            continuous: this.continuous.truncate(leftCutoff, rightCutoff),
            discrete: this.discrete.truncate(leftCutoff, rightCutoff),
        });
    }
    normalize() {
        if (this.isEmpty()) {
            return this;
        }
        if (this.continuous.isEmpty()) {
            return this.discrete.normalize().toMixed();
        }
        if (this.discrete.isEmpty()) {
            return this.continuous.normalize().toMixed();
        }
        const continuousIntegralSum = this.continuous.integralSum();
        const discreteIntegralSum = this.discrete.integralSum();
        const totalIntegralSum = continuousIntegralSum + discreteIntegralSum;
        const newContinuousSum = continuousIntegralSum / totalIntegralSum;
        const newDiscreteSum = discreteIntegralSum / totalIntegralSum;
        const normalizedContinuous = this.continuous
            .scaleBy(newContinuousSum / continuousIntegralSum)
            .withAdjustedIntegralSum(newContinuousSum);
        const normalizedDiscrete = this.discrete
            .scaleBy(newDiscreteSum / discreteIntegralSum)
            .withAdjustedIntegralSum(newDiscreteSum);
        return new MixedShape({
            continuous: normalizedContinuous,
            discrete: normalizedDiscrete,
            integralSumCache: 1,
        });
    }
    xToY(x) {
        const { continuous, discrete } = this.normalize();
        const c = continuous.xToY(x);
        const d = discrete.xToY(x);
        return MixedPoint.add(c, d);
    }
    toDiscreteProbabilityMassFraction() {
        const discreteIntegralSum = this.discrete.integralSum();
        const continuousIntegralSum = this.continuous.integralSum();
        const totalIntegralSum = discreteIntegralSum + continuousIntegralSum;
        return discreteIntegralSum / totalIntegralSum;
    }
    downsample(count) {
        const discreteIntegralSum = this.discrete.integralSum();
        const continuousIntegralSum = this.continuous.integralSum();
        const totalIntegralSum = discreteIntegralSum + continuousIntegralSum;
        const downsampledDiscrete = this.discrete.downsample(Math.floor((count * discreteIntegralSum) / totalIntegralSum));
        const downsampledContinuous = this.continuous.downsample(Math.floor((count * continuousIntegralSum) / totalIntegralSum));
        return new MixedShape({
            continuous: downsampledContinuous,
            discrete: downsampledDiscrete,
            integralSumCache: this.integralSumCache,
            integralCache: this.integralCache,
        });
    }
    integral() {
        if (!this._integralCache) {
            const continuousIntegral = this.continuous.integral();
            const discreteIntegral = Continuous.stepwiseToLinear(this.discrete.integral());
            this._integralCache = new ContinuousShape({
                xyShape: XYShape.PointwiseCombination.addCombine(XYShape.XtoY.continuousInterpolator("Linear", "UseOutermostPoints"), continuousIntegral.xyShape, discreteIntegral.xyShape),
            });
        }
        return this._integralCache;
    }
    integralSum() {
        return (this._integralSumCache ??= this.integral().lastY());
    }
    integralXtoY(f) {
        return XYShape.XtoY.linear(this.integral().xyShape, f);
    }
    integralYtoX(f) {
        return XYShape.YtoX.linear(this.integral().xyShape, f);
    }
    mapY(fn, integralSumCacheFn, integralCacheFn) {
        const discrete = this.discrete.mapY(fn, integralSumCacheFn, integralCacheFn);
        const continuous = this.continuous.mapY(fn, integralSumCacheFn, integralCacheFn);
        return new MixedShape({
            discrete,
            continuous,
            integralSumCache: this.integralSumCache === undefined
                ? undefined
                : integralSumCacheFn?.(this.integralSumCache),
            integralCache: this.integralCache === undefined
                ? undefined
                : integralCacheFn?.(this.integralCache),
        });
    }
    mapYResult(fn, integralSumCacheFn, integralCacheFn) {
        const discreteResult = this.discrete.mapYResult(fn, integralSumCacheFn, integralCacheFn);
        const continuousResult = this.continuous.mapYResult(fn, integralSumCacheFn, integralCacheFn);
        if (!continuousResult.ok) {
            return continuousResult;
        }
        if (!discreteResult.ok) {
            return discreteResult;
        }
        const continuous = continuousResult.value;
        const discrete = discreteResult.value;
        return Result.Ok(new MixedShape({
            discrete,
            continuous,
            integralSumCache: this.integralSumCache === undefined
                ? undefined
                : integralSumCacheFn?.(this.integralSumCache),
            integralCache: this.integralCache === undefined
                ? undefined
                : integralCacheFn?.(this.integralCache),
        }));
    }
    mean() {
        const discreteMean = this.discrete.mean();
        const continuousMean = this.continuous.mean();
        return ((discreteMean * this.discrete.integralSum() +
            continuousMean * this.continuous.integralSum()) /
            this.integralSum());
    }
    variance() {
        const discreteIntegralSum = this.discrete.integralSum();
        const continuousIntegralSum = this.continuous.integralSum();
        const totalIntegralSum = discreteIntegralSum + continuousIntegralSum;
        const getMeanOfSquares = ({ discrete, continuous }) => {
            const discreteMean = discrete.shapeMap(XYShape.T.square).mean();
            const continuousMean = continuous.getMeanOfSquares();
            return ((discreteMean * discreteIntegralSum +
                continuousMean * continuousIntegralSum) /
                totalIntegralSum);
        };
        switch (discreteIntegralSum / totalIntegralSum) {
            case 1:
                return this.discrete.variance();
            case 0:
                return this.continuous.variance();
            default:
                return XYShape.Analysis.getVarianceDangerously(this, (t) => t.mean(), (t) => getMeanOfSquares(t));
        }
    }
}
export const combineAlgebraically = (op, t1, t2) => {
    const ccConvResult = Continuous.combineAlgebraically(op, t1.continuous, t2.continuous);
    const dcConvResult = Continuous.combineAlgebraicallyWithDiscrete(op, t2.continuous, t1.discrete, "First");
    const cdConvResult = Continuous.combineAlgebraicallyWithDiscrete(op, t1.continuous, t2.discrete, "Second");
    const continuousConvResult = Continuous.sum([
        ccConvResult,
        dcConvResult,
        cdConvResult,
    ]);
    const discreteConvResult = Discrete.combineAlgebraically(op, t1.discrete, t2.discrete);
    const combinedIntegralSum = Common.combineIntegralSums((a, b) => a * b, t1.integralSumCache, t2.integralSumCache);
    return new MixedShape({
        discrete: discreteConvResult,
        continuous: continuousConvResult,
        integralSumCache: combinedIntegralSum,
        integralCache: undefined,
    });
};
export const combinePointwise = (t1, t2, fn, integralSumCachesFn = () => undefined, integralCachesFn = () => undefined) => {
    const reducedDiscrete = Discrete.combinePointwise(t1.toDiscrete(), t2.toDiscrete(), fn, integralSumCachesFn);
    const reducedContinuous = Continuous.combinePointwise(t1.toContinuous(), t2.toContinuous(), fn, undefined, integralSumCachesFn);
    const combinedIntegralSum = Common.combineIntegralSums(integralSumCachesFn, t1.integralSumCache, t2.integralSumCache);
    const combinedIntegral = Common.combineIntegrals(integralCachesFn, t1.integralCache, t2.integralCache);
    return Result.fmap(Result.merge(reducedContinuous, reducedDiscrete), ([continuous, discrete]) => new MixedShape({
        continuous,
        discrete,
        integralSumCache: combinedIntegralSum,
        integralCache: combinedIntegral,
    }));
};
export function buildMixedShape({ continuous, discrete, }) {
    continuous ??= new ContinuousShape({
        integralSumCache: 0,
        xyShape: { xs: [], ys: [] },
    });
    discrete ??= new DiscreteShape({
        integralSumCache: 0,
        xyShape: { xs: [], ys: [] },
    });
    const cLength = continuous.xyShape.xs.length;
    const dLength = discrete.xyShape.xs.length;
    if (cLength < 2 && dLength == 0) {
        return undefined;
    }
    else {
        return new MixedShape({ continuous, discrete });
    }
}
