// @quri/squiggle-lang/dist/PointSet/PointSetDist_Scoring.js
import { ComplexNumberError, PdfInvalidError, } from "../operationError.js";
import * as Result from "../utility/result.js";
import * as Mixed from "./Mixed.js";
const logFn = Math.log;
function minusScaledLogOfQuotient({ esti, answ, }) {
    const quot = esti / answ;
    return quot < 0.0
        ? Result.Err(new ComplexNumberError())
        : Result.Ok(-answ * logFn(quot));
}
export const WithDistAnswer = {
    integrand(estimateElement, answerElement) {
        if (answerElement === 0) {
            return Result.Ok(0);
        }
        else if (estimateElement === 0) {
            return Result.Ok(Infinity);
        }
        else {
            return minusScaledLogOfQuotient({
                esti: estimateElement,
                answ: answerElement,
            });
        }
    },
    sum({ estimate, answer, }) {
        return Result.fmap(Mixed.combinePointwise(estimate.toMixed(), answer.toMixed(), WithDistAnswer.integrand), (t) => t.integralSum());
    },
    sumWithPrior({ estimate, answer, prior, }) {
        const kl1 = WithDistAnswer.sum({ estimate, answer });
        const kl2 = WithDistAnswer.sum({ estimate: prior, answer });
        return Result.fmap(Result.merge(kl1, kl2), ([v1, v2]) => v1 - v2);
    },
};
export const WithScalarAnswer = {
    sum(mp) {
        return mp.continuous + mp.discrete;
    },
    score({ estimate, answer, }) {
        const _score = (estimatePdf, answer) => {
            const density = estimatePdf(answer);
            if (density === undefined) {
                return Result.Err(new PdfInvalidError());
            }
            else if (density < 0) {
                return Result.Err(new PdfInvalidError());
            }
            else if (density === 0) {
                return Result.Ok(Infinity);
            }
            else {
                return Result.Ok(-logFn(density));
            }
        };
        const estimatePdf = (x) => {
            if (estimate.toContinuous().isEmpty()) {
                return WithScalarAnswer.sum(estimate.toDiscrete().xToY(x));
            }
            else if (estimate.toDiscrete().isEmpty()) {
                return WithScalarAnswer.sum(estimate.toContinuous().xToY(x));
            }
            else {
                return undefined;
            }
        };
        return _score(estimatePdf, answer);
    },
    scoreWithPrior({ estimate, answer, prior, }) {
        return Result.fmap(Result.merge(WithScalarAnswer.score({ estimate, answer }), WithScalarAnswer.score({ estimate: prior, answer })), ([s1, s2]) => s1 - s2);
    },
};
