// @quri/squiggle-lang/dist/PointSet/PointSet.js
import * as Continuous from "../PointSet/Continuous.js";
import * as Discrete from "../PointSet/Discrete.js";
import * as Mixed from "../PointSet/Mixed.js";
import { ContinuousShape } from "./Continuous.js";
import { DiscreteShape } from "./Discrete.js";
export const convolutionOperationToFn = (op) => {
    return {
        Add: (x, y) => x + y,
        Multiply: (x, y) => x * y,
        Subtract: (x, y) => x - y,
    }[op];
};
export const combinePointwise = (t1, t2, fn, integralSumCachesFn = () => undefined, integralCachesFn = () => undefined) => {
    if (t1 instanceof ContinuousShape && t2 instanceof ContinuousShape) {
        return Continuous.combinePointwise(t1, t2, fn, undefined, integralSumCachesFn);
    }
    else if (t1 instanceof DiscreteShape && t2 instanceof DiscreteShape) {
        return Discrete.combinePointwise(t1, t2, fn, integralSumCachesFn);
    }
    else {
        return Mixed.combinePointwise(t1.toMixed(), t2.toMixed(), fn, integralSumCachesFn, integralCachesFn);
    }
};
export const isContinuous = (d) => d instanceof ContinuousShape;
export const isDiscrete = (d) => d instanceof DiscreteShape;
