// @quri/squiggle-lang/src/rescript/Distributions/SymbolicDist/SymbolicDistTypes.gen.tsx
/* TypeScript file generated from SymbolicDistTypes.res by genType. */
/* eslint-disable import/first */


// tslint:disable-next-line:interface-over-type-literal
export type normal = { readonly mean: number; readonly stdev: number };

// tslint:disable-next-line:interface-over-type-literal
export type lognormal = { readonly mu: number; readonly sigma: number };

// tslint:disable-next-line:interface-over-type-literal
export type uniform = { readonly low: number; readonly high: number };

// tslint:disable-next-line:interface-over-type-literal
export type beta = { readonly alpha: number; readonly beta: number };

// tslint:disable-next-line:interface-over-type-literal
export type exponential = { readonly rate: number };

// tslint:disable-next-line:interface-over-type-literal
export type cauchy = { readonly local: number; readonly scale: number };

// tslint:disable-next-line:interface-over-type-literal
export type triangular = {
  readonly low: number; 
  readonly medium: number; 
  readonly high: number
};

// tslint:disable-next-line:interface-over-type-literal
export type gamma = { readonly shape: number; readonly scale: number };

// tslint:disable-next-line:interface-over-type-literal
export type logistic = { readonly location: number; readonly scale: number };

// tslint:disable-next-line:interface-over-type-literal
export type bernoulli = { readonly p: number };

// tslint:disable-next-line:interface-over-type-literal
export type symbolicDist = 
    { NAME: "Normal"; VAL: normal }
  | { NAME: "Beta"; VAL: beta }
  | { NAME: "Lognormal"; VAL: lognormal }
  | { NAME: "Uniform"; VAL: uniform }
  | { NAME: "Exponential"; VAL: exponential }
  | { NAME: "Cauchy"; VAL: cauchy }
  | { NAME: "Triangular"; VAL: triangular }
  | { NAME: "Gamma"; VAL: gamma }
  | { NAME: "Float"; VAL: number }
  | { NAME: "Bernoulli"; VAL: bernoulli }
  | { NAME: "Logistic"; VAL: logistic };
